// koishi-plugin-chatluna/lib/llm-core/memory/message/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/llm-core/memory/message/index.ts
var message_exports = {};
__export(message_exports, {
  KoishiChatMessageHistory: () => KoishiChatMessageHistory
});
module.exports = __toCommonJS(message_exports);

// src/llm-core/memory/message/database_history.ts
var import_messages2 = require("@langchain/core/messages");
var import_chat_history = require("@langchain/core/chat_history");
var import_string2 = require("koishi-plugin-chatluna/utils/string");
var import_crypto = require("crypto");

// src/llm-core/agent/legacy-executor.ts
var import_messages = require("@langchain/core/messages");
var import_tool2 = require("@langchain/core/messages/tool");
var import_output_parsers = require("@langchain/core/output_parsers");
var import_runnables = require("@langchain/core/runnables");
var import_tools = require("@langchain/core/tools");
var import_koishi_plugin_chatluna2 = require("koishi-plugin-chatluna");
var import_base2 = require("koishi-plugin-chatluna/llm-core/chain/base");

// src/llm-core/agent/tool-observation.ts
var import_tool = require("@langchain/core/messages/tool");
var import_koishi_plugin_chatluna = require("koishi-plugin-chatluna");
var import_langchain = require("koishi-plugin-chatluna/utils/langchain");
var import_zod = require("zod");
function observationToMessageContent(observation) {
  return (0, import_tool.isDirectToolOutput)(observation) ? "" : observation;
}
__name(observationToMessageContent, "observationToMessageContent");

// src/llm-core/chain/infinite_context_chain.ts
var import_prompts = require("@langchain/core/prompts");
var import_base = require("koishi-plugin-chatluna/llm-core/chain/base");
var import_string = require("koishi-plugin-chatluna/utils/string");
var COMPRESS_PROMPT = import_prompts.PromptTemplate.fromTemplate(`You are a helpful AI assistant tasked with summarizing conversations.

When asked to summarize, provide a detailed but concise summary of the conversation.
Focus on information that would be helpful for continuing the conversation, including:
- What was done
- What is currently being worked on
- Which files are being modified
- What needs to be done next
- Key user requests, constraints, or preferences that should persist
- Important technical decisions and why they were made
- Tool calls that were made and their results (summarize the key outcomes)

Additional user instructions for this compression:
{instruction}

Some old tool result messages may say that the original tool output expired and was removed.
Treat those as intentional retention placeholders, not as meaningful tool output.

Your summary should be comprehensive enough to provide context but concise enough to be quickly understood.

Do not respond to any questions in the conversation, only output the summary.

Conversation:
{conversation_chunk}`);

// src/llm-core/memory/message/database_history.ts
var KoishiChatMessageHistory = class extends import_chat_history.BaseChatMessageHistory {
  constructor(ctx, conversationId, _maxMessagesCount, chatluna) {
    super();
    this._maxMessagesCount = _maxMessagesCount;
    this.chatluna = chatluna;
    this.conversationId = conversationId;
    this._ctx = ctx;
    this._chatHistory = [];
    this._additional_kwargs = {};
    this._updatedAt = /* @__PURE__ */ new Date(0);
  }
  static {
    __name(this, "KoishiChatMessageHistory");
  }
  // eslint-disable-next-line @typescript-eslint/naming-convention
  lc_namespace = ["llm-core", "memory", "message"];
  conversationId;
  _ctx;
  _latestId;
  _serializedChatHistory;
  _chatHistory;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  _additional_kwargs;
  _updatedAt;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  get additionalArgs() {
    return this._additional_kwargs;
  }
  async getMessages() {
    const latestUpdateTime = await this.getLatestUpdateTime();
    if (latestUpdateTime > this._updatedAt || this._chatHistory.length === 0) {
      this._chatHistory = await this._loadMessages();
    }
    return this._chatHistory;
  }
  async addUserMessage(message) {
    const humanMessage = new import_messages2.HumanMessage(message);
    await this.addMessage(humanMessage);
  }
  async addAIChatMessage(message) {
    const aiMessage = new import_messages2.AIMessage(message);
    await this.addMessage(aiMessage);
  }
  async addMessage(message) {
    await this.addMessages([message]);
  }
  async addMessages(messages) {
    if (messages.length === 0) {
      return;
    }
    await this.loadConversation();
    const serializedMessages = [];
    let parentId = this._latestId;
    for (const message of messages) {
      const serializedMessage = await serializeMessage(
        message,
        this.conversationId,
        parentId
      );
      serializedMessages.push(serializedMessage);
      parentId = serializedMessage.id;
    }
    await this._ctx.database.upsert("chatluna_message", serializedMessages);
    this._serializedChatHistory.push(...serializedMessages);
    this._chatHistory.push(...messages);
    this._latestId = serializedMessages[serializedMessages.length - 1].id;
    const updatedAt = /* @__PURE__ */ new Date();
    await this._trimMessages();
    this._updatedAt = updatedAt;
    await this._saveConversation(updatedAt);
  }
  async addAgentToolBatch(steps) {
    if (steps.length === 0) {
      return;
    }
    await this.addMessages(createAgentToolMessages(steps));
  }
  async replaceMessages(messages) {
    await this.loadConversation();
    const serializedMessages = [];
    let parentId = null;
    for (const message of messages) {
      const serializedMessage = await serializeMessage(
        message,
        this.conversationId,
        parentId
      );
      serializedMessages.push(serializedMessage);
      parentId = serializedMessage.id;
    }
    await this._ctx.database.remove("chatluna_message", {
      conversationId: this.conversationId
    });
    if (serializedMessages.length > 0) {
      await this._ctx.database.upsert(
        "chatluna_message",
        serializedMessages
      );
    }
    this._serializedChatHistory = serializedMessages;
    this._chatHistory = [...messages];
    this._latestId = serializedMessages[serializedMessages.length - 1]?.id ?? null;
    const updatedAt = /* @__PURE__ */ new Date();
    this._updatedAt = updatedAt;
    await this._saveConversation(updatedAt);
  }
  async clear() {
    await this._ctx.database.remove("chatluna_message", {
      conversationId: this.conversationId
    });
    await this._ctx.database.upsert("chatluna_conversation", [
      {
        id: this.conversationId,
        latestMessageId: null,
        updatedAt: /* @__PURE__ */ new Date()
      }
    ]);
    this._serializedChatHistory = [];
    this._chatHistory = [];
    this._latestId = null;
  }
  async delete() {
    await this._ctx.database.remove("chatluna_conversation", {
      id: this.conversationId
    });
  }
  async updateAdditionalArg(key, value) {
    await this.loadConversation();
    this._additional_kwargs[key] = value;
    await this._saveConversation();
  }
  async getAdditionalArg(key) {
    await this.loadConversation();
    return this._additional_kwargs[key];
  }
  async getAdditionalArgs() {
    await this.loadConversation();
    return this._additional_kwargs;
  }
  async deleteAdditionalArg(key) {
    await this.loadConversation();
    delete this._additional_kwargs[key];
    await this._saveConversation();
  }
  async removeAllToolAndFunctionMessages() {
    const messages = await this.getMessages();
    const filtered = messages.filter((message) => {
      const type = message.getType();
      return type !== "tool" && type !== "function";
    });
    if (filtered.length === messages.length) {
      return messages;
    }
    await this.replaceMessages(filtered);
    return filtered;
  }
  async overrideAdditionalArgs(kwargs) {
    await this.loadConversation();
    this._additional_kwargs = Object.assign(this._additional_kwargs, kwargs);
    await this._saveConversation();
  }
  async getLatestUpdateTime() {
    const conversation = (await this._ctx.database.get(
      "chatluna_conversation",
      {
        id: this.conversationId
      },
      ["updatedAt"]
    ))?.[0];
    return conversation?.updatedAt ?? /* @__PURE__ */ new Date(0);
  }
  async _loadMessages() {
    const queried = await this._ctx.database.get("chatluna_message", {
      conversationId: this.conversationId
    });
    const sorted = [];
    let currentMessageId = this._latestId;
    let isBad = false;
    const seen = /* @__PURE__ */ new Set();
    if (currentMessageId == null && queried.length > 0) {
      isBad = true;
    }
    while (currentMessageId != null && !isBad) {
      if (seen.has(currentMessageId)) {
        isBad = true;
        break;
      }
      seen.add(currentMessageId);
      const currentMessage = queried.find(
        (item) => item.id === currentMessageId
      );
      if (!currentMessage) {
        isBad = true;
        break;
      }
      sorted.unshift(currentMessage);
      currentMessageId = currentMessage.parentId;
    }
    if (isBad) {
      this._ctx.logger.warn(
        `Bad conversation detected for %s`,
        this.conversationId
      );
      sorted.length = 0;
      await this.clear();
    }
    this._serializedChatHistory = sorted;
    const promises = sorted.map(async (item) => {
      const args = JSON.parse(
        item.additional_kwargs_binary ? await (0, import_string2.gzipDecode)(item.additional_kwargs_binary) : "{}"
      );
      const responseMetadata = JSON.parse(
        item.response_metadata_binary ? await (0, import_string2.gzipDecode)(item.response_metadata_binary) : "{}"
      );
      responseMetadata.chatluna = {
        ...responseMetadata.chatluna ?? {},
        recordId: item.id,
        createdAt: item.createdAt?.toISOString()
      };
      let content;
      try {
        content = JSON.parse(
          item.content ? await (0, import_string2.gzipDecode)(item.content) : item.text
        );
      } catch {
        this._ctx.logger.warn(
          `Failed to deserialize message content for %s in %s, using fallback text.`,
          item.id,
          this.conversationId
        );
        content = typeof item.text === "string" ? item.text : "";
      }
      const fields = {
        content,
        id: item.rawId ?? void 0,
        name: item.name ?? void 0,
        tool_calls: item.tool_calls ?? void 0,
        tool_call_id: item.tool_call_id ?? void 0,
        response_metadata: responseMetadata,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        additional_kwargs: args
      };
      if (item.role === "system") {
        return new import_messages2.SystemMessage(fields);
      } else if (item.role === "human") {
        return new import_messages2.HumanMessage(fields);
      } else if (item.role === "ai") {
        return new import_messages2.AIMessage(fields);
      } else if (item.role === "function") {
        return new import_messages2.FunctionMessage(fields);
      } else if (item.role === "tool") {
        return new import_messages2.ToolMessage(fields);
      } else {
        throw new Error("Unknown role");
      }
    });
    return await Promise.all(promises);
  }
  async _loadConversation() {
    const conversation = (await this._ctx.database.get("chatluna_conversation", {
      id: this.conversationId
    }))?.[0];
    if (conversation) {
      this._latestId = conversation.latestMessageId ?? null;
      this._additional_kwargs = conversation.additional_kwargs != null ? JSON.parse(conversation.additional_kwargs) : {};
    } else {
      await this._ctx.database.create("chatluna_conversation", {
        id: this.conversationId,
        bindingKey: this.conversationId,
        title: "Conversation",
        model: this.chatluna.config.defaultModel,
        preset: this.chatluna.config.defaultPreset,
        chatMode: this.chatluna.config.defaultChatMode,
        createdBy: "system",
        createdAt: /* @__PURE__ */ new Date(),
        updatedAt: /* @__PURE__ */ new Date(),
        lastChatAt: /* @__PURE__ */ new Date(),
        status: "active",
        latestMessageId: null,
        additional_kwargs: null,
        compression: null,
        archivedAt: null,
        archiveId: null,
        legacyRoomId: null,
        legacyMeta: null,
        autoTitle: true
      });
    }
    if (!this._serializedChatHistory) {
      await this._loadMessages();
      this._updatedAt = conversation?.updatedAt ?? /* @__PURE__ */ new Date(0);
    }
  }
  async loadConversation() {
    if (!this._serializedChatHistory) {
      await this._loadConversation();
    }
  }
  async _trimMessages() {
    if (this._serializedChatHistory.length > this._maxMessagesCount) {
      const toDeleted = this._serializedChatHistory.splice(
        0,
        this._serializedChatHistory.length - this._maxMessagesCount
      );
      while (this._serializedChatHistory[0] != null && ["ai", "function", "tool"].includes(
        this._serializedChatHistory[0].role
      )) {
        const message = this._serializedChatHistory.shift();
        if (message) {
          toDeleted.push(message);
        }
      }
      await this._ctx.database.remove("chatluna_message", {
        id: toDeleted.map((item) => item.id)
      });
      const firstMessage = this._serializedChatHistory[0];
      this._latestId = this._serializedChatHistory[this._serializedChatHistory.length - 1]?.id ?? null;
      if (firstMessage) {
        firstMessage.parentId = null;
        await this._ctx.database.upsert("chatluna_message", [
          firstMessage
        ]);
      }
      this._chatHistory = await this._loadMessages();
    }
  }
  async _saveConversation(time = /* @__PURE__ */ new Date()) {
    const hasKwargs = this._additional_kwargs && Object.keys(this._additional_kwargs).length > 0;
    await this._ctx.database.upsert("chatluna_conversation", [
      {
        id: this.conversationId,
        latestMessageId: this._latestId,
        additional_kwargs: hasKwargs ? JSON.stringify(this._additional_kwargs) : null,
        updatedAt: time
      }
    ]);
  }
};
async function serializeMessage(message, conversationId, parentId) {
  const meta = readMessageMeta(message);
  const id = meta.recordId ?? (0, import_crypto.randomUUID)();
  const createdAt = meta.createdAt ? new Date(meta.createdAt) : /* @__PURE__ */ new Date();
  writeMessageMeta(message, {
    recordId: id,
    createdAt: createdAt.toISOString()
  });
  let additionalArgs = Object.assign({}, message.additional_kwargs);
  delete additionalArgs["preset"];
  delete additionalArgs["raw_content"];
  delete additionalArgs["type"];
  if (Object.keys(additionalArgs).length === 0) {
    additionalArgs = null;
  }
  let responseMetadata = Object.assign({}, message.response_metadata);
  if (Object.keys(responseMetadata).length === 0) {
    responseMetadata = null;
  }
  return {
    id,
    content: await (0, import_string2.gzipEncode)(JSON.stringify(message.content)).then(
      (buf) => (0, import_string2.bufferToArrayBuffer)(buf)
    ),
    parentId: parentId ?? null,
    role: message.getType(),
    name: message.name,
    tool_calls: message["tool_calls"],
    tool_call_id: message["tool_call_id"],
    additional_kwargs_binary: additionalArgs && Object.keys(additionalArgs).length > 0 ? await (0, import_string2.gzipEncode)(JSON.stringify(additionalArgs)).then(
      (buf) => (0, import_string2.bufferToArrayBuffer)(buf)
    ) : null,
    response_metadata_binary: responseMetadata && Object.keys(responseMetadata).length > 0 ? await (0, import_string2.gzipEncode)(JSON.stringify(responseMetadata)).then(
      (buf) => (0, import_string2.bufferToArrayBuffer)(buf)
    ) : null,
    rawId: message.id ?? null,
    conversationId,
    createdAt
  };
}
__name(serializeMessage, "serializeMessage");
function createAgentToolMessages(steps) {
  const reasoning = steps[0]?.action.reasoningContent;
  const message = steps[0]?.action.messageLog?.[0];
  return [
    new import_messages2.AIMessage({
      content: "",
      additional_kwargs: {
        ...message?.additional_kwargs ?? {},
        ...reasoning != null ? { reasoning_content: reasoning } : {}
      },
      tool_calls: steps.map((step) => ({
        id: step.action.toolCallId,
        name: step.action.tool,
        args: typeof step.action.toolInput !== "string" ? step.action.toolInput : { input: step.action.toolInput }
      }))
    }),
    ...steps.map(
      (step) => new import_messages2.ToolMessage({
        content: observationToMessageContent(step.observation),
        tool_call_id: step.action.toolCallId,
        name: step.action.tool
      })
    )
  ];
}
__name(createAgentToolMessages, "createAgentToolMessages");
function readMessageMeta(message) {
  const meta = message.response_metadata?.chatluna;
  return {
    recordId: typeof meta?.recordId === "string" && meta.recordId.length > 0 ? meta.recordId : void 0,
    createdAt: typeof meta?.createdAt === "string" && meta.createdAt.length > 0 ? meta.createdAt : void 0
  };
}
__name(readMessageMeta, "readMessageMeta");
function writeMessageMeta(message, meta) {
  message.response_metadata = {
    ...message.response_metadata ?? {},
    chatluna: {
      ...message.response_metadata?.chatluna ?? {},
      ...meta
    }
  };
}
__name(writeMessageMeta, "writeMessageMeta");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  KoishiChatMessageHistory
});
