// koishi-plugin-chatluna/lib/chains/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/chains/index.ts
var chains_exports = {};
__export(chains_exports, {
  ChainMiddleware: () => ChainMiddleware,
  ChainMiddlewareRunStatus: () => ChainMiddlewareRunStatus,
  ChatChain: () => ChatChain
});
module.exports = __toCommonJS(chains_exports);

// src/chains/chain.ts
var import_koishi = require("koishi");
var import_error = require("koishi-plugin-chatluna/utils/error");
var import_logger = require("koishi-plugin-chatluna/utils/logger");

// src/middlewares/system/lifecycle.ts
var lifecycleNames = [
  "lifecycle-check",
  "lifecycle-prepare",
  "lifecycle-handle_command",
  "lifecycle-request_conversation",
  "lifecycle-send"
];

// src/utils/time.ts
function formatDuration(ms) {
  if (ms < 1e3) {
    return `${ms}ms`;
  }
  const totalSeconds = Math.floor(ms / 1e3);
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor(totalSeconds % 3600 / 60);
  const seconds = totalSeconds % 60;
  if (hours > 0) {
    return `${hours}h ${minutes}m ${seconds}s`;
  }
  if (minutes > 0) {
    return `${minutes}m ${seconds}s`;
  }
  const decimal = (ms / 1e3).toFixed(2);
  return `${decimal}s`;
}
__name(formatDuration, "formatDuration");

// src/chains/chain.ts
var logger;
var ChatChain = class {
  constructor(ctx, config) {
    this.ctx = ctx;
    this.config = config;
    logger = (0, import_logger.createLogger)(ctx);
    this._graph = new ChatChainDependencyGraph();
    this._senders = [];
    const defaultChatChainSender = new DefaultChatChainSender(config);
    this._senders.push(
      (session, messages, context) => defaultChatChainSender.send(session, messages, context)
    );
  }
  static {
    __name(this, "ChatChain");
  }
  _graph;
  _senders;
  isSetErrorMessage = false;
  _createRecallThinkingMessage(context) {
    return async () => {
      if (!context.options?.thinkingTimeoutObject) return;
      const timeoutObj = context.options.thinkingTimeoutObject;
      clearTimeout(timeoutObj.timeout);
      timeoutObj.autoRecallTimeout && clearTimeout(timeoutObj.autoRecallTimeout);
      timeoutObj.recallFunc && await timeoutObj.recallFunc();
      timeoutObj.timeout = null;
      context.options.thinkingTimeoutObject = void 0;
    };
  }
  async receiveMessage(session, ctx) {
    const context = {
      config: this.config,
      message: session.content,
      ctx: ctx ?? this.ctx,
      session,
      options: {
        startedAt: Date.now()
      },
      send: /* @__PURE__ */ __name((message) => this.sendMessage(session, message, context), "send"),
      recallThinkingMessage: this._createRecallThinkingMessage(
        {}
      )
    };
    context.recallThinkingMessage = this._createRecallThinkingMessage(context);
    const result = await this._runMiddleware(session, context);
    await context.recallThinkingMessage();
    return result;
  }
  async receiveCommand(session, command, options = {}, ctx = this.ctx) {
    const context = {
      config: this.config,
      message: options?.message ?? session.content,
      ctx,
      session,
      command,
      send: /* @__PURE__ */ __name((message) => this.sendMessage(session, message, context), "send"),
      recallThinkingMessage: this._createRecallThinkingMessage(
        {}
      ),
      options: {
        ...options,
        startedAt: Date.now()
      }
    };
    context.recallThinkingMessage = this._createRecallThinkingMessage(context);
    const result = await this._runMiddleware(session, context);
    await context.recallThinkingMessage();
    return result;
  }
  middleware(name, middleware, ctx = this.ctx) {
    const result = new ChainMiddleware(name, middleware, this._graph);
    this._graph.addNode(result);
    const dispose = /* @__PURE__ */ __name(() => this._graph.removeNode(name), "dispose");
    ctx.effect(() => dispose);
    return result;
  }
  sender(sender) {
    this._senders.push(sender);
  }
  async _runMiddleware(session, context) {
    if (!this.isSetErrorMessage) {
      (0, import_error.setErrorFormatTemplate)(session.text("chatluna.error_message"));
      this.isSetErrorMessage = true;
    }
    const originMessage = context.message;
    const runLevels = this._graph.build();
    if (runLevels.length === 0) {
      return false;
    }
    let isOutputLog = false;
    for (const level of runLevels) {
      const results = await this._executeLevel(level, session, context);
      for (const result of results) {
        if (result.status === "stop") {
          await this._handleStopStatus(
            session,
            context,
            originMessage,
            isOutputLog
          );
          return false;
        }
        if (result.status === "error") {
          await this._handleMiddlewareError(
            session,
            result.middlewareName,
            result.error
          );
          return false;
        }
        if (result.output instanceof Array || typeof result.output === "string") {
          context.message = result.output;
        }
        if (result.shouldLog) {
          isOutputLog = true;
        }
      }
    }
    if (isOutputLog) {
      logger.debug("-".repeat(40) + "\n");
    }
    if (context.message != null && context.message !== originMessage) {
      await this.sendMessage(session, context.message, context);
    }
    return true;
  }
  async _executeLevel(middlewares, session, context) {
    const abortController = new AbortController();
    const results = [];
    let hasStopRequest = false;
    let hasError = false;
    const promises = middlewares.map(async (middleware, index) => {
      try {
        if (abortController.signal.aborted) {
          return {
            status: "success",
            output: 0 /* SKIPPED */,
            middlewareName: middleware.name,
            shouldLog: false
          };
        }
        const result = await this._executeMiddleware(
          middleware,
          session,
          context,
          abortController.signal
        );
        if (result.status === "stop" && !hasStopRequest) {
          hasStopRequest = true;
          abortController.abort();
        }
        if (result.status === "error" && !hasError) {
          hasError = true;
          abortController.abort();
        }
        results[index] = result;
        return result;
      } catch (error) {
        const errorResult = {
          status: "error",
          error,
          middlewareName: middleware.name,
          shouldLog: false
        };
        if (!hasError) {
          hasError = true;
          abortController.abort();
        }
        results[index] = errorResult;
        return errorResult;
      }
    });
    await Promise.all(promises);
    return results.filter((result) => result !== void 0);
  }
  async _executeMiddleware(middleware, session, context, abortSignal) {
    const startTime = Date.now();
    try {
      if (abortSignal?.aborted) {
        return {
          status: "success",
          output: 0 /* SKIPPED */,
          middlewareName: middleware.name,
          shouldLog: false
        };
      }
      const result = await middleware.run(session, context);
      const executionTime = Date.now() - startTime;
      const shouldLogTime = !middleware.name.startsWith("lifecycle-") && result !== 0 /* SKIPPED */ && middleware.name !== "allow_reply" && executionTime > 100;
      if (shouldLogTime) {
        logger.debug(
          `middleware %c executed in %s`,
          middleware.name,
          formatDuration(executionTime)
        );
      }
      if (result === 1 /* STOP */) {
        return {
          status: "stop",
          middlewareName: middleware.name,
          shouldLog: shouldLogTime
        };
      }
      return {
        status: "success",
        output: result,
        middlewareName: middleware.name,
        shouldLog: shouldLogTime
      };
    } catch (error) {
      return {
        status: "error",
        error,
        middlewareName: middleware.name,
        shouldLog: false
      };
    }
  }
  async sendMessage(session, message, context) {
    const messages = message instanceof Array ? message : [message];
    for (const sender of this._senders) {
      await sender(session, messages, context);
    }
  }
  async _handleStopStatus(session, context, originMessage, isOutputLog) {
    if (context.message != null && context.message !== originMessage) {
      await this.sendMessage(session, context.message, context);
    }
    if (isOutputLog) {
      logger.debug("-".repeat(40) + "\n");
    }
  }
  async _handleMiddlewareError(session, middlewareName, error) {
    if (error instanceof import_error.ChatLunaError) {
      const message = error.errorCode === import_error.ChatLunaErrorCode.ABORTED ? session.text("chatluna.aborted") : error.message;
      await this.sendMessage(session, message, void 0);
      return;
    }
    logger.error(`chat-chain: ${middlewareName} error ${error}`);
    logger.error(error);
    error.cause && logger.error(error.cause);
    logger.debug("-".repeat(40) + "\n");
    await this.sendMessage(
      session,
      session.text("chatluna.middleware_error", [
        middlewareName,
        error.message
      ]),
      void 0
    );
  }
};
var ChatChainDependencyGraph = class {
  static {
    __name(this, "ChatChainDependencyGraph");
  }
  _tasks = /* @__PURE__ */ new Map();
  _rules = [];
  _listeners = /* @__PURE__ */ new Map();
  _cachedOrder = null;
  _index = 0;
  addNode(middleware) {
    this._tasks.set(middleware.name, {
      name: middleware.name,
      middleware,
      index: this._index++
    });
    this._cachedOrder = null;
  }
  removeNode(name) {
    this._tasks.delete(name);
    for (let i = this._rules.length - 1; i >= 0; i--) {
      const rule = this._rules[i];
      if (rule.owner === name || rule.target === name) {
        this._rules.splice(i, 1);
      }
    }
    this._cachedOrder = null;
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  once(name, listener) {
    const listeners = this._listeners.get(name) ?? /* @__PURE__ */ new Set();
    listeners.add(listener);
    this._listeners.set(name, listeners);
  }
  before(taskA, taskB) {
    if (taskA instanceof ChainMiddleware) {
      taskA = taskA.name;
    }
    if (taskB instanceof ChainMiddleware) {
      taskB = taskB.name;
    }
    if (taskA && taskB) {
      this._rules.push({
        owner: taskA,
        target: taskB,
        type: "before",
        source: this._captureRuleSource()
      });
      this._cachedOrder = null;
    } else {
      throw new Error("Invalid tasks");
    }
  }
  after(taskA, taskB) {
    if (taskA instanceof ChainMiddleware) {
      taskA = taskA.name;
    }
    if (taskB instanceof ChainMiddleware) {
      taskB = taskB.name;
    }
    if (taskA && taskB) {
      this._rules.push({
        owner: taskA,
        target: taskB,
        type: "after",
        source: this._captureRuleSource()
      });
      this._cachedOrder = null;
    } else {
      throw new Error("Invalid tasks");
    }
  }
  getDependencies(task) {
    const deps = /* @__PURE__ */ new Set();
    for (const rule of this._rules) {
      if (rule.owner === task && rule.type === "after") {
        deps.add(rule.target);
      }
      if (rule.target === task && rule.type === "before") {
        deps.add(rule.owner);
      }
    }
    return deps;
  }
  getDependents(task) {
    const dependents = [];
    for (const rule of this._rules) {
      if (rule.owner === task && rule.type === "before") {
        dependents.push(rule.target);
      }
      if (rule.target === task && rule.type === "after") {
        dependents.push(rule.owner);
      }
    }
    return dependents;
  }
  build() {
    if (this._cachedOrder) {
      return this._cachedOrder;
    }
    for (const [, listeners] of this._listeners) {
      for (const listener of listeners) {
        listener();
      }
      listeners.clear();
    }
    const lifecycleSet = new Set(lifecycleNames);
    const nodes = [...this._tasks.values()].sort(
      (a, b) => a.index - b.index
    );
    const nodeNames = new Set(nodes.map((node) => node.name));
    const normalNodes = nodes.filter((node) => !lifecycleSet.has(node.name));
    const ranges = /* @__PURE__ */ new Map();
    const outgoing = /* @__PURE__ */ new Map();
    const incoming = /* @__PURE__ */ new Map();
    const slots = /* @__PURE__ */ new Map();
    const slotCause = /* @__PURE__ */ new Map();
    for (const node of normalNodes) {
      ranges.set(node.name, {
        min: 0,
        max: lifecycleNames.length,
        minRules: [],
        maxRules: []
      });
      outgoing.set(node.name, []);
      incoming.set(node.name, []);
      slots.set(node.name, 0);
    }
    for (const rule of this._rules) {
      if (!nodeNames.has(rule.owner)) {
        continue;
      }
      if (!lifecycleSet.has(rule.target) && !nodeNames.has(rule.target)) {
        throw new Error(
          `Unknown middleware "${rule.target}" referenced by ${this._formatRule(rule)}`
        );
      }
      if (lifecycleSet.has(rule.owner) && lifecycleSet.has(rule.target)) {
        continue;
      }
      if (lifecycleSet.has(rule.target) || lifecycleSet.has(rule.owner)) {
        const name = lifecycleSet.has(rule.target) ? rule.owner : rule.target;
        if (lifecycleSet.has(name)) {
          continue;
        }
        const range = ranges.get(name);
        if (!range) {
          continue;
        }
        const lifecycleName = lifecycleSet.has(rule.target) ? rule.target : rule.owner;
        const idx = lifecycleNames.indexOf(lifecycleName);
        const isAfter = lifecycleSet.has(rule.target) ? rule.type === "after" : rule.type === "before";
        if (isAfter) {
          range.min = Math.max(range.min, idx + 1);
          range.minRules.push(rule);
        } else {
          range.max = Math.min(range.max, idx);
          range.maxRules.push(rule);
        }
        slots.set(name, range.min);
        continue;
      }
      const edge = rule.type === "before" ? {
        from: rule.owner,
        to: rule.target,
        rule
      } : {
        from: rule.target,
        to: rule.owner,
        rule
      };
      outgoing.get(edge.from)?.push(edge);
      incoming.get(edge.to)?.push(edge);
    }
    const stack = [];
    const onStack = /* @__PURE__ */ new Set();
    const index = /* @__PURE__ */ new Map();
    const lowLink = /* @__PURE__ */ new Map();
    const cycles = [];
    let cursor = 0;
    const visit = /* @__PURE__ */ __name((name) => {
      index.set(name, cursor);
      lowLink.set(name, cursor);
      cursor += 1;
      stack.push(name);
      onStack.add(name);
      for (const edge of outgoing.get(name) ?? []) {
        const next = edge.to;
        if (!index.has(next)) {
          visit(next);
          lowLink.set(
            name,
            Math.min(lowLink.get(name), lowLink.get(next))
          );
          continue;
        }
        if (onStack.has(next)) {
          lowLink.set(
            name,
            Math.min(lowLink.get(name), index.get(next))
          );
        }
      }
      if (lowLink.get(name) !== index.get(name)) {
        return;
      }
      const group = [];
      while (true) {
        const current = stack.pop();
        onStack.delete(current);
        group.push(current);
        if (current === name) {
          break;
        }
      }
      if (group.length > 1 || (outgoing.get(group[0]) ?? []).some(
        (edge) => edge.to === group[0]
      )) {
        cycles.push(group);
      }
    }, "visit");
    for (const node of normalNodes) {
      if (!index.has(node.name)) {
        visit(node.name);
      }
    }
    if (cycles.length > 0) {
      const cycle = cycles[0];
      const cycleSet = new Set(cycle);
      const blocked = /* @__PURE__ */ new Set();
      const queue = [...cycle];
      while (queue.length > 0) {
        const current = queue.shift();
        for (const edge of outgoing.get(current) ?? []) {
          if (cycleSet.has(edge.to) || blocked.has(edge.to)) {
            continue;
          }
          blocked.add(edge.to);
          queue.push(edge.to);
        }
      }
      const order = new Map(nodes.map((node) => [node.name, node.index]));
      const cycleEdges = cycle.flatMap((name) => outgoing.get(name) ?? []).filter((edge) => cycleSet.has(edge.to)).sort(
        (a, b) => (order.get(a.from) ?? 0) - (order.get(b.from) ?? 0) || (order.get(a.to) ?? 0) - (order.get(b.to) ?? 0)
      );
      const blockedList = [...blocked].sort(
        (a, b) => (order.get(a) ?? 0) - (order.get(b) ?? 0)
      );
      throw new Error(
        [
          "Circular dependency detected in middleware graph.",
          `Cycle nodes: ${cycle.sort(
            (a, b) => (order.get(a) ?? 0) - (order.get(b) ?? 0)
          ).join(" -> ")}`,
          "Constraints:",
          ...cycleEdges.map(
            (edge) => `- ${edge.from} -> ${edge.to} via ${this._formatRule(edge.rule)}`
          ),
          blockedList.length > 0 ? `Blocked nodes: ${blockedList.join(", ")}` : ""
        ].filter((line) => line.length > 0).join("\n")
      );
    }
    const indegree = /* @__PURE__ */ new Map();
    for (const node of normalNodes) {
      indegree.set(node.name, incoming.get(node.name)?.length ?? 0);
    }
    const ready = normalNodes.filter((node) => indegree.get(node.name) === 0).map((node) => node.name);
    const topo = [];
    while (ready.length > 0) {
      ready.sort(
        (a, b) => (this._tasks.get(a)?.index ?? 0) - (this._tasks.get(b)?.index ?? 0)
      );
      const current = ready.shift();
      const currentSlot = slots.get(current) ?? 0;
      const range = ranges.get(current);
      if (range && currentSlot > range.max) {
        const path = [];
        let name = current;
        const visited = /* @__PURE__ */ new Set();
        while (slotCause.has(name) && !visited.has(name)) {
          visited.add(name);
          const edge = slotCause.get(name);
          path.unshift(
            `- ${edge.from} -> ${edge.to} via ${this._formatRule(edge.rule)}`
          );
          name = edge.from;
        }
        throw new Error(
          [
            `Cannot place middleware "${current}" in lifecycle order.`,
            `Resolved position: ${this._formatSlot(currentSlot)}`,
            `Latest allowed position: ${this._formatSlot(range.max)}`,
            range.minRules.length > 0 ? "Required after:" : "",
            ...range.minRules.map(
              (rule) => `- ${this._formatRule(rule)}`
            ),
            range.maxRules.length > 0 ? "Required before:" : "",
            ...range.maxRules.map(
              (rule) => `- ${this._formatRule(rule)}`
            ),
            path.length > 0 ? "Dependency chain:" : "",
            ...path
          ].filter((line) => line.length > 0).join("\n")
        );
      }
      topo.push(current);
      for (const edge of outgoing.get(current) ?? []) {
        if ((slots.get(edge.to) ?? 0) < currentSlot) {
          slots.set(edge.to, currentSlot);
          slotCause.set(edge.to, edge);
        }
        indegree.set(edge.to, (indegree.get(edge.to) ?? 0) - 1);
        if (indegree.get(edge.to) === 0) {
          ready.push(edge.to);
        }
      }
    }
    const levels = [];
    const slotGroups = /* @__PURE__ */ new Map();
    for (const name of topo) {
      const slot = slots.get(name) ?? 0;
      const group = slotGroups.get(slot) ?? [];
      group.push(name);
      slotGroups.set(slot, group);
    }
    for (let slot = 0; slot <= lifecycleNames.length; slot++) {
      const group = slotGroups.get(slot) ?? [];
      if (group.length > 0) {
        const groupSet = new Set(group);
        const groupIndegree = /* @__PURE__ */ new Map();
        for (const name of group) {
          groupIndegree.set(
            name,
            (incoming.get(name) ?? []).filter(
              (edge) => groupSet.has(edge.from)
            ).length
          );
        }
        let currentLevel = group.filter((name) => groupIndegree.get(name) === 0).sort(
          (a, b) => (this._tasks.get(a)?.index ?? 0) - (this._tasks.get(b)?.index ?? 0)
        );
        while (currentLevel.length > 0) {
          levels.push(
            currentLevel.map((name) => {
              const task = this._tasks.get(name);
              if (task?.middleware == null) {
                throw new Error(
                  `Missing middleware for task ${name}`
                );
              }
              return task.middleware;
            })
          );
          const nextLevel = [];
          for (const name of currentLevel) {
            for (const edge of outgoing.get(name) ?? []) {
              if (!groupSet.has(edge.to)) {
                continue;
              }
              groupIndegree.set(
                edge.to,
                (groupIndegree.get(edge.to) ?? 0) - 1
              );
              if (groupIndegree.get(edge.to) === 0) {
                nextLevel.push(edge.to);
              }
            }
          }
          currentLevel = nextLevel.sort(
            (a, b) => (this._tasks.get(a)?.index ?? 0) - (this._tasks.get(b)?.index ?? 0)
          );
        }
      }
      const lifecycle = this._tasks.get(lifecycleNames[slot])?.middleware;
      if (lifecycle) {
        levels.push([lifecycle]);
      }
    }
    this._cachedOrder = levels;
    return levels;
  }
  _captureRuleSource() {
    const stack = new Error().stack?.split("\n") ?? [];
    return stack.map((line) => line.trim()).find(
      (line) => line.length > 0 && line !== "Error" && !line.includes("ChatChainDependencyGraph.") && !line.includes("ChainMiddleware.") && !line.includes("chains\\chain.") && !line.includes("chains/chain.")
    );
  }
  _formatRule(rule) {
    const source = rule.source ? ` at ${rule.source}` : "";
    return `.${rule.type}('${rule.target}') declared by ${rule.owner}${source}`;
  }
  _formatSlot(slot) {
    if (slot <= 0) {
      return `before ${lifecycleNames[0]}`;
    }
    if (slot >= lifecycleNames.length) {
      return `after ${lifecycleNames[lifecycleNames.length - 1]}`;
    }
    return `between ${lifecycleNames[slot - 1]} and ${lifecycleNames[slot]}`;
  }
};
var ChainMiddleware = class {
  constructor(name, execute, graph) {
    this.name = name;
    this.execute = execute;
    this.graph = graph;
  }
  static {
    __name(this, "ChainMiddleware");
  }
  before(name) {
    this.graph.before(this.name, name);
    return this;
  }
  after(name) {
    this.graph.after(this.name, name);
    return this;
  }
  run(session, options) {
    return this.execute(session, options);
  }
};
var DefaultChatChainSender = class {
  constructor(config) {
    this.config = config;
  }
  static {
    __name(this, "DefaultChatChainSender");
  }
  processElements(elements) {
    return elements.filter((element) => {
      if (!element) return false;
      if (element.type === "img") {
        const src = element.attrs?.["src"];
        return !(typeof src === "string" && src.startsWith("attachment"));
      }
      return true;
    }).map((element) => {
      if (element.children?.length) {
        element.children = this.processElements(element.children);
      }
      return element;
    });
  }
  async send(session, messages, context) {
    if (!messages?.length) return;
    if (isElementArray(messages?.[0]) && messages[0][1]?.type === "markdown-qq") {
      await this.sendAsQQMarkdown(session, messages[0][0]);
      return;
    }
    if (this.config.isForwardMsg && this.getMessageText(messages).length > this.config.forwardMsgMinLength) {
      await this.sendAsForward(session, messages);
      return;
    }
    await this.sendAsNormal(session, messages, context);
  }
  async sendAsQQMarkdown(session, message) {
    const { user } = session.event;
    await session.bot.internal.sendPrivateMessage(
      user.id,
      {
        msg_type: 2,
        msg_seq: 1,
        msg_id: session.messageId,
        markdown: {
          content: message.attrs["content"]
        }
      }
    );
  }
  async sendAsForward(session, messages) {
    const sendMessages = this.convertToForwardMessages(messages);
    if (sendMessages.length < 1 || sendMessages.length === 1 && sendMessages.join().length === 0) {
      return;
    }
    await session.sendQueued(
      (0, import_koishi.h)("message", { forward: true }, ...sendMessages)
    );
  }
  convertToForwardMessages(messages) {
    const firstMsg = messages[0];
    if (Array.isArray(firstMsg)) {
      return messages.map((msg) => (0, import_koishi.h)("message", ...msg));
    }
    if (typeof firstMsg === "object") {
      return [(0, import_koishi.h)("message", ...messages)];
    }
    if (typeof firstMsg === "string") {
      return [import_koishi.h.text(firstMsg)];
    }
    throw new Error(`Unsupported message type: ${typeof firstMsg}`);
  }
  async sendAsNormal(session, messages, context) {
    for (const message of messages) {
      const messageFragment = await this.buildMessageFragment(
        session,
        message,
        context
      );
      if (!messageFragment?.length) continue;
      const processedFragment = this.processElements(messageFragment);
      await session.sendQueued(processedFragment);
    }
  }
  async buildMessageFragment(session, message, context) {
    const start = context?.options?.startedAt;
    const elapsed = start ? Date.now() - start : 0;
    const threshold = (this.config.replyQuoteThreshold ?? 0) * 1e3;
    const shouldAddQuote = this.config.isReplyWithAt && session.isDirect === false && session.messageId && elapsed >= threshold;
    const messageContent = this.convertMessageToArray(message);
    if (messageContent == null || messageContent.length < 1 || messageContent.length === 1 && messageContent.join().length === 0) {
      return;
    }
    if (!shouldAddQuote) {
      return messageContent;
    }
    const quote = (0, import_koishi.h)("quote", { id: session.messageId });
    const hasIncompatibleType = messageContent.some(
      (element) => element.type === "audio" || element.type === "message"
    );
    return hasIncompatibleType ? messageContent : [quote, ...messageContent];
  }
  convertMessageToArray(message) {
    if (Array.isArray(message)) {
      return message;
    }
    if (typeof message === "string") {
      return [import_koishi.h.text(message)];
    }
    return [message];
  }
  getMessageText(message) {
    return message.map((element) => {
      if (typeof element === "string") {
        return element;
      }
      if (Array.isArray(element)) {
        return import_koishi.h.select(element, "text").toString();
      }
      return element.toString();
    }).join(" ");
  }
};
var ChainMiddlewareRunStatus = /* @__PURE__ */ ((ChainMiddlewareRunStatus2) => {
  ChainMiddlewareRunStatus2[ChainMiddlewareRunStatus2["SKIPPED"] = 0] = "SKIPPED";
  ChainMiddlewareRunStatus2[ChainMiddlewareRunStatus2["STOP"] = 1] = "STOP";
  ChainMiddlewareRunStatus2[ChainMiddlewareRunStatus2["CONTINUE"] = 2] = "CONTINUE";
  return ChainMiddlewareRunStatus2;
})(ChainMiddlewareRunStatus || {});
function isElementArray(value) {
  return Array.isArray(value) && value.every(
    (item) => typeof item === "object" && item.attrs && item.type
  );
}
__name(isElementArray, "isElementArray");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ChainMiddleware,
  ChainMiddlewareRunStatus,
  ChatChain
});
