// koishi-plugin-chatluna/lib/utils/chatluna.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/utils/chatluna.ts
import { SystemMessage } from "@langchain/core/messages";
import { computed } from "@vue/reactivity";
import { ChatLunaChatPrompt } from "koishi-plugin-chatluna/llm-core/chain/prompt";
function computePreset(ctx, info, rawText) {
  return computed(
    () => ({
      triggerKeyword: [info.name],
      rawText,
      messages: rawText ? [new SystemMessage(rawText)] : [],
      config: info.promptMode === "preset" && info.preset ? ctx.chatluna.preset.getPreset(info.preset).value?.config ?? {} : {}
    })
  );
}
__name(computePreset, "computePreset");
function createChatPrompt(ctx, llm, preset) {
  return new ChatLunaChatPrompt({
    preset,
    tokenCounter: /* @__PURE__ */ __name((text) => llm.getNumTokens(text), "tokenCounter"),
    sendTokenLimit: llm.invocationParams().maxTokenLimit ?? llm.getModelMaxContextSize(),
    contextManager: ctx.chatluna.contextManager,
    promptRenderService: ctx.chatluna.promptRenderer
  });
}
__name(createChatPrompt, "createChatPrompt");
export {
  computePreset,
  createChatPrompt
};
