// koishi-plugin-chatluna/lib/utils/chatluna.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/chatluna.ts
var chatluna_exports = {};
__export(chatluna_exports, {
  computePreset: () => computePreset,
  createChatPrompt: () => createChatPrompt
});
module.exports = __toCommonJS(chatluna_exports);
var import_messages = require("@langchain/core/messages");
var import_reactivity = require("@vue/reactivity");
var import_prompt = require("koishi-plugin-chatluna/llm-core/chain/prompt");
function computePreset(ctx, info, rawText) {
  return (0, import_reactivity.computed)(
    () => ({
      triggerKeyword: [info.name],
      rawText,
      messages: rawText ? [new import_messages.SystemMessage(rawText)] : [],
      config: info.promptMode === "preset" && info.preset ? ctx.chatluna.preset.getPreset(info.preset).value?.config ?? {} : {}
    })
  );
}
__name(computePreset, "computePreset");
function createChatPrompt(ctx, llm, preset) {
  return new import_prompt.ChatLunaChatPrompt({
    preset,
    tokenCounter: /* @__PURE__ */ __name((text) => llm.getNumTokens(text), "tokenCounter"),
    sendTokenLimit: llm.invocationParams().maxTokenLimit ?? llm.getModelMaxContextSize(),
    contextManager: ctx.chatluna.contextManager,
    promptRenderService: ctx.chatluna.promptRenderer
  });
}
__name(createChatPrompt, "createChatPrompt");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  computePreset,
  createChatPrompt
});
