// koishi-plugin-chatluna/lib/utils/langchain.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/langchain.ts
var langchain_exports = {};
__export(langchain_exports, {
  isChatLunaUserMessage: () => isChatLunaUserMessage,
  isMessageContentAudio: () => isMessageContentAudio,
  isMessageContentComplex: () => isMessageContentComplex,
  isMessageContentFileUrl: () => isMessageContentFileUrl,
  isMessageContentImageUrl: () => isMessageContentImageUrl,
  isMessageContentText: () => isMessageContentText,
  isMessageContentVideo: () => isMessageContentVideo,
  markChatLunaUserMessage: () => markChatLunaUserMessage,
  truncateMessageContentUrls: () => truncateMessageContentUrls
});
module.exports = __toCommonJS(langchain_exports);
function markChatLunaUserMessage(msg) {
  msg.response_metadata = {
    ...msg.response_metadata ?? {},
    chatluna: {
      ...msg.response_metadata?.chatluna ?? {},
      source: "user"
    }
  };
}
__name(markChatLunaUserMessage, "markChatLunaUserMessage");
function isChatLunaUserMessage(msg) {
  const meta = msg.response_metadata?.chatluna;
  return meta?.source === "user";
}
__name(isChatLunaUserMessage, "isChatLunaUserMessage");
function isMessageContentImageUrl(message) {
  return message.type === "image_url" && message["image_url"] != null;
}
__name(isMessageContentImageUrl, "isMessageContentImageUrl");
function isMessageContentText(message) {
  return message.type === "text" && message.text != null;
}
__name(isMessageContentText, "isMessageContentText");
function isMessageContentFileUrl(message) {
  return message.type === "file_url" && message["file_url"] != null;
}
__name(isMessageContentFileUrl, "isMessageContentFileUrl");
function isMessageContentAudio(message) {
  return message.type === "audio_url" && message["audio_url"] != null;
}
__name(isMessageContentAudio, "isMessageContentAudio");
function isMessageContentVideo(message) {
  return message.type === "video_url" && message["video_url"] != null;
}
__name(isMessageContentVideo, "isMessageContentVideo");
function isMessageContentComplex(value) {
  if (value == null || typeof value !== "object") {
    return false;
  }
  const item = value;
  return isMessageContentText(item) || isMessageContentImageUrl(item) || isMessageContentFileUrl(item) || isMessageContentAudio(item) || isMessageContentVideo(item);
}
__name(isMessageContentComplex, "isMessageContentComplex");
function isMessageContentUrlPart(message) {
  return isMessageContentImageUrl(message) || isMessageContentFileUrl(message) || isMessageContentAudio(message) || isMessageContentVideo(message);
}
__name(isMessageContentUrlPart, "isMessageContentUrlPart");
function getMessageContentPartUrl(message) {
  if (isMessageContentImageUrl(message)) {
    return typeof message.image_url === "string" ? message.image_url : message.image_url.url;
  }
  if (isMessageContentFileUrl(message)) {
    return typeof message.file_url === "string" ? message.file_url : message.file_url.url;
  }
  if (isMessageContentAudio(message)) {
    return typeof message.audio_url === "string" ? message.audio_url : message.audio_url.url;
  }
  return typeof message.video_url === "string" ? message.video_url : message.video_url.url;
}
__name(getMessageContentPartUrl, "getMessageContentPartUrl");
function setMessageContentPartUrl(message, url) {
  if (isMessageContentImageUrl(message)) {
    if (typeof message.image_url === "string") {
      message.image_url = url;
      return;
    }
    message.image_url.url = url;
    return;
  }
  if (isMessageContentFileUrl(message)) {
    if (typeof message.file_url === "string") {
      message.file_url = url;
      return;
    }
    message.file_url.url = url;
    return;
  }
  if (isMessageContentAudio(message)) {
    if (typeof message.audio_url === "string") {
      message.audio_url = url;
      return;
    }
    message.audio_url.url = url;
    return;
  }
  if (typeof message.video_url === "string") {
    message.video_url = url;
    return;
  }
  message.video_url.url = url;
}
__name(setMessageContentPartUrl, "setMessageContentPartUrl");
function truncateMessageContentUrls(content, maxLength = 100) {
  if (!Array.isArray(content)) {
    return content;
  }
  return content.map((part) => {
    if (!isMessageContentUrlPart(part)) {
      return part;
    }
    const url = getMessageContentPartUrl(part);
    if (url.startsWith("http") || url.length <= maxLength) {
      return part;
    }
    setMessageContentPartUrl(
      part,
      url.substring(0, maxLength) + " ..." + url.length + " chars"
    );
    return part;
  });
}
__name(truncateMessageContentUrls, "truncateMessageContentUrls");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  isChatLunaUserMessage,
  isMessageContentAudio,
  isMessageContentComplex,
  isMessageContentFileUrl,
  isMessageContentImageUrl,
  isMessageContentText,
  isMessageContentVideo,
  markChatLunaUserMessage,
  truncateMessageContentUrls
});
