// koishi-plugin-chatluna/lib/utils/queue.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/utils/queue.ts
import {
  ChatLunaError,
  ChatLunaErrorCode
} from "koishi-plugin-chatluna/utils/error";
import { ObjectLock } from "koishi-plugin-chatluna/utils/lock";
import { withResolver } from "koishi-plugin-chatluna/utils/promise";
var TIME_MINUTE = 60 * 1e3;
var RequestIdQueue = class {
  static {
    __name(this, "RequestIdQueue");
  }
  _queue = {};
  _queueLocks = {};
  // Per-key runnable window size used when waking the next batch.
  _limits = {};
  _maxQueueSize = 50;
  _queueTimeout;
  constructor(queueTimeout = TIME_MINUTE * 3) {
    this._queueTimeout = queueTimeout;
    const timer = setInterval(() => this.cleanup(), queueTimeout);
    timer.unref?.();
  }
  async add(key, requestId) {
    const currentLength = this._queue[key]?.length ?? 0;
    if (currentLength >= this._maxQueueSize) {
      throw new ChatLunaError(ChatLunaErrorCode.QUEUE_OVERFLOW);
    }
    if (!this._queueLocks[key]) {
      this._queueLocks[key] = new ObjectLock(this._queueTimeout);
    }
    const { promise, resolve, reject } = withResolver();
    const queueItem = {
      requestId,
      timestamp: Date.now(),
      active: false,
      timeout: this._queueTimeout,
      notifyPromise: { promise, resolve, reject }
    };
    let isFirst = false;
    try {
      await this._queueLocks[key].runLocked(async () => {
        if (!this._queue[key]) {
          this._queue[key] = [];
        }
        const existingIndex = this._queue[key].findIndex(
          (item) => item.requestId === requestId
        );
        if (existingIndex !== -1) {
          return;
        }
        if (this._queue[key].length >= this._maxQueueSize) {
          throw new ChatLunaError(ChatLunaErrorCode.QUEUE_OVERFLOW);
        }
        this._queue[key].push(queueItem);
        isFirst = this._queue[key].length === 1;
        if (isFirst) {
          queueItem.active = true;
        }
      });
      if (isFirst) {
        resolve();
      }
    } catch (error) {
      reject(error);
      throw error;
    }
  }
  async remove(key, requestId) {
    if (!this._queue[key]) return;
    if (!this._queueLocks[key]) {
      this._queueLocks[key] = new ObjectLock(this._queueTimeout);
    }
    const lock = this._queueLocks[key];
    const items = [];
    try {
      await lock.runLocked(async () => {
        if (!this._queue[key]) return;
        const index = this._queue[key].findIndex(
          (item) => item.requestId === requestId
        );
        if (index === -1) return;
        this._queue[key].splice(index, 1);
        if (this._queue[key].length === 0) {
          delete this._queue[key];
          delete this._limits[key];
          return;
        }
        const limit = this._limits[key] ?? 1;
        for (let idx = 0; idx < this._queue[key].length && idx < limit; idx++) {
          const item = this._queue[key][idx];
          if (item.active) {
            continue;
          }
          item.active = true;
          items.push(item);
        }
      });
      items.forEach((item) => item.notifyPromise.resolve());
    } catch (error) {
      console.error("Error in remove operation:", error);
    }
  }
  async wait(key, requestId, maxConcurrent, timeout = this._queueTimeout) {
    if (!this._queue[key]) {
      await this.add(key, requestId);
    }
    if (!this._queueLocks[key]) {
      this._queueLocks[key] = new ObjectLock(this._queueTimeout);
    }
    const lock = this._queueLocks[key];
    const limit = maxConcurrent > 0 ? maxConcurrent : 1;
    let item;
    let shouldExecute = false;
    await lock.runLocked(async () => {
      if (!this._queue[key]) return;
      this._limits[key] = limit;
      const index = this._queue[key].findIndex(
        (item2) => item2.requestId === requestId
      );
      if (index === -1) return;
      item = this._queue[key][index];
      item.timeout = index < limit ? timeout : Math.max(
        this._queueTimeout,
        Math.ceil(index / limit) * timeout
      );
      if (index < limit) {
        item.active = true;
        shouldExecute = true;
      }
    });
    if (shouldExecute || item == null) {
      return;
    }
    let timeoutId;
    const timeoutError = new Error(
      `Queue wait timeout after ${item.timeout}ms`
    );
    try {
      const timeoutPromise = new Promise((_, reject) => {
        timeoutId = setTimeout(() => {
          reject(timeoutError);
        }, item.timeout);
      });
      await Promise.race([item.notifyPromise.promise, timeoutPromise]);
    } catch (error) {
      await this.remove(key, requestId).catch(() => {
      });
      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }
  async cleanup() {
    const now = Date.now();
    const keys = Object.keys(this._queue);
    for (const key of keys) {
      if (!this._queueLocks[key]) {
        this._queueLocks[key] = new ObjectLock(this._queueTimeout);
      }
      const lock = this._queueLocks[key];
      const expired = [];
      const items = [];
      await lock.runLocked(async () => {
        if (!this._queue[key]) return;
        for (const item of this._queue[key]) {
          if (!item.active && now - item.timestamp >= item.timeout) {
            expired.push(item);
          }
        }
        if (expired.length === 0) {
          return;
        }
        this._queue[key] = this._queue[key].filter(
          (item) => item.active || now - item.timestamp < item.timeout
        );
        if (this._queue[key].length === 0) {
          delete this._queue[key];
          delete this._limits[key];
          return;
        }
        const limit = this._limits[key] ?? 1;
        for (let idx = 0; idx < this._queue[key].length && idx < limit; idx++) {
          const item = this._queue[key][idx];
          if (item.active) {
            continue;
          }
          item.active = true;
          items.push(item);
        }
      });
      expired.forEach((item) => {
        item.notifyPromise.reject(
          new Error(`Queue wait timeout after ${item.timeout}ms`)
        );
      });
      items.forEach((item) => item.notifyPromise.resolve());
    }
  }
  async getQueueLength(key) {
    if (!this._queueLocks[key]) {
      this._queueLocks[key] = new ObjectLock(this._queueTimeout);
    }
    return await this._queueLocks[key].runLocked(
      async () => this._queue[key]?.length ?? 0
    );
  }
  async getQueueStatus(key) {
    if (!this._queueLocks[key]) {
      this._queueLocks[key] = new ObjectLock(this._queueTimeout);
    }
    return await this._queueLocks[key].runLocked(async () => ({
      length: this._queue[key]?.length ?? 0,
      items: this._queue[key]?.map((item) => ({
        requestId: item.requestId,
        age: Date.now() - item.timestamp
      })) ?? []
    }));
  }
};
export {
  RequestIdQueue
};
