// koishi-plugin-chatluna/lib/utils/schema.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/utils/schema.ts
import { h, Schema } from "koishi";
import { computed, watch } from "@vue/reactivity";
import { ModelType } from "koishi-plugin-chatluna/llm-core/platform/types";

// src/utils/model.ts
function parseRawModelName(modelName) {
  if (modelName == null || modelName.trim().length < 1) {
    return [void 0, void 0];
  }
  const value = modelName.trim();
  const index = value.indexOf("/");
  if (index === -1) {
    return [void 0, value];
  }
  if (index === 0 || index === value.length - 1) {
    return [void 0, void 0];
  }
  return [value.slice(0, index), value.slice(index + 1)];
}
__name(parseRawModelName, "parseRawModelName");

// src/utils/schema.ts
function compareSchemaName(a, b) {
  const [aPlatform, aName] = parseRawModelName(a);
  const [bPlatform, bName] = parseRawModelName(b);
  return (aPlatform ?? "").localeCompare(bPlatform ?? "") || (aName ?? "").localeCompare(bName ?? "");
}
__name(compareSchemaName, "compareSchemaName");
function modelSchema(ctx, createNotification = false) {
  const modelNames = getModelSchemas(ctx.chatluna.platform);
  const notification = createNotification ? ctx.notifier?.create({
    content: h.parse(
      "您当前没有配置模型，请前往 https://chatluna.chat/guide/configure-model-platform/introduction.html 了解如何安装模型适配器。"
    ),
    type: "warning"
  }) : void 0;
  const watcher = watch(
    modelNames,
    (modelNames2) => {
      ctx.schema.set("model", Schema.union(modelNames2));
      if (modelNames2.length > 1 && notification) {
        notification?.update({
          content: `当前共加载了 ${modelNames2.length} 个模型。`,
          type: "success"
        });
      } else {
        notification?.update({
          content: h.parse(
            "您当前没有配置模型，请前往 https://chatluna.chat/guide/configure-model-platform/introduction.html 了解如何安装模型适配器。"
          ),
          type: "warning"
        });
      }
    },
    {
      immediate: true
    }
  );
  const stop = /* @__PURE__ */ __name(() => watcher.stop(), "stop");
  ctx.effect(() => stop);
}
__name(modelSchema, "modelSchema");
function embeddingsSchema(ctx) {
  const modelNames = getModelSchemas(
    ctx.chatluna.platform,
    ModelType.embeddings
  );
  const watcher = watch(
    modelNames,
    (modelNames2) => {
      ctx.schema.set("embeddings", Schema.union(modelNames2));
    },
    {
      immediate: true
    }
  );
  const stop = /* @__PURE__ */ __name(() => watcher.stop(), "stop");
  ctx.effect(() => stop);
}
__name(embeddingsSchema, "embeddingsSchema");
function rerankerSchema(ctx) {
  const modelNames = getModelSchemas(
    ctx.chatluna.platform,
    ModelType.reranker
  );
  const watcher = watch(
    modelNames,
    (modelNames2) => {
      ctx.schema.set("reranker", Schema.union(modelNames2));
    },
    {
      immediate: true
    }
  );
  const stop = /* @__PURE__ */ __name(() => watcher.stop(), "stop");
  ctx.effect(() => stop);
}
__name(rerankerSchema, "rerankerSchema");
function chatChainSchema(ctx) {
  const modelNames = getChatChainNames(ctx.chatluna.platform);
  const watcher = watch(
    modelNames,
    (modelNames2) => {
      ctx.schema.set("chat-mode", Schema.union(modelNames2));
    },
    {
      immediate: true
    }
  );
  const stop = /* @__PURE__ */ __name(() => watcher.stop(), "stop");
  ctx.effect(() => stop);
}
__name(chatChainSchema, "chatChainSchema");
function vectorStoreSchema(ctx) {
  const vectorStoreNames = getVectorStores(ctx, ctx.chatluna.platform);
  const watcher = watch(
    vectorStoreNames,
    (vectorStoreNames2) => {
      ctx.schema.set("vector-store", Schema.union(vectorStoreNames2));
    },
    {
      immediate: true
    }
  );
  const stop = /* @__PURE__ */ __name(() => watcher.stop(), "stop");
  ctx.effect(() => stop);
}
__name(vectorStoreSchema, "vectorStoreSchema");
function listModelNames(service, type = ModelType.llm, includeNone = false) {
  const models = service.listAllModels(type);
  return computed(() => {
    const names = models.value.map(
      (model) => model.platform + "/" + model.name
    );
    return includeNone ? ["无"].concat(names).sort(compareSchemaName) : names.sort(compareSchemaName);
  });
}
__name(listModelNames, "listModelNames");
function getModelSchemas(service, type = ModelType.llm) {
  const models = listModelNames(service, type, true);
  return computed(
    () => models.value.map((model) => Schema.const(model).description(model))
  );
}
__name(getModelSchemas, "getModelSchemas");
function getVectorStores(ctx, service) {
  const vectorStoreNamesRef = service.vectorStores;
  return computed(
    () => ["无"].concat(vectorStoreNamesRef.value).sort(compareSchemaName).map((name) => Schema.const(name).description(name))
  );
}
__name(getVectorStores, "getVectorStores");
function getChatChainNames(service) {
  const chains = service.chatChains;
  return computed(
    () => chains.value.slice().sort((a, b) => compareSchemaName(a.name, b.name)).map((info) => Schema.const(info.name).i18n(info.description))
  );
}
__name(getChatChainNames, "getChatChainNames");
export {
  chatChainSchema,
  embeddingsSchema,
  listModelNames,
  modelSchema,
  rerankerSchema,
  vectorStoreSchema
};
