// koishi-plugin-chatluna/lib/utils/string.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/utils/string.ts
import { h } from "koishi";
import { gunzip, gzip } from "zlib";
import { promisify } from "util";
import { chatLunaFetch } from "koishi-plugin-chatluna/utils/request";
import crypto from "node:crypto";
import { transformMessageContentToElements } from "koishi-plugin-chatluna/utils/koishi";
import {
  isMessageContentImageUrl,
  isMessageContentText
} from "koishi-plugin-chatluna/utils/langchain";
var gzipAsync = promisify(gzip);
var gunzipAsync = promisify(gunzip);
function fuzzyQuery(source, keywords) {
  for (const keyword of keywords) {
    const match = source.includes(keyword);
    if (match) {
      return true;
    }
  }
  return false;
}
__name(fuzzyQuery, "fuzzyQuery");
var MIME_TYPE_BY_EXTENSION = {
  png: "image/png",
  jpg: "image/jpeg",
  jpeg: "image/jpeg",
  gif: "image/gif",
  webp: "image/webp",
  svg: "image/svg+xml",
  bmp: "image/bmp",
  ico: "image/x-icon",
  html: "text/html",
  htm: "text/html",
  css: "text/css",
  txt: "text/plain",
  md: "text/markdown",
  xml: "text/xml",
  csv: "text/csv",
  rtf: "text/rtf",
  js: "text/javascript",
  mjs: "text/javascript",
  json: "application/json",
  pdf: "application/pdf",
  mp4: "video/mp4",
  mpeg: "video/mpeg",
  mov: "video/mov",
  avi: "video/avi",
  flv: "video/x-flv",
  mpg: "video/mpg",
  webm: "video/webm",
  wmv: "video/wmv",
  "3gp": "video/3gpp",
  "3gpp": "video/3gpp",
  mp3: "audio/mpeg",
  amr: "audio/amr",
  aiff: "audio/aiff",
  aac: "audio/aac",
  flac: "audio/flac",
  wav: "audio/wav",
  ogg: "audio/ogg",
  m4a: "audio/mp4"
};
function getMimeTypeFromExtension(ext) {
  const normalized = ext?.trim().toLowerCase().replace(/^\./, "");
  if (!normalized) {
    return null;
  }
  return MIME_TYPE_BY_EXTENSION[normalized] ?? null;
}
__name(getMimeTypeFromExtension, "getMimeTypeFromExtension");
function getMimeTypeFromSource(sourceUrl, fileName) {
  const candidates = [fileName, sourceUrl];
  for (const candidate of candidates) {
    if (!candidate) {
      continue;
    }
    const source = candidate.split("?")[0].split("#")[0].toLowerCase();
    const filePart = source.split(/[\\/]/).pop();
    if (!filePart) {
      continue;
    }
    const dotIndex = filePart.lastIndexOf(".");
    const extension = dotIndex >= 0 ? filePart.slice(dotIndex + 1) : filePart;
    const mimeType = getMimeTypeFromExtension(extension);
    if (mimeType) {
      return mimeType;
    }
  }
  return null;
}
__name(getMimeTypeFromSource, "getMimeTypeFromSource");
function getImageMimeType(ext) {
  const mimeType = getMimeTypeFromExtension(ext);
  return mimeType?.startsWith("image/") ? mimeType : "image/jpeg";
}
__name(getImageMimeType, "getImageMimeType");
function getImageType(buffer, pure = false, checkIsImage = true) {
  if (buffer.length < 12) {
    return checkIsImage ? void 0 : pure ? "jpg" : "image/jpeg";
  }
  if (buffer[0] === 137 && buffer[1] === 80 && buffer[2] === 78 && buffer[3] === 71 && buffer[4] === 13 && buffer[5] === 10 && buffer[6] === 26 && buffer[7] === 10) {
    return pure ? "png" : "image/png";
  }
  if (buffer[0] === 255 && buffer[1] === 216 && buffer[2] === 255) {
    return pure ? "jpg" : "image/jpeg";
  }
  if (buffer[0] === 71 && buffer[1] === 73 && buffer[2] === 70 && buffer[3] === 56) {
    return pure ? "gif" : "image/gif";
  }
  if (buffer[0] === 82 && buffer[1] === 73 && buffer[2] === 70 && buffer[3] === 70 && buffer[8] === 87 && buffer[9] === 69 && buffer[10] === 66 && buffer[11] === 80) {
    return pure ? "webp" : "image/webp";
  }
  if (checkIsImage) {
    return void 0;
  }
  return pure ? "jpg" : "image/jpeg";
}
__name(getImageType, "getImageType");
function getMessageContent(message) {
  if (typeof message === "string") {
    return message;
  }
  if (message == null) {
    return "";
  }
  const buffer = [];
  for (const part of message) {
    if (part.type === "text") {
      buffer.push(part.text);
    }
  }
  return buffer.join("");
}
__name(getMessageContent, "getMessageContent");
function sanitizeToolLogString(input) {
  return input.replace(/data:[^;]+;base64,[A-Za-z0-9+/=]+/g, "[BASE64_DATA_URL]").replace(
    /("thoughtSignature"\s*:\s*")[^"]+(")/g,
    "$1[THOUGHT_SIGNATURE]$2"
  ).replace(
    /("data"\s*:\s*")[A-Za-z0-9+/=]{128,}(")/g,
    "$1[BASE64_DATA]$2"
  ).substring(0, 2e3) + "...";
}
__name(sanitizeToolLogString, "sanitizeToolLogString");
function sanitizeToolLogValue(value, visited = /* @__PURE__ */ new Set()) {
  if (typeof value === "string") {
    return sanitizeToolLogString(value);
  }
  if (value != null && typeof value === "object") {
    if (visited.has(value)) {
      return "[CIRCULAR]";
    }
    visited.add(value);
    try {
      if (Array.isArray(value)) {
        return value.map((item) => sanitizeToolLogValue(item, visited));
      }
      const result = {};
      for (const [key, inner] of Object.entries(
        value
      )) {
        if (key === "thoughtSignature" && typeof inner === "string") {
          result[key] = "[THOUGHT_SIGNATURE]";
          continue;
        }
        if (key === "data" && typeof inner === "string" && /^[A-Za-z0-9+/=]{128,}$/.test(inner)) {
          result[key] = "[BASE64_DATA]";
          continue;
        }
        result[key] = sanitizeToolLogValue(inner, visited);
      }
      return result;
    } finally {
      visited.delete(value);
    }
  }
  return value;
}
__name(sanitizeToolLogValue, "sanitizeToolLogValue");
function getNotEmptyString(...texts) {
  for (const text of texts) {
    if (text && text?.length > 0) {
      return text;
    }
  }
}
__name(getNotEmptyString, "getNotEmptyString");
function getCurrentWeekday() {
  const daysOfWeek = [
    "Sunday",
    "Monday",
    "Tuesday",
    "Wednesday",
    "Thursday",
    "Friday",
    "Saturday"
  ];
  const currentDate = /* @__PURE__ */ new Date();
  return daysOfWeek[currentDate.getDay()];
}
__name(getCurrentWeekday, "getCurrentWeekday");
var getTimeInUTC = /* @__PURE__ */ __name((offset) => {
  const date = /* @__PURE__ */ new Date();
  date.setMinutes(date.getMinutes() + offset * 60);
  return date.toISOString().substring(11, 8);
}, "getTimeInUTC");
var getTimeDiffFormat = /* @__PURE__ */ __name((time1, time2) => {
  const diff = Math.abs(time1 - time2);
  const days = Math.floor(diff / (1e3 * 60 * 60 * 24));
  const hours = Math.floor(diff % (1e3 * 60 * 60 * 24) / (1e3 * 60 * 60));
  const minutes = Math.floor(diff % (1e3 * 60 * 60) / (1e3 * 60));
  const parts = [];
  if (days > 0) parts.push(`${days} day${days > 1 ? "s" : ""}`);
  if (hours > 0) parts.push(`${hours} hour${hours > 1 ? "s" : ""}`);
  if (minutes > 0) parts.push(`${minutes} minute${minutes > 1 ? "s" : ""}`);
  return parts.join(", ") || "now";
}, "getTimeDiffFormat");
var getTimeDiff = /* @__PURE__ */ __name((time1, time2) => {
  return getTimeDiffFormat(
    new Date(time1).getTime(),
    new Date(time2).getTime()
  );
}, "getTimeDiff");
var selectFromList = /* @__PURE__ */ __name((args, isPick, seed = "") => {
  const items = args.split(",").map((item) => item.trim());
  if (isPick) {
    const hash = crypto.createHash("sha1").update(`${seed}:${args}`).digest();
    return items[hash.readUInt32BE(0) % items.length];
  }
  return items[Math.floor(Math.random() * items.length)];
}, "selectFromList");
var rollDice = /* @__PURE__ */ __name((formula) => {
  const parts = formula.split("d");
  let count = 1;
  if (parts.length > 1 && !isNaN(Number(parts[0]))) {
    count = parseInt(parts[0], 10);
  }
  const lastPart = parts[parts.length - 1].split("+");
  let add = 0;
  if (lastPart.length > 1 && !isNaN(Number(lastPart[1]))) {
    add = parseInt(lastPart[1], 10);
  }
  const range = !isNaN(Number(lastPart[0])) ? parseInt(lastPart[0], 10) : 1;
  return Math.floor(Math.random() * (count * range - count + 1)) + count + add;
}, "rollDice");
var fetchUrl = /* @__PURE__ */ __name(async (url, method = "GET", body = null, textLength = 1e3) => {
  const response = await chatLunaFetch(url, {
    method,
    body
  });
  const text = await response.text();
  if (text.length > textLength) {
    return text.substring(0, textLength);
  }
  return text;
}, "fetchUrl");
var PresetPostHandler = class {
  constructor(ctx, config, object) {
    this.ctx = ctx;
    this.config = config;
    this.prefix = object.prefix;
    this.postfix = object.postfix;
    this.variables = object.variables ?? {};
    this.censor = object.censor;
    this._compileVariables();
  }
  static {
    __name(this, "PresetPostHandler");
  }
  prefix;
  postfix;
  variables;
  censor;
  compiledVariables;
  async handler(session, data) {
    let content = data;
    const variables = {};
    if (this.compiledVariables) {
      for (const [key, value] of Object.entries(this.compiledVariables)) {
        const match = content.match(value);
        if (!match) {
          continue;
        }
        variables[key] = match[1];
      }
    }
    const censor = this.ctx.censor;
    if (censor && (this.config.censor || this.censor)) {
      content = await censor.transform([h.text(content)], session).then((element) => element.join(""));
    }
    let displayContent = content;
    if (this.prefix) {
      const startIndex = content.indexOf(this.prefix);
      if (startIndex !== -1) {
        displayContent = content.substring(
          startIndex + this.prefix.length
        );
      }
    }
    if (this.postfix) {
      const endIndex = displayContent.lastIndexOf(this.postfix);
      if (endIndex !== -1) {
        displayContent = displayContent.substring(0, endIndex);
      }
    }
    return { content, variables, displayContent };
  }
  _compileVariables() {
    if (!this.variables) {
      return;
    }
    this.compiledVariables = {};
    for (const [key, value] of Object.entries(this.variables)) {
      this.compiledVariables[key] = new RegExp(value);
    }
  }
};
async function gzipEncode(text, encoding = "buffer") {
  const buffer = await gzipAsync(text);
  return encoding === "buffer" ? buffer : buffer.toString(encoding);
}
__name(gzipEncode, "gzipEncode");
async function gzipDecode(data, inputEncoding = "base64") {
  const buffer = typeof data === "string" ? Buffer.from(data, inputEncoding) : data;
  return (await gunzipAsync(buffer)).toString("utf8");
}
__name(gzipDecode, "gzipDecode");
function bufferToArrayBuffer(buffer) {
  const arrayBuffer = new ArrayBuffer(buffer.length);
  const view = new Uint8Array(arrayBuffer);
  for (let i = 0; i < buffer.length; i++) {
    view[i] = buffer[i];
  }
  return arrayBuffer;
}
__name(bufferToArrayBuffer, "bufferToArrayBuffer");
async function hashString(text, length = 8) {
  const hash = await crypto.webcrypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(text)
  );
  const hashArray = Array.from(new Uint8Array(hash));
  const hashString2 = hashArray.map((b) => b.toString(16).padStart(2, "0")).join("");
  return hashString2.substring(0, length);
}
__name(hashString, "hashString");
function getSystemPromptVariables(session, config, conversation) {
  const lastActiveAt = conversation.lastChatAt ?? conversation.updatedAt;
  return {
    name: config.botNames[0],
    date: (/* @__PURE__ */ new Date()).toLocaleString(),
    bot_id: session.bot.selfId,
    is_group: !session.isDirect,
    is_private: session.isDirect,
    group_id: session.guildId ?? session.event?.guild?.id,
    group_name: session.event?.guild?.name || session.guildId,
    user_id: session.author?.user?.id ?? session.event?.user?.id ?? session.userId ?? "0",
    platform: session.platform,
    user: getNotEmptyString(
      session.author?.nick,
      session.author?.name,
      session.event.user?.name,
      session.username
    ),
    built: {
      preset: conversation.preset,
      platform: session.platform,
      conversationId: conversation.id
    },
    noop: "",
    time: (/* @__PURE__ */ new Date()).toLocaleTimeString(),
    weekday: getCurrentWeekday(),
    idle_duration: getTimeDiffFormat(
      (/* @__PURE__ */ new Date()).getTime(),
      lastActiveAt.getTime()
    )
  };
}
__name(getSystemPromptVariables, "getSystemPromptVariables");
function formatToolCall(tool, arg, log) {
  let rawArg = arg;
  if (Object.keys(rawArg).length === 1) {
    rawArg = rawArg?.input ?? rawArg?.arguments ?? rawArg;
  }
  if (typeof rawArg !== "string") {
    rawArg = JSON.stringify(rawArg, null, 2) || "";
  }
  return `{
  tool: '${tool}',
  arg: '${rawArg}',
  log: '${log}'
}`;
}
__name(formatToolCall, "formatToolCall");
async function formatUserPromptString(config, presetTemplate, session, prompt, conversation) {
  return await session.app.chatluna.promptRenderer.renderTemplate(
    presetTemplate.formatUserPromptString,
    {
      sender_id: session.author?.user?.id ?? session.event?.user?.id ?? "0",
      sender: getNotEmptyString(
        session.author?.nick,
        session.author?.name,
        session.event.user?.name,
        session.username
      ),
      prompt,
      ...getSystemPromptVariables(session, config, conversation)
    },
    {
      configurable: {
        session
      }
    }
  );
}
__name(formatUserPromptString, "formatUserPromptString");
export {
  PresetPostHandler,
  bufferToArrayBuffer,
  fetchUrl,
  formatToolCall,
  formatUserPromptString,
  fuzzyQuery,
  getCurrentWeekday,
  getImageMimeType,
  getImageType,
  getMessageContent,
  getMimeTypeFromSource,
  getNotEmptyString,
  getSystemPromptVariables,
  getTimeDiff,
  getTimeDiffFormat,
  getTimeInUTC,
  gzipDecode,
  gzipEncode,
  hashString,
  isMessageContentImageUrl,
  isMessageContentText,
  rollDice,
  sanitizeToolLogString,
  sanitizeToolLogValue,
  selectFromList,
  transformMessageContentToElements
};
