// koishi-plugin-chatluna/lib/utils/virtual_session.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/utils/virtual_session.ts
import { h } from "koishi";
import { transformMessageContentToElements } from "koishi-plugin-chatluna/utils/koishi";
import { getMessageContent } from "koishi-plugin-chatluna/utils/string";
function buildVirtualSession(bot, routing, action) {
  const event = {
    type: "message",
    platform: routing.platform,
    selfId: routing.selfId,
    timestamp: Date.now(),
    channel: {
      id: routing.channelId ?? routing.guildId ?? routing.userId,
      type: routing.isDirect ? 1 : 0
    },
    guild: routing.guildId == null ? void 0 : { id: routing.guildId },
    user: {
      id: routing.userId,
      name: routing.username ?? action.messageName ?? "trigger"
    },
    message: {
      content: getMessageContent(action.message),
      elements: typeof action.message === "string" ? [h.text(action.message)] : transformMessageContentToElements(action.message)
    }
  };
  return bot.session(event);
}
__name(buildVirtualSession, "buildVirtualSession");
export {
  buildVirtualSession
};
