// koishi-plugin-chatluna/lib/utils/virtual_session.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/utils/virtual_session.ts
var virtual_session_exports = {};
__export(virtual_session_exports, {
  buildVirtualSession: () => buildVirtualSession
});
module.exports = __toCommonJS(virtual_session_exports);
var import_koishi = require("koishi");
var import_koishi2 = require("koishi-plugin-chatluna/utils/koishi");
var import_string = require("koishi-plugin-chatluna/utils/string");
function buildVirtualSession(bot, routing, action) {
  const event = {
    type: "message",
    platform: routing.platform,
    selfId: routing.selfId,
    timestamp: Date.now(),
    channel: {
      id: routing.channelId ?? routing.guildId ?? routing.userId,
      type: routing.isDirect ? 1 : 0
    },
    guild: routing.guildId == null ? void 0 : { id: routing.guildId },
    user: {
      id: routing.userId,
      name: routing.username ?? action.messageName ?? "trigger"
    },
    message: {
      content: (0, import_string.getMessageContent)(action.message),
      elements: typeof action.message === "string" ? [import_koishi.h.text(action.message)] : (0, import_koishi2.transformMessageContentToElements)(action.message)
    }
  };
  return bot.session(event);
}
__name(buildVirtualSession, "buildVirtualSession");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  buildVirtualSession
});
