// @chatluna/shared-prompt-renderer/lib/index.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ChatLunaPromptRenderer: () => ChatLunaPromptRenderer
});
module.exports = __toCommonJS(index_exports);

// src/parser.ts
function parseExpression(input) {
  const lexer = new ExpressionLexer(input);
  const tokens = lexer.tokenize();
  const parser = new ExpressionParser(tokens);
  return parser.parse();
}
__name(parseExpression, "parseExpression");
var ExpressionParser = class {
  static {
    __name(this, "ExpressionParser");
  }
  tokens;
  pos = 0;
  constructor(tokens) {
    this.tokens = tokens;
  }
  parse() {
    const expr = this.parseConditional();
    if (this.peek().type !== "eof") {
      throw new Error("Unexpected token after expression");
    }
    return expr;
  }
  parseConditional() {
    const expr = this.parseLogicalOr();
    const token = this.peek();
    if (token.type === "punctuation" && token.value === "?") {
      this.advance();
      const consequent = this.parseConditional();
      const colonToken = this.peek();
      if (colonToken.type !== "punctuation" || colonToken.value !== ":") {
        throw new Error('Expected ":" in conditional expression');
      }
      this.advance();
      const alternate = this.parseConditional();
      return {
        type: "conditional",
        test: expr,
        consequent,
        alternate
      };
    }
    return expr;
  }
  parseLogicalOr() {
    let left = this.parseLogicalAnd();
    while (true) {
      const token = this.peek();
      if (token.type === "operator" && token.value === "||") {
        this.advance();
        const right = this.parseLogicalAnd();
        left = { type: "binary", operator: "||", left, right };
      } else {
        break;
      }
    }
    return left;
  }
  parseLogicalAnd() {
    let left = this.parseEquality();
    while (true) {
      const token = this.peek();
      if (token.type === "operator" && token.value === "&&") {
        this.advance();
        const right = this.parseEquality();
        left = { type: "binary", operator: "&&", left, right };
      } else {
        break;
      }
    }
    return left;
  }
  parseEquality() {
    let left = this.parseRelational();
    while (true) {
      const token = this.peek();
      if (token.type === "operator" && (token.value === "==" || token.value === "!=")) {
        const op = token.value;
        this.advance();
        const right = this.parseRelational();
        left = { type: "binary", operator: op, left, right };
      } else {
        break;
      }
    }
    return left;
  }
  parseRelational() {
    let left = this.parseAdditive();
    while (true) {
      const token = this.peek();
      if (token.type === "operator" && (token.value === "<=" || token.value === ">=" || token.value === "<" || token.value === ">")) {
        const op = token.value;
        this.advance();
        const right = this.parseAdditive();
        left = { type: "binary", operator: op, left, right };
      } else {
        break;
      }
    }
    return left;
  }
  parseAdditive() {
    let left = this.parseMultiplicative();
    while (true) {
      const token = this.peek();
      if (token.type === "operator" && (token.value === "+" || token.value === "-")) {
        const op = token.value;
        this.advance();
        const right = this.parseMultiplicative();
        left = { type: "binary", operator: op, left, right };
      } else {
        break;
      }
    }
    return left;
  }
  parseMultiplicative() {
    let left = this.parseUnary();
    while (true) {
      const token = this.peek();
      if (token.type === "operator" && (token.value === "*" || token.value === "/" || token.value === "%")) {
        const op = token.value;
        this.advance();
        const right = this.parseUnary();
        left = { type: "binary", operator: op, left, right };
      } else {
        break;
      }
    }
    return left;
  }
  parseUnary() {
    const token = this.peek();
    if (token.type === "operator" && (token.value === "!" || token.value === "-" || token.value === "+")) {
      const op = token.value;
      this.advance();
      const argument = this.parseUnary();
      return { type: "unary", operator: op, argument };
    }
    return this.parsePostfix();
  }
  parsePostfix() {
    let expr = this.parsePrimary();
    while (true) {
      const token = this.peek();
      if (token.type === "punctuation" && token.value === ".") {
        this.advance();
        const propToken = this.peek();
        if (propToken.type !== "identifier") {
          throw new Error('Expected identifier after "."');
        }
        const property = propToken.name;
        this.advance();
        expr = { type: "member", object: expr, property };
      } else if (token.type === "punctuation" && token.value === "[") {
        this.advance();
        const index = this.parseConditional();
        const closeBracket = this.peek();
        if (closeBracket.type !== "punctuation" || closeBracket.value !== "]") {
          throw new Error('Expected "]"');
        }
        this.advance();
        expr = { type: "index", object: expr, index };
      } else if (token.type === "punctuation" && token.value === "(") {
        this.advance();
        const args = this.parseArgumentList();
        const closeParen = this.peek();
        if (closeParen.type !== "punctuation" || closeParen.value !== ")") {
          throw new Error('Expected ")"');
        }
        this.advance();
        const callee = expr.type === "identifier" ? expr.name : expr;
        expr = {
          type: "call",
          callee,
          arguments: args
        };
      } else {
        break;
      }
    }
    return expr;
  }
  parsePrimary() {
    const token = this.peek();
    if (token.type === "string") {
      this.advance();
      return { type: "literal", value: token.value };
    }
    if (token.type === "number") {
      this.advance();
      return { type: "literal", value: token.value };
    }
    if (token.type === "keyword") {
      this.advance();
      if (token.value === "true") {
        return { type: "literal", value: true };
      } else if (token.value === "false") {
        return { type: "literal", value: false };
      } else if (token.value === "null") {
        return { type: "literal", value: null };
      }
    }
    if (token.type === "identifier") {
      this.advance();
      return { type: "identifier", name: token.name };
    }
    if (token.type === "punctuation" && token.value === "(") {
      this.advance();
      const expr = this.parseConditional();
      const closeParen = this.peek();
      if (closeParen.type !== "punctuation" || closeParen.value !== ")") {
        throw new Error('Expected ")"');
      }
      this.advance();
      return expr;
    }
    throw new Error(`Unexpected token: ${JSON.stringify(token)}`);
  }
  parseArgumentList() {
    const args = [];
    const token = this.peek();
    if (token.type === "punctuation" && token.value === ")") {
      return args;
    }
    while (true) {
      args.push(this.parseConditional());
      const nextToken = this.peek();
      if (nextToken.type === "punctuation" && nextToken.value === ",") {
        this.advance();
      } else {
        break;
      }
    }
    return args;
  }
  peek() {
    return this.tokens[this.pos] || { type: "eof" };
  }
  advance() {
    this.pos++;
  }
};
var TemplateParser = class {
  static {
    __name(this, "TemplateParser");
  }
  tokens;
  pos = 0;
  constructor(tokens) {
    this.tokens = tokens;
  }
  parse() {
    const ast = [];
    while (!this.isAtEnd()) {
      const token = this.parseToken();
      if (token) {
        ast.push(token);
      }
    }
    return ast;
  }
  parseToken() {
    const current = this.peek();
    if (!current) return null;
    if (current.type === "text") {
      this.advance();
      return { type: "text", value: current.value };
    }
    if (current.type === "tag") {
      this.advance();
      return this.parseTag(current.content);
    }
    return null;
  }
  parseTag(content) {
    const trimmed = content.trim();
    if (trimmed.startsWith("if ")) {
      return this.parseIfBlock(trimmed);
    }
    if (trimmed.startsWith("for ")) {
      return this.parseForBlock(trimmed);
    }
    if (trimmed.startsWith("while ")) {
      return this.parseWhileBlock(trimmed);
    }
    if (trimmed.startsWith("repeat ")) {
      return this.parseRepeatBlock(trimmed);
    }
    try {
      const expression = parseExpression(trimmed);
      return { type: "expression", expression };
    } catch (e) {
      return { type: "text", value: `{${content}}` };
    }
  }
  parseIfBlock(header) {
    const condition = parseExpression(header.slice(3).trim());
    const consequent = [];
    const elseIfs = [];
    let alternate;
    let currentBlock = consequent;
    while (!this.isAtEnd()) {
      const current = this.peek();
      if (current?.type === "tag") {
        const tag = current.content.trim();
        if (tag === "/if") {
          this.advance();
          break;
        } else if (tag.startsWith("elseif ")) {
          this.advance();
          const elseIfCondition = parseExpression(tag.slice(7).trim());
          const elseIfBlock = [];
          elseIfs.push({
            condition: elseIfCondition,
            consequent: elseIfBlock
          });
          currentBlock = elseIfBlock;
          continue;
        } else if (tag === "else") {
          this.advance();
          alternate = [];
          currentBlock = alternate;
          continue;
        }
      }
      const token = this.parseToken();
      if (token) {
        currentBlock.push(token);
      }
    }
    return {
      type: "if",
      condition,
      consequent,
      elseIfs: elseIfs.length > 0 ? elseIfs : void 0,
      alternate
    };
  }
  parseForBlock(header) {
    const content = header.slice(4).trim();
    const inIndex = content.indexOf(" in ");
    if (inIndex === -1) {
      return { type: "text", value: `{${header}}` };
    }
    const variable = content.slice(0, inIndex).trim();
    const iterableContent = content.slice(inIndex + 4).trim();
    const iterable = parseExpression(iterableContent);
    const body = [];
    while (!this.isAtEnd()) {
      const current = this.peek();
      if (current?.type === "tag" && current.content.trim() === "/for") {
        this.advance();
        break;
      }
      const token = this.parseToken();
      if (token) {
        body.push(token);
      }
    }
    return { type: "for", variable, iterable, body };
  }
  parseWhileBlock(header) {
    const condition = parseExpression(header.slice(6).trim());
    const body = [];
    while (!this.isAtEnd()) {
      const current = this.peek();
      if (current?.type === "tag" && current.content.trim() === "/while") {
        this.advance();
        break;
      }
      const token = this.parseToken();
      if (token) {
        body.push(token);
      }
    }
    return { type: "while", condition, body };
  }
  parseRepeatBlock(header) {
    const count = parseExpression(header.slice(7).trim());
    const body = [];
    while (!this.isAtEnd()) {
      const current = this.peek();
      if (current?.type === "tag" && current.content.trim() === "/repeat") {
        this.advance();
        break;
      }
      const token = this.parseToken();
      if (token) {
        body.push(token);
      }
    }
    return { type: "repeat", count, body };
  }
  peek() {
    return this.tokens[this.pos];
  }
  advance() {
    this.pos++;
  }
  isAtEnd() {
    return this.pos >= this.tokens.length;
  }
};

// src/lexer.ts
var LRUCache = class {
  static {
    __name(this, "LRUCache");
  }
  cache = /* @__PURE__ */ new Map();
  maxSize;
  constructor(maxSize) {
    this.maxSize = maxSize;
  }
  get(key) {
    const value = this.cache.get(key);
    if (value !== void 0) {
      this.cache.delete(key);
      this.cache.set(key, value);
    }
    return value;
  }
  set(key, value) {
    this.cache.delete(key);
    this.cache.set(key, value);
    if (this.cache.size > this.maxSize) {
      const firstKey = this.cache.keys().next().value;
      this.cache.delete(firstKey);
    }
  }
  clear() {
    this.cache.clear();
  }
};
var tokenizeCache = new LRUCache(1e3);
function tokenize(input) {
  const cached = tokenizeCache.get(input);
  if (cached) {
    return cached;
  }
  const lexer = new TemplateLexer(input);
  const lexerTokens = lexer.tokenize();
  const parser = new TemplateParser(lexerTokens);
  const tokens = parser.parse();
  tokenizeCache.set(input, tokens);
  return tokens;
}
__name(tokenize, "tokenize");
var TemplateLexer = class {
  static {
    __name(this, "TemplateLexer");
  }
  input;
  pos = 0;
  tokens = [];
  constructor(input) {
    this.input = input;
  }
  tokenize() {
    while (!this.isAtEnd()) {
      this.scanToken();
    }
    return this.tokens;
  }
  scanToken() {
    const char = this.peek();
    if (this.peek(2) === "{{") {
      this.addText("{");
      this.advance(2);
      return;
    }
    if (this.peek(2) === "}}") {
      this.addText("}");
      this.advance(2);
      return;
    }
    if (char === "{") {
      this.scanTag();
      return;
    }
    this.scanText();
  }
  scanText() {
    let text = "";
    while (!this.isAtEnd()) {
      const char = this.peek();
      const next2 = this.peek(2);
      if (char === "{" || next2 === "{{" || next2 === "}}") {
        break;
      }
      text += char;
      this.advance();
    }
    if (text) {
      this.addText(text);
    }
  }
  scanTag() {
    this.advance();
    const content = this.readTagContent();
    if (this.peek() === "}") {
      this.advance();
      this.tokens.push({ type: "tag", content });
    } else {
      this.addText("{" + content);
    }
  }
  readTagContent() {
    let content = "";
    let depth = 1;
    let inString = false;
    let stringDelim = "";
    while (!this.isAtEnd() && depth > 0) {
      const char = this.peek();
      if (inString) {
        content += char;
        this.advance();
        if (char === "\\" && !this.isAtEnd()) {
          content += this.peek();
          this.advance();
        } else if (char === stringDelim) {
          inString = false;
          stringDelim = "";
        }
      } else {
        if (char === '"' || char === "'") {
          inString = true;
          stringDelim = char;
          content += char;
          this.advance();
        } else if (char === "{") {
          depth++;
          content += char;
          this.advance();
        } else if (char === "}") {
          depth--;
          if (depth > 0) {
            content += char;
            this.advance();
          }
        } else {
          content += char;
          this.advance();
        }
      }
    }
    return content;
  }
  addText(value) {
    const last = this.tokens[this.tokens.length - 1];
    if (last && last.type === "text") {
      last.value += value;
    } else {
      this.tokens.push({ type: "text", value });
    }
  }
  peek(count = 1) {
    if (count === 1) {
      return this.input[this.pos] || "";
    }
    return this.input.slice(this.pos, this.pos + count);
  }
  advance(count = 1) {
    this.pos += count;
  }
  isAtEnd() {
    return this.pos >= this.input.length;
  }
};
var ExpressionLexer = class {
  static {
    __name(this, "ExpressionLexer");
  }
  input;
  pos = 0;
  constructor(input) {
    this.input = input.trim();
  }
  tokenize() {
    const tokens = [];
    while (!this.isAtEnd()) {
      this.skipWhitespace();
      if (this.isAtEnd()) break;
      const token = this.scanToken();
      if (token) {
        tokens.push(token);
      }
    }
    tokens.push({ type: "eof" });
    return tokens;
  }
  scanToken() {
    const char = this.peek();
    if (char === '"' || char === "'") {
      return this.scanStringLiteral();
    }
    if (this.isDigit(char)) {
      return this.scanNumberLiteral();
    }
    if (this.isIdentifierStart(char)) {
      return this.scanIdentifierOrKeyword();
    }
    const next2 = this.peek(2);
    if (next2 === "==" || next2 === "!=" || next2 === "<=" || next2 === ">=" || next2 === "&&" || next2 === "||") {
      this.advance(2);
      return { type: "operator", value: next2 };
    }
    if (char === "+" || char === "-" || char === "*" || char === "/" || char === "%" || char === "<" || char === ">" || char === "!") {
      this.advance();
      return { type: "operator", value: char };
    }
    if (char === "(" || char === ")" || char === "[" || char === "]" || char === "." || char === "," || char === "?" || char === ":") {
      this.advance();
      return { type: "punctuation", value: char };
    }
    throw new Error(`Unexpected character: ${char}`);
  }
  scanStringLiteral() {
    const quote = this.peek();
    this.advance();
    let value = "";
    while (this.pos < this.input.length) {
      const char = this.peek();
      if (char === quote) {
        this.advance();
        return { type: "string", value };
      } else if (char === "\\") {
        this.advance();
        const escaped = this.peek();
        switch (escaped) {
          case "n":
            value += "\n";
            break;
          case "t":
            value += "	";
            break;
          case "r":
            value += "\r";
            break;
          case "\\":
            value += "\\";
            break;
          case '"':
            value += '"';
            break;
          case "'":
            value += "'";
            break;
          default:
            value += escaped;
        }
        this.advance();
      } else {
        value += char;
        this.advance();
      }
    }
    throw new Error("Unterminated string literal");
  }
  scanNumberLiteral() {
    let value = "";
    while (this.pos < this.input.length && this.isDigit(this.peek())) {
      value += this.peek();
      this.advance();
    }
    if (this.peek() === ".") {
      value += ".";
      this.advance();
      while (this.pos < this.input.length && this.isDigit(this.peek())) {
        value += this.peek();
        this.advance();
      }
    }
    return { type: "number", value: parseFloat(value) };
  }
  scanIdentifierOrKeyword() {
    let name = "";
    while (this.pos < this.input.length && this.isIdentifierPart(this.peek())) {
      name += this.peek();
      this.advance();
    }
    if (name === "true" || name === "false" || name === "null") {
      return { type: "keyword", value: name };
    }
    return { type: "identifier", name };
  }
  peek(count = 1) {
    if (count === 1) {
      return this.input[this.pos] || "";
    }
    return this.input.slice(this.pos, this.pos + count);
  }
  advance(count = 1) {
    this.pos += count;
  }
  skipWhitespace() {
    while (this.pos < this.input.length && /\s/.test(this.input[this.pos])) {
      this.pos++;
    }
  }
  isAtEnd() {
    return this.pos >= this.input.length;
  }
  isDigit(char) {
    return char >= "0" && char <= "9";
  }
  isIdentifierStart(char) {
    return char >= "a" && char <= "z" || char >= "A" && char <= "Z" || char === "_" || char === "$";
  }
  isIdentifierPart(char) {
    return this.isIdentifierStart(char) || this.isDigit(char);
  }
};

// src/evaluator.ts
async function evaluateExpression(expr, variables, functionProviders, configurable) {
  switch (expr.type) {
    case "literal":
      return expr.value;
    case "identifier": {
      const value = variables[expr.name];
      if (value === void 0) {
        const func = functionProviders[expr.name];
        if (func) {
          try {
            return await func([], variables, configurable);
          } catch (e) {
            console.warn(
              `Failed to call function "${expr.name}" as variable:`,
              e
            );
            return "";
          }
        }
        return "";
      }
      if (typeof value === "function") {
        return value();
      }
      return value;
    }
    case "member": {
      const obj = await evaluateExpression(
        expr.object,
        variables,
        functionProviders,
        configurable
      );
      if (obj == null) return "";
      const value = obj[expr.property];
      if (typeof value === "function") {
        return value.call(obj);
      }
      return value ?? "";
    }
    case "index": {
      const obj = await evaluateExpression(
        expr.object,
        variables,
        functionProviders,
        configurable
      );
      const index = await evaluateExpression(
        expr.index,
        variables,
        functionProviders,
        configurable
      );
      if (obj == null) return "";
      const value = obj[index];
      if (typeof value === "function") {
        return value.call(obj);
      }
      return value ?? "";
    }
    case "call": {
      let func;
      if (typeof expr.callee === "string") {
        func = functionProviders[expr.callee];
        if (!func) {
          throw new Error(`Unknown function: ${expr.callee}`);
        }
      } else {
        const callableValue = await evaluateExpression(
          expr.callee,
          variables,
          functionProviders,
          configurable
        );
        if (typeof callableValue === "function") {
          func = /* @__PURE__ */ __name(async (args2, vars, cfg) => {
            return await callableValue(...args2);
          }, "func");
        } else {
          throw new Error(
            `Expression is not callable: ${JSON.stringify(expr.callee)}`
          );
        }
      }
      const args = await Promise.all(
        expr.arguments.map(
          (arg) => evaluateExpression(
            arg,
            variables,
            functionProviders,
            configurable
          )
        )
      );
      const stringArgs = args.map((arg) => String(arg));
      return await func(stringArgs, variables, configurable);
    }
    case "binary": {
      const left = await evaluateExpression(
        expr.left,
        variables,
        functionProviders,
        configurable
      );
      const right = await evaluateExpression(
        expr.right,
        variables,
        functionProviders,
        configurable
      );
      switch (expr.operator) {
        case "+":
          return left + right;
        case "-":
          return left - right;
        case "*":
          return left * right;
        case "/":
          return left / right;
        case "%":
          return left % right;
        case "==":
          return left === right;
        case "!=":
          return left !== right;
        case "<":
          return left < right;
        case ">":
          return left > right;
        case "<=":
          return left <= right;
        case ">=":
          return left >= right;
        case "&&":
          return left && right;
        case "||":
          return left || right;
        default:
          throw new Error(`Unknown operator: ${expr.operator}`);
      }
    }
    case "unary": {
      const argument = await evaluateExpression(
        expr.argument,
        variables,
        functionProviders,
        configurable
      );
      switch (expr.operator) {
        case "!":
          return !argument;
        case "-":
          return -argument;
        case "+":
          return +argument;
        default:
          throw new Error(`Unknown operator: ${expr.operator}`);
      }
    }
    case "conditional": {
      const test = await evaluateExpression(
        expr.test,
        variables,
        functionProviders,
        configurable
      );
      if (test) {
        return await evaluateExpression(
          expr.consequent,
          variables,
          functionProviders,
          configurable
        );
      } else {
        return await evaluateExpression(
          expr.alternate,
          variables,
          functionProviders,
          configurable
        );
      }
    }
    default:
      throw new Error(`Unknown expression type: ${expr.type}`);
  }
}
__name(evaluateExpression, "evaluateExpression");
async function isTruthy(expr, variables, functionProviders, configurable) {
  const value = await evaluateExpression(
    expr,
    variables,
    functionProviders,
    configurable
  );
  return !!value;
}
__name(isTruthy, "isTruthy");

// src/utils.ts
async function renderInternal(source, variables, extensions, configurable, detectedVariables, currentDepth, maxDepth) {
  if (currentDepth >= maxDepth) {
    return source;
  }
  const allVariables = currentDepth === 0 ? getAllVariables(variables, extensions.variableProviders) : variables;
  const functionProviders = extensions.functionProviders ?? {};
  const tokens = tokenize(source);
  const results = [];
  for (const token of tokens) {
    const result = await renderToken(
      token,
      allVariables,
      functionProviders,
      configurable,
      detectedVariables,
      currentDepth,
      maxDepth
    );
    results.push(result);
  }
  return results.join("");
}
__name(renderInternal, "renderInternal");
async function renderToken(token, variables, functionProviders, configurable, detectedVariables, currentDepth, maxDepth) {
  if (currentDepth >= maxDepth) {
    return "";
  }
  switch (token.type) {
    case "text":
      return token.value;
    case "expression": {
      try {
        const value = await evaluateExpression(
          token.expression,
          variables,
          functionProviders,
          configurable
        );
        collectVariables(token.expression, detectedVariables);
        return String(value ?? "");
      } catch (e) {
        return "";
      }
    }
    case "if": {
      const condition = await isTruthy(
        token.condition,
        variables,
        functionProviders,
        configurable
      );
      collectVariables(token.condition, detectedVariables);
      let branch;
      if (condition) {
        branch = token.consequent;
      } else if (token.elseIfs && token.elseIfs.length > 0) {
        for (const elseIf of token.elseIfs) {
          const elseIfCondition = await isTruthy(
            elseIf.condition,
            variables,
            functionProviders,
            configurable
          );
          collectVariables(elseIf.condition, detectedVariables);
          if (elseIfCondition) {
            branch = elseIf.consequent;
            break;
          }
        }
        if (!branch) {
          branch = token.alternate;
        }
      } else {
        branch = token.alternate;
      }
      if (!branch) return "";
      const results = [];
      for (const subToken of branch) {
        const result = await renderToken(
          subToken,
          variables,
          functionProviders,
          configurable,
          detectedVariables,
          currentDepth + 1,
          maxDepth
        );
        results.push(result);
      }
      return results.join("");
    }
    case "for": {
      const iterable = await evaluateExpression(
        token.iterable,
        variables,
        functionProviders,
        configurable
      );
      collectVariables(token.iterable, detectedVariables);
      if (!Array.isArray(iterable)) {
        return "";
      }
      const results = [];
      for (const item of iterable) {
        const loopVariables = Object.create(variables);
        loopVariables[token.variable] = item;
        const bodyResults = [];
        for (const subToken of token.body) {
          const result = await renderToken(
            subToken,
            loopVariables,
            functionProviders,
            configurable,
            detectedVariables,
            currentDepth + 1,
            maxDepth
          );
          bodyResults.push(result);
        }
        results.push(bodyResults.join(""));
      }
      return results.join("");
    }
    case "while": {
      const results = [];
      const MAX_ITERATIONS = 1e4;
      let iterationCount = 0;
      while (iterationCount < MAX_ITERATIONS) {
        const conditionValue = await isTruthy(
          token.condition,
          variables,
          functionProviders,
          configurable
        );
        collectVariables(token.condition, detectedVariables);
        if (!conditionValue) break;
        const bodyResults = [];
        for (const subToken of token.body) {
          const result = await renderToken(
            subToken,
            variables,
            functionProviders,
            configurable,
            detectedVariables,
            currentDepth + 1,
            maxDepth
          );
          bodyResults.push(result);
        }
        results.push(bodyResults.join(""));
        iterationCount++;
      }
      if (iterationCount >= MAX_ITERATIONS) {
        console.warn(
          `While loop exceeded maximum iterations (${MAX_ITERATIONS}). Breaking to prevent infinite loop.`
        );
      }
      return results.join("");
    }
    case "repeat": {
      const countValue = await evaluateExpression(
        token.count,
        variables,
        functionProviders,
        configurable
      );
      collectVariables(token.count, detectedVariables);
      const count = Number(countValue);
      if (isNaN(count) || count < 0) {
        console.warn(
          `Repeat count must be a number, got: ${countValue}. Skipping repeat.`
        );
        return "";
      }
      const results = [];
      const iterations = Math.floor(count);
      for (let i = 0; i < iterations; i++) {
        const bodyResults = [];
        for (const subToken of token.body) {
          const result = await renderToken(
            subToken,
            variables,
            functionProviders,
            configurable,
            detectedVariables,
            currentDepth + 1,
            maxDepth
          );
          bodyResults.push(result);
        }
        results.push(bodyResults.join(""));
      }
      return results.join("");
    }
    default:
      return "";
  }
}
__name(renderToken, "renderToken");
function collectVariables(expr, detectedVariables) {
  switch (expr.type) {
    case "identifier":
      detectedVariables.push(expr.name);
      break;
    case "member":
      collectVariables(expr.object, detectedVariables);
      break;
    case "index":
      collectVariables(expr.object, detectedVariables);
      collectVariables(expr.index, detectedVariables);
      break;
    case "call":
      if (typeof expr.callee === "string") {
        detectedVariables.push(expr.callee);
      } else {
        collectVariables(expr.callee, detectedVariables);
      }
      expr.arguments.forEach(
        (arg) => collectVariables(arg, detectedVariables)
      );
      break;
    case "binary":
      collectVariables(expr.left, detectedVariables);
      collectVariables(expr.right, detectedVariables);
      break;
    case "unary":
      collectVariables(expr.argument, detectedVariables);
      break;
    case "conditional":
      collectVariables(expr.test, detectedVariables);
      collectVariables(expr.consequent, detectedVariables);
      collectVariables(expr.alternate, detectedVariables);
      break;
  }
}
__name(collectVariables, "collectVariables");
function getAllVariables(variables, variableProviders = []) {
  let allVariables = { ...variables };
  for (const provider of variableProviders) {
    const providerVariables = provider();
    allVariables = { ...allVariables, ...providerVariables };
  }
  return allVariables;
}
__name(getAllVariables, "getAllVariables");

// src/renderer.ts
var ChatLunaPromptRenderer = class {
  static {
    __name(this, "ChatLunaPromptRenderer");
  }
  _variableProviders = [];
  _functionProviders = {};
  _staticVariables = {};
  registerVariableProvider(provider) {
    this._variableProviders.push(provider);
    return () => {
      const index = this._variableProviders.indexOf(provider);
      if (index !== -1) {
        this._variableProviders.splice(index, 1);
      }
    };
  }
  registerFunctionProvider(name, provider) {
    this._functionProviders[name] = provider;
    return () => {
      delete this._functionProviders[name];
    };
  }
  setStaticVariable(name, value) {
    this._staticVariables[name] = value;
  }
  getStaticVariable(name) {
    return this._staticVariables[name];
  }
  removeStaticVariable(name) {
    delete this._staticVariables[name];
  }
  async render(source, variables = {}, options = {}) {
    const { extensions = {}, configurable = {}, maxDepth = 10 } = options;
    const mergedExtensions = {
      variableProviders: [
        ...this._variableProviders,
        ...extensions.variableProviders ?? []
      ],
      functionProviders: {
        ...this._functionProviders,
        ...extensions.functionProviders ?? {}
      }
    };
    variables = Object.assign({}, this._staticVariables, variables);
    const detectedVariables = [];
    const result = await renderInternal(
      source,
      variables,
      mergedExtensions,
      configurable,
      detectedVariables,
      0,
      maxDepth
    );
    return {
      text: result,
      variables: Array.from(new Set(detectedVariables))
    };
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ChatLunaPromptRenderer
});
