// koishi-plugin-chatluna/lib/services/chat.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/locales/zh-CN.schema.plugin.yml
var require_zh_CN_schema_plugin = __commonJS({
  "src/locales/zh-CN.schema.plugin.yml"(exports2, module2) {
    module2.exports = { $inner: [{ $desc: "全局配置", $inner: { chatConcurrentMaxSize: "当前适配器支持的模型最大并发聊天数。", chatTimeLimit: "每小时的 API 调用次数限制。", maxRetries: "模型请求失败后的最大重试次数。", timeout: "模型请求的超时时间（毫秒）。", configMode: { $desc: "请求配置模式。", $inner: ["顺序配置模式（当前配置无效时，自动切换至下一个可用配置）。", "负载均衡模式（轮询使用所有可用配置）。"] }, proxyMode: { $desc: "当前插件的代理设置模式。", $inner: ["遵循 ChatLuna 主插件的全局代理设置", "禁用代理", "使用自定义代理设置"] } } }, { proxyAddress: "当前插件的自定义代理地址。仅在使用自定义代理设置时生效。" }] };
  }
});

// src/locales/en-US.schema.plugin.yml
var require_en_US_schema_plugin = __commonJS({
  "src/locales/en-US.schema.plugin.yml"(exports2, module2) {
    module2.exports = { $inner: [{ $desc: "Global Configuration", $inner: { chatConcurrentMaxSize: "Max concurrent chats for current adapter models.", chatTimeLimit: "API calls per hour limit (calls/hour).", maxRetries: "Max retries on model request failure.", timeout: "Model request timeout (ms).", configMode: { $desc: "Request config mode.", $inner: ["Sequential (auto-switch to next valid config on failure).", "Load balancing (rotate through all available configs)."] }, proxyMode: { $desc: "Plugin proxy mode.", $inner: ["Use ChatLuna global proxy settings", "Disable proxy", "Use custom proxy"] } } }, { proxyAddress: 'Custom proxy for plugin. Only used when custom proxy mode is selected. (e.g., "http://127.0.0.1:7897" or "socks5://proxy.example.com:1080")' }] };
  }
});

// src/services/chat.ts
var chat_exports = {};
__export(chat_exports, {
  AdminRequiredError: () => AdminRequiredError,
  ChatLunaContextManagerService: () => ChatLunaContextManagerService,
  ChatLunaPlugin: () => ChatLunaPlugin,
  ChatLunaPromptRenderService: () => ChatLunaPromptRenderService,
  ChatLunaService: () => ChatLunaService,
  ConstraintDisabledError: () => ConstraintDisabledError,
  ConstraintFixedError: () => ConstraintFixedError,
  ConstraintLockedError: () => ConstraintLockedError,
  ConversationNotFoundError: () => ConversationNotFoundError,
  ConversationResolutionError: () => ConversationResolutionError,
  InvalidChatModeError: () => InvalidChatModeError,
  MessageTransformer: () => MessageTransformer,
  STAGE_ORDER: () => STAGE_ORDER,
  applyPresetLane: () => applyPresetLane,
  computeBaseBindingKey: () => computeBaseBindingKey,
  getBaseBindingKey: () => getBaseBindingKey,
  getPresetLane: () => getPresetLane,
  toMessages: () => toMessages
});
module.exports = __toCommonJS(chat_exports);
var import_manager = require("@langchain/core/callbacks/manager");
var import_fs = __toESM(require("fs"), 1);
var import_path3 = __toESM(require("path"), 1);
var import_koishi5 = require("koishi");
var import_count_tokens2 = require("koishi-plugin-chatluna/llm-core/utils/count_tokens");
var import_app = require("koishi-plugin-chatluna/llm-core/chat/app");

// src/cache.ts
var import_koishi = require("koishi");
var Cache = class {
  constructor(ctx, config, tableName) {
    this.config = config;
    this.tableName = tableName;
    this._cache = new DatabaseCache(ctx);
    ctx.on("ready", async () => {
      await this._cache.clear("chathub/keys");
      await this._cache.clear("chathub/chat_limit");
    });
  }
  static {
    __name(this, "Cache");
  }
  _cache;
  get(tableNameOrId, id) {
    if (typeof id === "string") {
      return this._cache.get(tableNameOrId, id);
    }
    return this._cache.get(this.tableName, tableNameOrId);
  }
  set(tableNameOrId, idOrValue, value) {
    if (value != null) {
      return this._cache.set(tableNameOrId, idOrValue, value);
    }
    return this._cache.set(this.tableName, tableNameOrId, idOrValue);
  }
  delete(tableNameOrId, id) {
    if (typeof id === "string") {
      return this._cache.delete(tableNameOrId, id);
    }
    return this._cache.delete(this.tableName, tableNameOrId);
  }
  async clear(tableName) {
    if (tableName) {
      await this._cache.clear(tableName);
    } else {
      await this._cache.clear(this.tableName);
    }
  }
};
var DatabaseCache = class {
  constructor(ctx) {
    this.ctx = ctx;
    ctx.model.extend(
      "cache",
      {
        table: "string(63)",
        key: "string(63)",
        value: "text",
        expire: "timestamp"
      },
      {
        primary: ["table", "key"]
      }
    );
    ctx.setInterval(async () => {
      await ctx.database.remove("cache", { expire: { $lt: /* @__PURE__ */ new Date() } });
    }, 10 * import_koishi.Time.minute);
  }
  static {
    __name(this, "DatabaseCache");
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  encode(data) {
    return JSON.stringify(data);
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  decode(record) {
    return JSON.parse(record);
  }
  async clear(table) {
    await this.ctx.database.remove("cache", { table });
  }
  async get(table, key) {
    const [entry] = await this.ctx.database.get("cache", { table, key }, [
      "expire",
      "value"
    ]);
    if (!entry) return;
    if (entry.expire && +entry.expire < Date.now()) return;
    return this.decode(entry.value);
  }
  async set(table, key, value, maxAge = 1e3 * 60 * 60 * 24) {
    const expire = maxAge ? new Date(Date.now() + maxAge) : null;
    await this.ctx.database.upsert("cache", [
      {
        table,
        key,
        value: this.encode(value),
        expire
      }
    ]);
  }
  async delete(table, key) {
    await this.ctx.database.remove("cache", { table, key });
  }
  async *keys(table) {
    const entries = await this.ctx.database.get("cache", { table }, [
      "expire",
      "key"
    ]);
    yield* entries.filter((entry) => !entry.expire || +entry.expire > Date.now()).map((entry) => entry.key);
  }
  async *values(table) {
    const entries = await this.ctx.database.get("cache", { table }, [
      "expire",
      "value"
    ]);
    yield* entries.filter((entry) => !entry.expire || +entry.expire > Date.now()).map((entry) => this.decode(entry.value));
  }
  async *entries(table) {
    const entries = await this.ctx.database.get("cache", { table }, [
      "expire",
      "key",
      "value"
    ]);
    yield* entries.filter((entry) => !entry.expire || +entry.expire > Date.now()).map((entry) => [entry.key, this.decode(entry.value)]);
  }
};

// src/chains/chain.ts
var import_koishi2 = require("koishi");
var import_error = require("koishi-plugin-chatluna/utils/error");
var import_logger = require("koishi-plugin-chatluna/utils/logger");

// src/middlewares/system/lifecycle.ts
var lifecycleNames = [
  "lifecycle-check",
  "lifecycle-prepare",
  "lifecycle-handle_command",
  "lifecycle-request_conversation",
  "lifecycle-send"
];

// src/utils/time.ts
function formatDuration(ms) {
  if (ms < 1e3) {
    return `${ms}ms`;
  }
  const totalSeconds = Math.floor(ms / 1e3);
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor(totalSeconds % 3600 / 60);
  const seconds = totalSeconds % 60;
  if (hours > 0) {
    return `${hours}h ${minutes}m ${seconds}s`;
  }
  if (minutes > 0) {
    return `${minutes}m ${seconds}s`;
  }
  const decimal = (ms / 1e3).toFixed(2);
  return `${decimal}s`;
}
__name(formatDuration, "formatDuration");

// src/chains/chain.ts
var logger;
var ChatChain = class {
  constructor(ctx, config) {
    this.ctx = ctx;
    this.config = config;
    logger = (0, import_logger.createLogger)(ctx);
    this._graph = new ChatChainDependencyGraph();
    this._senders = [];
    const defaultChatChainSender = new DefaultChatChainSender(config);
    this._senders.push(
      (session, messages, context) => defaultChatChainSender.send(session, messages, context)
    );
  }
  static {
    __name(this, "ChatChain");
  }
  _graph;
  _senders;
  isSetErrorMessage = false;
  _createRecallThinkingMessage(context) {
    return async () => {
      if (!context.options?.thinkingTimeoutObject) return;
      const timeoutObj = context.options.thinkingTimeoutObject;
      clearTimeout(timeoutObj.timeout);
      timeoutObj.autoRecallTimeout && clearTimeout(timeoutObj.autoRecallTimeout);
      timeoutObj.recallFunc && await timeoutObj.recallFunc();
      timeoutObj.timeout = null;
      context.options.thinkingTimeoutObject = void 0;
    };
  }
  async receiveMessage(session, ctx) {
    const context = {
      config: this.config,
      message: session.content,
      ctx: ctx ?? this.ctx,
      session,
      options: {
        startedAt: Date.now()
      },
      send: /* @__PURE__ */ __name((message) => this.sendMessage(session, message, context), "send"),
      recallThinkingMessage: this._createRecallThinkingMessage(
        {}
      )
    };
    context.recallThinkingMessage = this._createRecallThinkingMessage(context);
    const result = await this._runMiddleware(session, context);
    await context.recallThinkingMessage();
    return result;
  }
  async receiveCommand(session, command, options = {}, ctx = this.ctx) {
    const context = {
      config: this.config,
      message: options?.message ?? session.content,
      ctx,
      session,
      command,
      send: /* @__PURE__ */ __name((message) => this.sendMessage(session, message, context), "send"),
      recallThinkingMessage: this._createRecallThinkingMessage(
        {}
      ),
      options: {
        ...options,
        startedAt: Date.now()
      }
    };
    context.recallThinkingMessage = this._createRecallThinkingMessage(context);
    const result = await this._runMiddleware(session, context);
    await context.recallThinkingMessage();
    return result;
  }
  middleware(name, middleware, ctx = this.ctx) {
    const result = new ChainMiddleware(name, middleware, this._graph);
    this._graph.addNode(result);
    const dispose = /* @__PURE__ */ __name(() => this._graph.removeNode(name), "dispose");
    ctx.effect(() => dispose);
    return result;
  }
  sender(sender) {
    this._senders.push(sender);
  }
  async _runMiddleware(session, context) {
    if (!this.isSetErrorMessage) {
      (0, import_error.setErrorFormatTemplate)(session.text("chatluna.error_message"));
      this.isSetErrorMessage = true;
    }
    const originMessage = context.message;
    const runLevels = this._graph.build();
    if (runLevels.length === 0) {
      return false;
    }
    let isOutputLog = false;
    for (const level of runLevels) {
      const results = await this._executeLevel(level, session, context);
      for (const result of results) {
        if (result.status === "stop") {
          await this._handleStopStatus(
            session,
            context,
            originMessage,
            isOutputLog
          );
          return false;
        }
        if (result.status === "error") {
          await this._handleMiddlewareError(
            session,
            result.middlewareName,
            result.error
          );
          return false;
        }
        if (result.output instanceof Array || typeof result.output === "string") {
          context.message = result.output;
        }
        if (result.shouldLog) {
          isOutputLog = true;
        }
      }
    }
    if (isOutputLog) {
      logger.debug("-".repeat(40) + "\n");
    }
    if (context.message != null && context.message !== originMessage) {
      await this.sendMessage(session, context.message, context);
    }
    return true;
  }
  async _executeLevel(middlewares, session, context) {
    const abortController = new AbortController();
    const results = [];
    let hasStopRequest = false;
    let hasError = false;
    const promises = middlewares.map(async (middleware, index) => {
      try {
        if (abortController.signal.aborted) {
          return {
            status: "success",
            output: 0 /* SKIPPED */,
            middlewareName: middleware.name,
            shouldLog: false
          };
        }
        const result = await this._executeMiddleware(
          middleware,
          session,
          context,
          abortController.signal
        );
        if (result.status === "stop" && !hasStopRequest) {
          hasStopRequest = true;
          abortController.abort();
        }
        if (result.status === "error" && !hasError) {
          hasError = true;
          abortController.abort();
        }
        results[index] = result;
        return result;
      } catch (error) {
        const errorResult = {
          status: "error",
          error,
          middlewareName: middleware.name,
          shouldLog: false
        };
        if (!hasError) {
          hasError = true;
          abortController.abort();
        }
        results[index] = errorResult;
        return errorResult;
      }
    });
    await Promise.all(promises);
    return results.filter((result) => result !== void 0);
  }
  async _executeMiddleware(middleware, session, context, abortSignal) {
    const startTime = Date.now();
    try {
      if (abortSignal?.aborted) {
        return {
          status: "success",
          output: 0 /* SKIPPED */,
          middlewareName: middleware.name,
          shouldLog: false
        };
      }
      const result = await middleware.run(session, context);
      const executionTime = Date.now() - startTime;
      const shouldLogTime = !middleware.name.startsWith("lifecycle-") && result !== 0 /* SKIPPED */ && middleware.name !== "allow_reply" && executionTime > 100;
      if (shouldLogTime) {
        logger.debug(
          `middleware %c executed in %s`,
          middleware.name,
          formatDuration(executionTime)
        );
      }
      if (result === 1 /* STOP */) {
        return {
          status: "stop",
          middlewareName: middleware.name,
          shouldLog: shouldLogTime
        };
      }
      return {
        status: "success",
        output: result,
        middlewareName: middleware.name,
        shouldLog: shouldLogTime
      };
    } catch (error) {
      return {
        status: "error",
        error,
        middlewareName: middleware.name,
        shouldLog: false
      };
    }
  }
  async sendMessage(session, message, context) {
    const messages = message instanceof Array ? message : [message];
    for (const sender of this._senders) {
      await sender(session, messages, context);
    }
  }
  async _handleStopStatus(session, context, originMessage, isOutputLog) {
    if (context.message != null && context.message !== originMessage) {
      await this.sendMessage(session, context.message, context);
    }
    if (isOutputLog) {
      logger.debug("-".repeat(40) + "\n");
    }
  }
  async _handleMiddlewareError(session, middlewareName, error) {
    if (error instanceof import_error.ChatLunaError) {
      const message = error.errorCode === import_error.ChatLunaErrorCode.ABORTED ? session.text("chatluna.aborted") : error.message;
      await this.sendMessage(session, message, void 0);
      return;
    }
    logger.error(`chat-chain: ${middlewareName} error ${error}`);
    logger.error(error);
    error.cause && logger.error(error.cause);
    logger.debug("-".repeat(40) + "\n");
    await this.sendMessage(
      session,
      session.text("chatluna.middleware_error", [
        middlewareName,
        error.message
      ]),
      void 0
    );
  }
};
var ChatChainDependencyGraph = class {
  static {
    __name(this, "ChatChainDependencyGraph");
  }
  _tasks = /* @__PURE__ */ new Map();
  _rules = [];
  _listeners = /* @__PURE__ */ new Map();
  _cachedOrder = null;
  _index = 0;
  addNode(middleware) {
    this._tasks.set(middleware.name, {
      name: middleware.name,
      middleware,
      index: this._index++
    });
    this._cachedOrder = null;
  }
  removeNode(name) {
    this._tasks.delete(name);
    for (let i = this._rules.length - 1; i >= 0; i--) {
      const rule = this._rules[i];
      if (rule.owner === name || rule.target === name) {
        this._rules.splice(i, 1);
      }
    }
    this._cachedOrder = null;
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  once(name, listener) {
    const listeners = this._listeners.get(name) ?? /* @__PURE__ */ new Set();
    listeners.add(listener);
    this._listeners.set(name, listeners);
  }
  before(taskA, taskB) {
    if (taskA instanceof ChainMiddleware) {
      taskA = taskA.name;
    }
    if (taskB instanceof ChainMiddleware) {
      taskB = taskB.name;
    }
    if (taskA && taskB) {
      this._rules.push({
        owner: taskA,
        target: taskB,
        type: "before",
        source: this._captureRuleSource()
      });
      this._cachedOrder = null;
    } else {
      throw new Error("Invalid tasks");
    }
  }
  after(taskA, taskB) {
    if (taskA instanceof ChainMiddleware) {
      taskA = taskA.name;
    }
    if (taskB instanceof ChainMiddleware) {
      taskB = taskB.name;
    }
    if (taskA && taskB) {
      this._rules.push({
        owner: taskA,
        target: taskB,
        type: "after",
        source: this._captureRuleSource()
      });
      this._cachedOrder = null;
    } else {
      throw new Error("Invalid tasks");
    }
  }
  getDependencies(task) {
    const deps = /* @__PURE__ */ new Set();
    for (const rule of this._rules) {
      if (rule.owner === task && rule.type === "after") {
        deps.add(rule.target);
      }
      if (rule.target === task && rule.type === "before") {
        deps.add(rule.owner);
      }
    }
    return deps;
  }
  getDependents(task) {
    const dependents = [];
    for (const rule of this._rules) {
      if (rule.owner === task && rule.type === "before") {
        dependents.push(rule.target);
      }
      if (rule.target === task && rule.type === "after") {
        dependents.push(rule.owner);
      }
    }
    return dependents;
  }
  build() {
    if (this._cachedOrder) {
      return this._cachedOrder;
    }
    for (const [, listeners] of this._listeners) {
      for (const listener of listeners) {
        listener();
      }
      listeners.clear();
    }
    const lifecycleSet = new Set(lifecycleNames);
    const nodes = [...this._tasks.values()].sort(
      (a, b) => a.index - b.index
    );
    const nodeNames = new Set(nodes.map((node) => node.name));
    const normalNodes = nodes.filter((node) => !lifecycleSet.has(node.name));
    const ranges = /* @__PURE__ */ new Map();
    const outgoing = /* @__PURE__ */ new Map();
    const incoming = /* @__PURE__ */ new Map();
    const slots = /* @__PURE__ */ new Map();
    const slotCause = /* @__PURE__ */ new Map();
    for (const node of normalNodes) {
      ranges.set(node.name, {
        min: 0,
        max: lifecycleNames.length,
        minRules: [],
        maxRules: []
      });
      outgoing.set(node.name, []);
      incoming.set(node.name, []);
      slots.set(node.name, 0);
    }
    for (const rule of this._rules) {
      if (!nodeNames.has(rule.owner)) {
        continue;
      }
      if (!lifecycleSet.has(rule.target) && !nodeNames.has(rule.target)) {
        throw new Error(
          `Unknown middleware "${rule.target}" referenced by ${this._formatRule(rule)}`
        );
      }
      if (lifecycleSet.has(rule.owner) && lifecycleSet.has(rule.target)) {
        continue;
      }
      if (lifecycleSet.has(rule.target) || lifecycleSet.has(rule.owner)) {
        const name = lifecycleSet.has(rule.target) ? rule.owner : rule.target;
        if (lifecycleSet.has(name)) {
          continue;
        }
        const range = ranges.get(name);
        if (!range) {
          continue;
        }
        const lifecycleName = lifecycleSet.has(rule.target) ? rule.target : rule.owner;
        const idx = lifecycleNames.indexOf(lifecycleName);
        const isAfter = lifecycleSet.has(rule.target) ? rule.type === "after" : rule.type === "before";
        if (isAfter) {
          range.min = Math.max(range.min, idx + 1);
          range.minRules.push(rule);
        } else {
          range.max = Math.min(range.max, idx);
          range.maxRules.push(rule);
        }
        slots.set(name, range.min);
        continue;
      }
      const edge = rule.type === "before" ? {
        from: rule.owner,
        to: rule.target,
        rule
      } : {
        from: rule.target,
        to: rule.owner,
        rule
      };
      outgoing.get(edge.from)?.push(edge);
      incoming.get(edge.to)?.push(edge);
    }
    const stack = [];
    const onStack = /* @__PURE__ */ new Set();
    const index = /* @__PURE__ */ new Map();
    const lowLink = /* @__PURE__ */ new Map();
    const cycles = [];
    let cursor = 0;
    const visit = /* @__PURE__ */ __name((name) => {
      index.set(name, cursor);
      lowLink.set(name, cursor);
      cursor += 1;
      stack.push(name);
      onStack.add(name);
      for (const edge of outgoing.get(name) ?? []) {
        const next = edge.to;
        if (!index.has(next)) {
          visit(next);
          lowLink.set(
            name,
            Math.min(lowLink.get(name), lowLink.get(next))
          );
          continue;
        }
        if (onStack.has(next)) {
          lowLink.set(
            name,
            Math.min(lowLink.get(name), index.get(next))
          );
        }
      }
      if (lowLink.get(name) !== index.get(name)) {
        return;
      }
      const group = [];
      while (true) {
        const current = stack.pop();
        onStack.delete(current);
        group.push(current);
        if (current === name) {
          break;
        }
      }
      if (group.length > 1 || (outgoing.get(group[0]) ?? []).some(
        (edge) => edge.to === group[0]
      )) {
        cycles.push(group);
      }
    }, "visit");
    for (const node of normalNodes) {
      if (!index.has(node.name)) {
        visit(node.name);
      }
    }
    if (cycles.length > 0) {
      const cycle = cycles[0];
      const cycleSet = new Set(cycle);
      const blocked = /* @__PURE__ */ new Set();
      const queue = [...cycle];
      while (queue.length > 0) {
        const current = queue.shift();
        for (const edge of outgoing.get(current) ?? []) {
          if (cycleSet.has(edge.to) || blocked.has(edge.to)) {
            continue;
          }
          blocked.add(edge.to);
          queue.push(edge.to);
        }
      }
      const order = new Map(nodes.map((node) => [node.name, node.index]));
      const cycleEdges = cycle.flatMap((name) => outgoing.get(name) ?? []).filter((edge) => cycleSet.has(edge.to)).sort(
        (a, b) => (order.get(a.from) ?? 0) - (order.get(b.from) ?? 0) || (order.get(a.to) ?? 0) - (order.get(b.to) ?? 0)
      );
      const blockedList = [...blocked].sort(
        (a, b) => (order.get(a) ?? 0) - (order.get(b) ?? 0)
      );
      throw new Error(
        [
          "Circular dependency detected in middleware graph.",
          `Cycle nodes: ${cycle.sort(
            (a, b) => (order.get(a) ?? 0) - (order.get(b) ?? 0)
          ).join(" -> ")}`,
          "Constraints:",
          ...cycleEdges.map(
            (edge) => `- ${edge.from} -> ${edge.to} via ${this._formatRule(edge.rule)}`
          ),
          blockedList.length > 0 ? `Blocked nodes: ${blockedList.join(", ")}` : ""
        ].filter((line) => line.length > 0).join("\n")
      );
    }
    const indegree = /* @__PURE__ */ new Map();
    for (const node of normalNodes) {
      indegree.set(node.name, incoming.get(node.name)?.length ?? 0);
    }
    const ready = normalNodes.filter((node) => indegree.get(node.name) === 0).map((node) => node.name);
    const topo = [];
    while (ready.length > 0) {
      ready.sort(
        (a, b) => (this._tasks.get(a)?.index ?? 0) - (this._tasks.get(b)?.index ?? 0)
      );
      const current = ready.shift();
      const currentSlot = slots.get(current) ?? 0;
      const range = ranges.get(current);
      if (range && currentSlot > range.max) {
        const path4 = [];
        let name = current;
        const visited = /* @__PURE__ */ new Set();
        while (slotCause.has(name) && !visited.has(name)) {
          visited.add(name);
          const edge = slotCause.get(name);
          path4.unshift(
            `- ${edge.from} -> ${edge.to} via ${this._formatRule(edge.rule)}`
          );
          name = edge.from;
        }
        throw new Error(
          [
            `Cannot place middleware "${current}" in lifecycle order.`,
            `Resolved position: ${this._formatSlot(currentSlot)}`,
            `Latest allowed position: ${this._formatSlot(range.max)}`,
            range.minRules.length > 0 ? "Required after:" : "",
            ...range.minRules.map(
              (rule) => `- ${this._formatRule(rule)}`
            ),
            range.maxRules.length > 0 ? "Required before:" : "",
            ...range.maxRules.map(
              (rule) => `- ${this._formatRule(rule)}`
            ),
            path4.length > 0 ? "Dependency chain:" : "",
            ...path4
          ].filter((line) => line.length > 0).join("\n")
        );
      }
      topo.push(current);
      for (const edge of outgoing.get(current) ?? []) {
        if ((slots.get(edge.to) ?? 0) < currentSlot) {
          slots.set(edge.to, currentSlot);
          slotCause.set(edge.to, edge);
        }
        indegree.set(edge.to, (indegree.get(edge.to) ?? 0) - 1);
        if (indegree.get(edge.to) === 0) {
          ready.push(edge.to);
        }
      }
    }
    const levels = [];
    const slotGroups = /* @__PURE__ */ new Map();
    for (const name of topo) {
      const slot = slots.get(name) ?? 0;
      const group = slotGroups.get(slot) ?? [];
      group.push(name);
      slotGroups.set(slot, group);
    }
    for (let slot = 0; slot <= lifecycleNames.length; slot++) {
      const group = slotGroups.get(slot) ?? [];
      if (group.length > 0) {
        const groupSet = new Set(group);
        const groupIndegree = /* @__PURE__ */ new Map();
        for (const name of group) {
          groupIndegree.set(
            name,
            (incoming.get(name) ?? []).filter(
              (edge) => groupSet.has(edge.from)
            ).length
          );
        }
        let currentLevel = group.filter((name) => groupIndegree.get(name) === 0).sort(
          (a, b) => (this._tasks.get(a)?.index ?? 0) - (this._tasks.get(b)?.index ?? 0)
        );
        while (currentLevel.length > 0) {
          levels.push(
            currentLevel.map((name) => {
              const task = this._tasks.get(name);
              if (task?.middleware == null) {
                throw new Error(
                  `Missing middleware for task ${name}`
                );
              }
              return task.middleware;
            })
          );
          const nextLevel = [];
          for (const name of currentLevel) {
            for (const edge of outgoing.get(name) ?? []) {
              if (!groupSet.has(edge.to)) {
                continue;
              }
              groupIndegree.set(
                edge.to,
                (groupIndegree.get(edge.to) ?? 0) - 1
              );
              if (groupIndegree.get(edge.to) === 0) {
                nextLevel.push(edge.to);
              }
            }
          }
          currentLevel = nextLevel.sort(
            (a, b) => (this._tasks.get(a)?.index ?? 0) - (this._tasks.get(b)?.index ?? 0)
          );
        }
      }
      const lifecycle = this._tasks.get(lifecycleNames[slot])?.middleware;
      if (lifecycle) {
        levels.push([lifecycle]);
      }
    }
    this._cachedOrder = levels;
    return levels;
  }
  _captureRuleSource() {
    const stack = new Error().stack?.split("\n") ?? [];
    return stack.map((line) => line.trim()).find(
      (line) => line.length > 0 && line !== "Error" && !line.includes("ChatChainDependencyGraph.") && !line.includes("ChainMiddleware.") && !line.includes("chains\\chain.") && !line.includes("chains/chain.")
    );
  }
  _formatRule(rule) {
    const source = rule.source ? ` at ${rule.source}` : "";
    return `.${rule.type}('${rule.target}') declared by ${rule.owner}${source}`;
  }
  _formatSlot(slot) {
    if (slot <= 0) {
      return `before ${lifecycleNames[0]}`;
    }
    if (slot >= lifecycleNames.length) {
      return `after ${lifecycleNames[lifecycleNames.length - 1]}`;
    }
    return `between ${lifecycleNames[slot - 1]} and ${lifecycleNames[slot]}`;
  }
};
var ChainMiddleware = class {
  constructor(name, execute, graph) {
    this.name = name;
    this.execute = execute;
    this.graph = graph;
  }
  static {
    __name(this, "ChainMiddleware");
  }
  before(name) {
    this.graph.before(this.name, name);
    return this;
  }
  after(name) {
    this.graph.after(this.name, name);
    return this;
  }
  run(session, options) {
    return this.execute(session, options);
  }
};
var DefaultChatChainSender = class {
  constructor(config) {
    this.config = config;
  }
  static {
    __name(this, "DefaultChatChainSender");
  }
  processElements(elements) {
    return elements.filter((element) => {
      if (!element) return false;
      if (element.type === "img") {
        const src = element.attrs?.["src"];
        return !(typeof src === "string" && src.startsWith("attachment"));
      }
      return true;
    }).map((element) => {
      if (element.children?.length) {
        element.children = this.processElements(element.children);
      }
      return element;
    });
  }
  async send(session, messages, context) {
    if (!messages?.length) return;
    if (isElementArray(messages?.[0]) && messages[0][1]?.type === "markdown-qq") {
      await this.sendAsQQMarkdown(session, messages[0][0]);
      return;
    }
    if (this.config.isForwardMsg && this.getMessageText(messages).length > this.config.forwardMsgMinLength) {
      await this.sendAsForward(session, messages);
      return;
    }
    await this.sendAsNormal(session, messages, context);
  }
  async sendAsQQMarkdown(session, message) {
    const { user } = session.event;
    await session.bot.internal.sendPrivateMessage(
      user.id,
      {
        msg_type: 2,
        msg_seq: 1,
        msg_id: session.messageId,
        markdown: {
          content: message.attrs["content"]
        }
      }
    );
  }
  async sendAsForward(session, messages) {
    const sendMessages = this.convertToForwardMessages(messages);
    if (sendMessages.length < 1 || sendMessages.length === 1 && sendMessages.join().length === 0) {
      return;
    }
    await session.sendQueued(
      (0, import_koishi2.h)("message", { forward: true }, ...sendMessages)
    );
  }
  convertToForwardMessages(messages) {
    const firstMsg = messages[0];
    if (Array.isArray(firstMsg)) {
      return messages.map((msg) => (0, import_koishi2.h)("message", ...msg));
    }
    if (typeof firstMsg === "object") {
      return [(0, import_koishi2.h)("message", ...messages)];
    }
    if (typeof firstMsg === "string") {
      return [import_koishi2.h.text(firstMsg)];
    }
    throw new Error(`Unsupported message type: ${typeof firstMsg}`);
  }
  async sendAsNormal(session, messages, context) {
    for (const message of messages) {
      const messageFragment = await this.buildMessageFragment(
        session,
        message,
        context
      );
      if (!messageFragment?.length) continue;
      const processedFragment = this.processElements(messageFragment);
      await session.sendQueued(processedFragment);
    }
  }
  async buildMessageFragment(session, message, context) {
    const start = context?.options?.startedAt;
    const elapsed = start ? Date.now() - start : 0;
    const threshold = (this.config.replyQuoteThreshold ?? 0) * 1e3;
    const shouldAddQuote = this.config.isReplyWithAt && session.isDirect === false && session.messageId && elapsed >= threshold;
    const messageContent = this.convertMessageToArray(message);
    if (messageContent == null || messageContent.length < 1 || messageContent.length === 1 && messageContent.join().length === 0) {
      return;
    }
    if (!shouldAddQuote) {
      return messageContent;
    }
    const quote = (0, import_koishi2.h)("quote", { id: session.messageId });
    const hasIncompatibleType = messageContent.some(
      (element) => element.type === "audio" || element.type === "message"
    );
    return hasIncompatibleType ? messageContent : [quote, ...messageContent];
  }
  convertMessageToArray(message) {
    if (Array.isArray(message)) {
      return message;
    }
    if (typeof message === "string") {
      return [import_koishi2.h.text(message)];
    }
    return [message];
  }
  getMessageText(message) {
    return message.map((element) => {
      if (typeof element === "string") {
        return element;
      }
      if (Array.isArray(element)) {
        return import_koishi2.h.select(element, "text").toString();
      }
      return element.toString();
    }).join(" ");
  }
};
function isElementArray(value) {
  return Array.isArray(value) && value.every(
    (item) => typeof item === "object" && item.attrs && item.type
  );
}
__name(isElementArray, "isElementArray");

// src/services/chat.ts
var import_agent = require("koishi-plugin-chatluna/llm-core/agent");
var import_config = require("koishi-plugin-chatluna/llm-core/platform/config");
var import_model2 = require("koishi-plugin-chatluna/llm-core/platform/model");
var import_rerank = require("koishi-plugin-chatluna/llm-core/platform/rerank");
var import_service = require("koishi-plugin-chatluna/llm-core/platform/service");
var import_types6 = require("koishi-plugin-chatluna/llm-core/platform/types");
var import_preset = require("koishi-plugin-chatluna/preset");
var import_error5 = require("koishi-plugin-chatluna/utils/error");

// src/services/message_transform.ts
var import_koishi_plugin_chatluna = require("koishi-plugin-chatluna");
var import_error2 = require("koishi-plugin-chatluna/utils/error");
var import_string = require("koishi-plugin-chatluna/utils/string");
var MessageTransformer = class {
  constructor(_config) {
    this._config = _config;
  }
  static {
    __name(this, "MessageTransformer");
  }
  _beforeTransformFunctions = [];
  _transformFunctions = /* @__PURE__ */ new Map();
  async transform(session, elements, model, message = {
    content: "",
    name: session.username,
    additional_kwargs: {}
  }, options = {
    quote: false,
    includeQuoteReply: true
  }) {
    await this._runBeforeTransform(
      session,
      elements,
      message,
      model,
      options
    );
    const sourceElementString = elements.map((h2) => h2.toString(true)).join();
    const quoteElementString = ((session.quote && session.quote.elements) ?? []).map((h2) => h2.toString(true)).join();
    for (const element of elements) {
      await this._processElement(session, element, message, model);
    }
    if (session.quote && !options.quote && options.includeQuoteReply && sourceElementString !== quoteElementString) {
      const quoteMessage = await this.transform(
        session,
        session.quote.elements ?? [],
        model,
        {
          content: "",
          name: session.username,
          additional_kwargs: {}
        },
        {
          quote: true,
          includeQuoteReply: options.includeQuoteReply
        }
      );
      const extractText = /* @__PURE__ */ __name((content) => {
        if (typeof content === "string") return content;
        return Array.isArray(content) ? content.filter((item) => (0, import_string.isMessageContentText)(item)).map((item) => item.text).join("") : "";
      }, "extractText");
      const extractImages = /* @__PURE__ */ __name((content) => Array.isArray(content) ? content.filter((item) => (0, import_string.isMessageContentImageUrl)(item)) : [], "extractImages");
      const quoteText = extractText(quoteMessage.content);
      const quoteImages = extractImages(quoteMessage.content);
      const hasImages = extractImages(message.content).length > 0 || quoteImages.length > 0;
      const quoteUsername = session.quote.user?.name || session.quote.user?.id || "Unknown";
      const quoteTimestamp = session.quote.timestamp ? new Date(session.quote.timestamp).toLocaleString("zh-CN", {
        year: "numeric",
        month: "2-digit",
        day: "2-digit",
        hour: "2-digit",
        minute: "2-digit",
        second: "2-digit",
        hour12: false
      }) : "";
      const quoteSaid = session.text("chatluna.quote_said");
      const quoteHeader = quoteTimestamp ? `${quoteTimestamp} ${quoteUsername}` : quoteUsername;
      if (hasImages) {
        if (typeof message.content === "string") {
          message.content = message.content.trim().length > 0 ? [{ type: "text", text: message.content }] : [];
        }
        if (quoteText && quoteText !== "[image]") {
          const currentText = extractText(message.content);
          const quotedContent = `Referenced message: [${quoteHeader} ${quoteSaid}："${quoteText}"]

User's message: ${currentText}`;
          message.content = message.content.filter(
            (item) => item.type !== "text"
          );
          message.content.unshift({
            type: "text",
            text: quotedContent
          });
        }
        message.content = [...quoteImages, ...message.content];
      } else if (quoteText && quoteText !== "[image]") {
        const currentText = extractText(message.content);
        message.content = `Referenced message: [${quoteHeader} ${quoteSaid}："${quoteText}"]

User's message: ${currentText}`;
      }
    }
    return message;
  }
  before(transformFunction, priority = 0) {
    const wrapper = {
      func: transformFunction,
      priority
    };
    const insertIndex = this._beforeTransformFunctions.findIndex(
      (item) => item.priority > priority
    );
    if (insertIndex === -1) {
      this._beforeTransformFunctions.push(wrapper);
    } else {
      this._beforeTransformFunctions.splice(insertIndex, 0, wrapper);
    }
    return () => {
      const index = this._beforeTransformFunctions.findIndex(
        (item) => item.func === transformFunction
      );
      if (index === -1) return;
      this._beforeTransformFunctions.splice(index, 1);
    };
  }
  intercept(type, transformFunction, priority = 0) {
    const functions = this._transformFunctions.get(type);
    if (type === "text" && functions?.length) {
      throw new import_error2.ChatLunaError(
        import_error2.ChatLunaErrorCode.UNKNOWN_ERROR,
        new Error("text transform function already exists")
      );
    }
    const wrapper = {
      func: transformFunction,
      priority
    };
    if (!functions) {
      this._transformFunctions.set(type, [wrapper]);
    } else {
      const insertIndex = functions.findIndex(
        (item) => item.priority > priority
      );
      if (insertIndex === -1) {
        functions.push(wrapper);
      } else {
        functions.splice(insertIndex, 0, wrapper);
      }
    }
    return () => {
      const currentFunctions = this._transformFunctions.get(type);
      if (!currentFunctions) return;
      const index = currentFunctions.findIndex(
        (item) => item.func === transformFunction
      );
      if (index === -1) return;
      if (currentFunctions.length === 1) {
        this._transformFunctions.delete(type);
      } else {
        currentFunctions.splice(index, 1);
      }
    };
  }
  replace(type, transformFunction) {
    if (type === "text") {
      throw new import_error2.ChatLunaError(
        import_error2.ChatLunaErrorCode.UNKNOWN_ERROR,
        new Error("text transform function cannot be replaced")
      );
    }
    const functions = this._transformFunctions.get(type);
    if (functions == null || functions.length === 0) {
      import_koishi_plugin_chatluna.logger?.warn(
        `transform function for ${type} not exists. Check your installed plugins.`
      );
    }
    this._transformFunctions.set(type, [
      { func: transformFunction, priority: 0 }
    ]);
    return () => {
      this._transformFunctions.delete(type);
    };
  }
  has(type) {
    const functions = this._transformFunctions.get(type);
    return functions != null && functions.length > 0;
  }
  async _processElement(session, element, message, model) {
    const transformFunctions = this._transformFunctions.get(element.type);
    if (!transformFunctions?.length) {
      if (element.children?.length) {
        await this.transform(
          session,
          element.children,
          model,
          message,
          {
            quote: false,
            includeQuoteReply: true
          }
        );
      }
      return;
    }
    const hasChildren = !!element.children?.length;
    for (const { func: transformFunction } of transformFunctions) {
      const result = await transformFunction(
        session,
        element,
        message,
        model
      );
      if (result !== false) return;
      if (hasChildren) {
        await this.transform(
          session,
          element.children,
          model,
          message,
          {
            quote: false,
            includeQuoteReply: true
          }
        );
        return;
      }
    }
    if (hasChildren) {
      await this.transform(session, element.children, model, message, {
        quote: false,
        includeQuoteReply: false
      });
    }
  }
  async _runBeforeTransform(session, elements, message, model, options) {
    for (const { func: transformFunction } of this._beforeTransformFunctions) {
      await transformFunction(session, elements, message, model, options);
    }
  }
};

// src/services/conversation.ts
var import_crypto2 = require("crypto");
var import_promises2 = __toESM(require("fs/promises"), 1);
var import_path2 = __toESM(require("path"), 1);
var import_types2 = require("koishi-plugin-chatluna/llm-core/platform/types");
var import_count_tokens = require("koishi-plugin-chatluna/llm-core/utils/count_tokens");

// src/utils/archive.ts
var import_crypto = require("crypto");
var import_promises = __toESM(require("fs/promises"), 1);
var import_path = __toESM(require("path"), 1);

// src/utils/compression.ts
var import_zlib = require("zlib");
var import_util = require("util");
var gzipAsync = (0, import_util.promisify)(import_zlib.gzip);
var gunzipAsync = (0, import_util.promisify)(import_zlib.gunzip);
async function gzipEncode(data, encoding = "buffer") {
  const result = await gzipAsync(data);
  return encoding === "buffer" ? result : result.toString(encoding);
}
__name(gzipEncode, "gzipEncode");
async function gzipDecode(data, encoding = "base64") {
  const buffer = typeof data === "string" ? Buffer.from(data, encoding) : ArrayBuffer.isView(data) ? Buffer.from(data.buffer, data.byteOffset, data.byteLength) : Buffer.from(data);
  return (await gunzipAsync(buffer)).toString();
}
__name(gzipDecode, "gzipDecode");
function bufferToArrayBuffer(buffer) {
  return buffer.buffer.slice(
    buffer.byteOffset,
    buffer.byteOffset + buffer.byteLength
  );
}
__name(bufferToArrayBuffer, "bufferToArrayBuffer");

// src/utils/archive.ts
async function removeArchive(ctx, archiveId) {
  if (archiveId == null) {
    return;
  }
  const archive = await ctx.chatluna.conversation.getArchive(archiveId);
  if (archive?.path) {
    await import_promises.default.rm(archive.path, {
      recursive: true,
      force: true
    });
  }
  await ctx.database.remove("chatluna_archive", {
    id: archiveId
  });
}
__name(removeArchive, "removeArchive");
async function unbindConversation(ctx, conversationId) {
  const [active, last] = await Promise.all([
    ctx.database.get("chatluna_binding", {
      activeConversationId: conversationId
    }),
    ctx.database.get("chatluna_binding", {
      lastConversationId: conversationId
    })
  ]);
  const bindings = Array.from(
    new Map(
      [...active, ...last].map(
        (item) => [item.bindingKey, item]
      )
    ).values()
  );
  for (const binding of bindings) {
    await ctx.database.upsert("chatluna_binding", [
      {
        bindingKey: binding.bindingKey,
        activeConversationId: binding.activeConversationId === conversationId ? null : binding.activeConversationId,
        lastConversationId: binding.lastConversationId === conversationId ? null : binding.lastConversationId,
        updatedAt: /* @__PURE__ */ new Date()
      }
    ]);
  }
}
__name(unbindConversation, "unbindConversation");
async function readArchivePayload(archivePath) {
  const stat = await import_promises.default.stat(archivePath);
  if (stat.isDirectory()) {
    const manifest = JSON.parse(
      await import_promises.default.readFile(import_path.default.join(archivePath, "manifest.json"), "utf8")
    );
    const conversation = JSON.parse(
      await import_promises.default.readFile(
        import_path.default.join(archivePath, "conversation.json"),
        "utf8"
      )
    );
    const messageBuffer = await import_promises.default.readFile(
      import_path.default.join(archivePath, "messages.jsonl.gz")
    );
    if (manifest.size !== messageBuffer.byteLength) {
      throw new Error("Archive payload size mismatch.");
    }
    if (manifest.checksum != null && manifest.checksum.length > 0) {
      const checksum = (0, import_crypto.createHash)("sha256").update(messageBuffer).digest("hex");
      if (checksum !== manifest.checksum) {
        throw new Error("Archive payload checksum mismatch.");
      }
    }
    const messages = (await gzipDecode(messageBuffer)).split("\n").filter((line) => line.length > 0).map((line) => JSON.parse(line));
    return {
      formatVersion: manifest.formatVersion,
      exportedAt: manifest.createdAt,
      conversation,
      messages
    };
  }
  return JSON.parse(
    await gzipDecode(await import_promises.default.readFile(archivePath))
  );
}
__name(readArchivePayload, "readArchivePayload");
function serializeConversation(conversation) {
  return {
    ...conversation,
    createdAt: conversation.createdAt.toISOString(),
    updatedAt: conversation.updatedAt.toISOString(),
    lastChatAt: conversation.lastChatAt ? conversation.lastChatAt.toISOString() : null,
    archivedAt: conversation.archivedAt ? conversation.archivedAt.toISOString() : null
  };
}
__name(serializeConversation, "serializeConversation");
function deserializeConversation(conversation) {
  return {
    ...conversation,
    createdAt: new Date(conversation.createdAt),
    updatedAt: new Date(conversation.updatedAt),
    lastChatAt: conversation.lastChatAt ? new Date(conversation.lastChatAt) : null,
    archivedAt: conversation.archivedAt ? new Date(conversation.archivedAt) : null
  };
}
__name(deserializeConversation, "deserializeConversation");
function serializeMessage(message) {
  return {
    ...message,
    content: serializeBinary(message.content),
    additional_kwargs_binary: serializeBinary(
      message.additional_kwargs_binary
    ),
    response_metadata_binary: serializeBinary(
      message.response_metadata_binary
    ),
    createdAt: message.createdAt?.toISOString() ?? null
  };
}
__name(serializeMessage, "serializeMessage");
function deserializeMessage(message) {
  return {
    ...message,
    content: deserializeBinary(message.content),
    additional_kwargs_binary: deserializeBinary(
      message.additional_kwargs_binary
    ),
    response_metadata_binary: deserializeBinary(
      message.response_metadata_binary
    ),
    createdAt: message.createdAt ? new Date(message.createdAt) : null
  };
}
__name(deserializeMessage, "deserializeMessage");
function serializeBinary(value) {
  if (value == null) {
    return null;
  }
  return Buffer.from(value).toString("base64");
}
__name(serializeBinary, "serializeBinary");
function deserializeBinary(value) {
  if (value == null || value.length === 0) {
    return null;
  }
  return bufferToArrayBuffer(Buffer.from(value, "base64"));
}
__name(deserializeBinary, "deserializeBinary");

// src/types.ts
var ConversationResolutionError = class extends Error {
  constructor(code) {
    super(
      code === "ambiguous_target" ? "Conversation target is ambiguous." : "Conversation does not belong to current route."
    );
    this.code = code;
  }
  static {
    __name(this, "ConversationResolutionError");
  }
};
var ConversationNotFoundError = class extends Error {
  static {
    __name(this, "ConversationNotFoundError");
  }
  constructor() {
    super("Conversation not found.");
    this.name = "ConversationNotFoundError";
  }
};
var ConstraintLockedError = class extends Error {
  constructor(action) {
    super(`Conversation ${action} is locked by constraint.`);
    this.action = action;
    this.name = "ConstraintLockedError";
  }
  static {
    __name(this, "ConstraintLockedError");
  }
};
var ConstraintDisabledError = class extends Error {
  constructor(action) {
    super(`Conversation ${action} is disabled by constraint.`);
    this.action = action;
    this.name = "ConstraintDisabledError";
  }
  static {
    __name(this, "ConstraintDisabledError");
  }
};
var FIXED_FIELD_LABEL = {
  model: "Model",
  preset: "Preset",
  chatMode: "Chat mode"
};
var ConstraintFixedError = class extends Error {
  constructor(field, value) {
    super(`${FIXED_FIELD_LABEL[field]} is fixed to ${value}.`);
    this.field = field;
    this.value = value;
    this.name = "ConstraintFixedError";
  }
  static {
    __name(this, "ConstraintFixedError");
  }
};
var InvalidChatModeError = class extends Error {
  constructor(mode) {
    super(`Chat mode ${mode} not found.`);
    this.mode = mode;
    this.name = "InvalidChatModeError";
  }
  static {
    __name(this, "InvalidChatModeError");
  }
};
var AdminRequiredError = class extends Error {
  static {
    __name(this, "AdminRequiredError");
  }
  constructor() {
    super("Conversation management requires administrator permission.");
    this.name = "AdminRequiredError";
  }
};
function getBaseBindingKey(bindingKey) {
  const idx = bindingKey.indexOf(":preset:");
  return idx < 0 ? bindingKey : bindingKey.slice(0, idx);
}
__name(getBaseBindingKey, "getBaseBindingKey");
function getPresetLane(bindingKey) {
  const idx = bindingKey.indexOf(":preset:");
  return idx < 0 ? void 0 : bindingKey.slice(idx + 8);
}
__name(getPresetLane, "getPresetLane");
function computeBaseBindingKey(session, routeMode, routeKey) {
  const platform = session.platform;
  if (platform == null) {
    throw new Error("Session platform is missing.");
  }
  const selfId = session.selfId;
  if (selfId == null) {
    throw new Error("Session selfId is missing.");
  }
  if (routeMode === "custom") {
    if (routeKey == null || routeKey.length === 0) {
      throw new Error("Custom route key is missing.");
    }
    return `custom:${routeKey}`;
  }
  if (routeMode === "shared") {
    const guildId2 = session.guildId ?? session.channelId;
    if (guildId2 == null) {
      throw new Error("Shared conversation route requires guildId.");
    }
    return `shared:${platform}:${selfId}:${guildId2}`;
  }
  const userId = session.userId;
  if (userId == null) {
    throw new Error("Personal conversation route requires userId.");
  }
  if (session.isDirect) {
    return `personal:${platform}:${selfId}:direct:${userId}`;
  }
  const guildId = session.guildId ?? session.channelId;
  if (guildId == null) {
    throw new Error("Personal conversation route requires guildId.");
  }
  return `personal:${platform}:${selfId}:${guildId}:${userId}`;
}
__name(computeBaseBindingKey, "computeBaseBindingKey");
function applyPresetLane(bindingKey, presetLane) {
  if (presetLane == null || presetLane.length === 0) {
    return bindingKey;
  }
  return `${getBaseBindingKey(bindingKey)}:preset:${presetLane}`;
}
__name(applyPresetLane, "applyPresetLane");

// src/utils/conversation.ts
function getFallbackBindingKeys(session, bindingKey) {
  const idx = bindingKey.indexOf(":preset:");
  const suffix = idx >= 0 ? bindingKey.slice(idx) : "";
  if (bindingKey.startsWith("custom:")) {
    return [];
  }
  const guildOrChannel = session.guildId ?? session.channelId ?? "unknown";
  return session.isDirect ? [`personal:legacy:legacy:direct:${session.userId}${suffix}`] : bindingKey.startsWith("shared:") ? [
    `shared:legacy:legacy:${guildOrChannel}${suffix}`,
    `personal:legacy:legacy:${guildOrChannel}:${session.userId}${suffix}`
  ] : [
    `personal:legacy:legacy:${guildOrChannel}:${session.userId}${suffix}`,
    `shared:legacy:legacy:${guildOrChannel}${suffix}`
  ];
}
__name(getFallbackBindingKeys, "getFallbackBindingKeys");
function getLookupKeys(session, bindingKey, allPresetLanes = false) {
  const keys = /* @__PURE__ */ new Set();
  keys.add(allPresetLanes ? getBaseBindingKey(bindingKey) : bindingKey);
  for (const key of getFallbackBindingKeys(session, bindingKey)) {
    keys.add(allPresetLanes ? getBaseBindingKey(key) : key);
  }
  return Array.from(keys);
}
__name(getLookupKeys, "getLookupKeys");
function pickBindingKey(resolved, conversation) {
  return conversation.bindingKey === resolved.bindingKey ? resolved.bindingKey : conversation.bindingKey;
}
__name(pickBindingKey, "pickBindingKey");

// src/services/conversation.ts
var import_langchain = require("koishi-plugin-chatluna/utils/langchain");

// src/utils/message_content.ts
function getMessageContent(message) {
  if (typeof message === "string") {
    return message;
  }
  if (message == null) {
    return "";
  }
  const buffer = [];
  for (const part of message) {
    if (part.type === "text") {
      buffer.push(part.text);
    }
  }
  return buffer.join("");
}
__name(getMessageContent, "getMessageContent");

// src/services/conversation.ts
var import_koishi3 = require("koishi-plugin-chatluna/utils/koishi");
var import_lock = require("koishi-plugin-chatluna/utils/lock");
var import_error3 = require("koishi-plugin-chatluna/utils/error");
var EMPTY_MODEL_NAMES = /* @__PURE__ */ new Set(["", "无", "empty"]);
var FIXED_FIELDS = [
  { key: "model", constraintKey: "fixedModel", label: "model" },
  { key: "preset", constraintKey: "fixedPreset", label: "preset" },
  { key: "chatMode", constraintKey: "fixedChatMode", label: "chatMode" }
];
var ConversationService = class {
  constructor(ctx, config, runtime, platform) {
    this.ctx = ctx;
    this.config = config;
    this.runtime = runtime;
    this.platform = platform;
  }
  static {
    __name(this, "ConversationService");
  }
  _bindingLocks = /* @__PURE__ */ new Map();
  _titleLocks = /* @__PURE__ */ new Map();
  async getConversation(id) {
    return this.firstRow("chatluna_conversation", { id });
  }
  async getBinding(bindingKey) {
    return this.firstRow("chatluna_binding", { bindingKey });
  }
  async getArchive(id) {
    return this.firstRow("chatluna_archive", { id });
  }
  async getArchiveByConversationId(conversationId) {
    return this.firstRow("chatluna_archive", { conversationId });
  }
  async listConstraints() {
    return (await this.ctx.database.get(
      "chatluna_constraint",
      {}
    )).filter((c) => c.enabled !== false).sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));
  }
  async resolveConstraint(session, options = {}) {
    let constraints = (await this.listConstraints()).filter(
      (c) => isConstraintMatched(c, session)
    );
    const routed = constraints.find((c) => c.routeMode != null);
    let routeMode = routed?.routeMode ?? this.getDefaultRouteMode(session);
    let baseKey = computeBaseBindingKey(
      session,
      routeMode,
      routed?.routeKey
    );
    let bindingKey = this.normalizeBindingKey(
      baseKey,
      options.bindingKey,
      options.presetLane
    );
    if (options.bindingKey != null) {
      constraints = constraints.filter(
        (c) => !c.name.startsWith("managed:")
      );
      const managed = await this.getManagedConstraintByBindingKey(bindingKey);
      if (managed != null) {
        constraints.unshift(managed);
      }
      baseKey = getBaseBindingKey(bindingKey);
      routeMode = baseKey.startsWith("shared:") ? "shared" : baseKey.startsWith("personal:") ? "personal" : "custom";
    }
    const activePresetLane = firstDefined(constraints, "activePresetLane");
    const presetLane = options.presetLane ?? (options.useRoutePresetLane ? activePresetLane : void 0);
    if (options.bindingKey == null || !options.bindingKey.includes(":preset:")) {
      bindingKey = applyPresetLane(baseKey, presetLane);
    }
    const lane = getPresetLane(bindingKey);
    return {
      routeMode,
      baseKey,
      bindingKey,
      constraints,
      activePresetLane,
      defaultModel: firstDefined(constraints, "defaultModel") ?? this.config.defaultModel,
      defaultPreset: lane ?? firstDefined(constraints, "defaultPreset") ?? this.config.defaultPreset,
      defaultChatMode: firstDefined(constraints, "defaultChatMode") ?? this.config.defaultChatMode,
      fixedModel: firstDefined(constraints, "fixedModel"),
      fixedPreset: firstDefined(constraints, "fixedPreset"),
      fixedChatMode: firstDefined(constraints, "fixedChatMode"),
      lockConversation: firstBoolean(
        constraints,
        "lockConversation",
        false
      ),
      allowNew: firstBoolean(constraints, "allowNew", true),
      allowSwitch: firstBoolean(constraints, "allowSwitch", true),
      allowArchive: firstBoolean(constraints, "allowArchive", true),
      allowExport: firstBoolean(constraints, "allowExport", true),
      manageMode: firstDefined(constraints, "manageMode") ?? "admin"
    };
  }
  normalizeBindingKey(baseKey, explicit, presetLane) {
    if (explicit == null) {
      return applyPresetLane(baseKey, presetLane);
    }
    return explicit.includes(":preset:") ? explicit : applyPresetLane(explicit, presetLane);
  }
  async resolveConversationContext(session, options = {}) {
    const constraint = await this.resolveConstraint(session, options);
    const matched = options.bindingKey == null ? await this.resolveBindingForKey(
      session,
      constraint.bindingKey
    ) : {
      bindingKey: constraint.bindingKey,
      binding: await this.getBinding(constraint.bindingKey)
    };
    const binding = matched?.binding;
    const bindingKey = matched?.bindingKey ?? constraint.bindingKey;
    const conversation = options.conversationId ? await this.getConversation(options.conversationId) : binding?.activeConversationId ? await this.getConversation(binding.activeConversationId) : void 0;
    const allowed = conversation != null && await hasConversationPermission(
      this.ctx,
      session,
      conversation,
      "view",
      bindingKey
    ) ? conversation : null;
    return this.buildContext(
      constraint,
      bindingKey,
      binding ?? null,
      allowed
    );
  }
  async resolveConversation(session, options = {}) {
    const mode = options.mode ?? "context";
    const resolved = await this.resolveConversationContext(session, options);
    if (mode === "context") {
      return {
        ...resolved,
        mode,
        conversationId: resolved.conversation?.id ?? null
      };
    }
    if (mode === "active") {
      return this.resolveActiveMode(session, resolved, options);
    }
    return this.resolveTargetMode(session, resolved, options);
  }
  async resolveActiveMode(session, initial, options) {
    let current = initial;
    if (current.constraint.lockConversation && current.binding?.activeConversationId != null) {
      const conversation2 = await this.getConversation(
        current.binding.activeConversationId
      );
      if (conversation2 != null && conversation2.status !== "deleted" && conversation2.status !== "broken" && await hasConversationPermission(
        this.ctx,
        session,
        conversation2,
        "view",
        current.bindingKey
      )) {
        current = this.buildContext(
          current.constraint,
          current.bindingKey,
          current.binding,
          conversation2
        );
      }
    }
    if (current.conversation != null) {
      if (current.conversation.status !== "archived") {
        return {
          ...current,
          mode: "active",
          conversationId: current.conversation.id
        };
      }
      await assertManageAllowed(session, current.constraint);
      if (!current.constraint.allowArchive) {
        throw new ConstraintDisabledError("restore");
      }
      const restored = await this.restoreConversation(session, {
        conversationId: current.conversation.id
      });
      const refreshed = this.buildContext(
        current.constraint,
        current.bindingKey,
        current.binding,
        restored
      );
      return {
        ...refreshed,
        mode: "active",
        conversationId: restored.id
      };
    }
    if (!current.constraint.allowNew) {
      throw new ConstraintDisabledError("create");
    }
    const conversation = await this.createConversation(session, {
      bindingKey: current.bindingKey,
      preset: current.effectivePreset,
      model: current.effectiveModel,
      chatMode: current.effectiveChatMode,
      title: current.presetLane ?? "New Conversation"
    });
    return {
      ...current,
      mode: "active",
      conversation,
      conversationId: conversation.id
    };
  }
  async resolveTargetMode(session, resolved, options) {
    const mode = "target";
    const finalize = /* @__PURE__ */ __name((conversation) => ({
      ...resolved,
      mode,
      conversation,
      conversationId: conversation?.id ?? null
    }), "finalize");
    const promoteTarget = /* @__PURE__ */ __name(async (conversation) => {
      const target2 = await this.resolveConversationContext(session, {
        ...options,
        bindingKey: conversation.bindingKey,
        conversationId: conversation.id,
        presetLane: getPresetLane(conversation.bindingKey),
        useRoutePresetLane: false
      });
      return {
        ...target2,
        mode,
        conversation: target2.conversation ?? conversation,
        conversationId: conversation.id
      };
    }, "promoteTarget");
    if (options.conversationId != null) {
      const conversation = await this.getConversation(
        options.conversationId
      );
      if (conversation == null || conversation.status === "deleted" || conversation.status === "broken") {
        return finalize(null);
      }
      const inLookupKeys = options.allPresetLanes === true && getLookupKeys(
        session,
        resolved.constraint.bindingKey,
        true
      ).includes(getBaseBindingKey(conversation.bindingKey));
      if (!inLookupKeys && !await hasConversationPermission(
        this.ctx,
        session,
        conversation,
        options.permission ?? "view",
        resolved.bindingKey
      )) {
        throw new ConversationResolutionError("target_outside_route");
      }
      return promoteTarget(conversation);
    }
    if (options.targetConversation == null) {
      return {
        ...resolved,
        mode,
        conversationId: resolved.conversation?.id ?? null
      };
    }
    const target = options.targetConversation.trim();
    if (target.length === 0) {
      return finalize(null);
    }
    const entries = await this.listConversationEntries(session, {
      presetLane: options.presetLane,
      allPresetLanes: options.allPresetLanes,
      includeArchived: options.includeArchived
    });
    const normalized = target.toLocaleLowerCase();
    const local = matchTargetConversation(
      target,
      normalized,
      entries.map((item) => item.conversation),
      entries
    );
    if (local != null) {
      return promoteTarget(local);
    }
    const global = await this.findAccessibleConversations(session, {
      ...options,
      bindingKey: resolved.bindingKey,
      includeArchived: options.includeArchived,
      query: normalized,
      exactId: target
    });
    const remote = matchTargetConversation(target, normalized, global);
    if (remote != null) {
      return promoteTarget(remote);
    }
    return finalize(null);
  }
  buildContext(constraint, bindingKey, binding, conversation) {
    const effectiveModel = this.pickModel(constraint, conversation);
    const withModel = conversation != null && effectiveModel != null ? { ...conversation, model: effectiveModel } : conversation;
    return {
      bindingKey,
      presetLane: getPresetLane(bindingKey),
      binding,
      conversation: withModel,
      effectiveModel,
      effectivePreset: constraint.fixedPreset ?? conversation?.preset ?? getPresetLane(bindingKey) ?? constraint.defaultPreset ?? this.config.defaultPreset,
      effectiveChatMode: constraint.fixedChatMode ?? conversation?.chatMode ?? constraint.defaultChatMode ?? this.config.defaultChatMode,
      constraint
    };
  }
  async resolveBindingForKey(session, bindingKey) {
    const binding = await this.getBinding(bindingKey);
    if (binding != null) {
      return { bindingKey, binding };
    }
    for (const key of getFallbackBindingKeys(session, bindingKey)) {
      const legacyBinding = await this.getBinding(key);
      if (legacyBinding != null) {
        return { bindingKey: key, binding: legacyBinding };
      }
    }
    return null;
  }
  async ensureActiveConversation(session, options = {}) {
    const resolved = await this.resolveConversation(session, {
      ...options,
      mode: "active"
    });
    return resolved;
  }
  async createConversation(session, options) {
    this.checkChatMode(options.chatMode);
    return runLock(this._bindingLocks, options.bindingKey, async () => {
      const now = /* @__PURE__ */ new Date();
      const conversation = {
        id: (0, import_crypto2.randomUUID)(),
        seq: await this.allocateConversationSeq(options.bindingKey),
        bindingKey: options.bindingKey,
        title: options.title,
        model: options.model,
        preset: options.preset,
        chatMode: options.chatMode,
        createdBy: session.userId,
        createdAt: now,
        updatedAt: now,
        lastChatAt: now,
        status: "active",
        latestMessageId: null,
        additional_kwargs: null,
        compression: null,
        archivedAt: null,
        archiveId: null,
        legacyRoomId: null,
        legacyMeta: null,
        autoTitle: true
      };
      await this.ctx.root.parallel(
        "chatluna/before-conversation-create",
        {
          conversation,
          bindingKey: options.bindingKey
        }
      );
      await this.ctx.database.create(
        "chatluna_conversation",
        conversation
      );
      if (options.setActive !== false) {
        await this.setActiveConversation(
          options.bindingKey,
          conversation.id
        );
      }
      await this.ctx.root.parallel("chatluna/after-conversation-create", {
        conversation,
        bindingKey: options.bindingKey
      });
      return conversation;
    });
  }
  async setActiveConversation(bindingKey, conversationId) {
    const current = await this.getBinding(bindingKey);
    const prev = current?.activeConversationId;
    const payload = {
      bindingKey,
      activeConversationId: conversationId,
      lastConversationId: prev != null && prev !== conversationId ? prev : current?.lastConversationId ?? null,
      updatedAt: /* @__PURE__ */ new Date()
    };
    await this.ctx.database.upsert("chatluna_binding", [payload]);
    await this.ctx.root.parallel("chatluna/after-binding-update", {
      binding: payload,
      previousConversationId: prev ?? null
    });
    return payload;
  }
  async touchConversation(conversationId, patch = {}) {
    const update = {
      updatedAt: patch.updatedAt ?? /* @__PURE__ */ new Date()
    };
    for (const key in patch) {
      const value = patch[key];
      if (value !== void 0) {
        update[key] = value;
      }
    }
    await this.ctx.database.set(
      "chatluna_conversation",
      { id: conversationId },
      update
    );
    return this.getConversation(conversationId);
  }
  async claimAutoTitle(conversationId) {
    return runLock(this._titleLocks, conversationId, async () => {
      const conversation = await this.getConversation(conversationId);
      if (conversation == null || !conversation.autoTitle) {
        return false;
      }
      await this.touchConversation(conversationId, { autoTitle: false });
      return true;
    });
  }
  async listConversations(session, options = {}) {
    const resolved = await this.resolveConversation(session, {
      ...options,
      mode: "context"
    });
    const keys = getLookupKeys(
      session,
      resolved.constraint.bindingKey,
      options.allPresetLanes
    );
    const all = await this.ctx.database.get(
      "chatluna_conversation",
      options.allPresetLanes ? {} : { bindingKey: keys.length === 1 ? keys[0] : { $in: keys } }
    );
    const conversations = options.allPresetLanes ? all.filter(
      (c) => keys.some(
        (key) => c.bindingKey === key || c.bindingKey.startsWith(key + ":preset:")
      )
    ) : all;
    const filtered = conversations.filter(
      (c) => c.status !== "deleted" && c.status !== "broken" && (options.includeArchived || c.status !== "archived")
    );
    const merged = new Set(filtered.map((c) => c.bindingKey)).size > 1;
    return filtered.sort((a, b) => {
      const seq = (a.seq ?? 0) - (b.seq ?? 0);
      if (seq !== 0) return seq;
      if (merged) {
        const key = a.bindingKey.localeCompare(b.bindingKey);
        if (key !== 0) return key;
      }
      const created = a.createdAt.getTime() - b.createdAt.getTime();
      if (created !== 0) return created;
      return a.id.localeCompare(b.id);
    });
  }
  async listConversationEntries(session, options = {}) {
    const conversations = await this.listConversations(session, options);
    const merged = new Set(conversations.map((c) => c.bindingKey)).size > 1;
    return conversations.map((conversation, idx) => ({
      conversation,
      displaySeq: merged ? idx + 1 : conversation.seq ?? 0
    }));
  }
  async getTarget(session, options, permission, includeArchived = false) {
    const resolved = await this.resolveConversation(session, {
      ...options,
      includeArchived,
      permission,
      mode: "target"
    });
    await assertManageAllowed(session, resolved.constraint);
    const conversation = resolved.conversation;
    if (conversation == null) {
      throw new ConversationNotFoundError();
    }
    const managed = await this.getManagedConstraintByBindingKey(
      conversation.bindingKey
    );
    if (managed != null) {
      await assertManageAllowed(session, managed);
    }
    return { resolved, conversation, managed };
  }
  async switchConversation(session, options) {
    const { resolved, conversation, managed } = await this.getTarget(
      session,
      options,
      "manage"
    );
    const current = options.allPresetLanes ? await this.resolveConversation(session, {
      useRoutePresetLane: true,
      mode: "context"
    }) : resolved;
    assertActionAllowed("switch", resolved, managed, {
      needsAllow: "switch"
    });
    const previousConversation = current.binding?.activeConversationId ? await this.getConversation(current.binding.activeConversationId) : null;
    const sameRoute = options.allPresetLanes === true && getLookupKeys(
      session,
      resolved.constraint.bindingKey,
      true
    ).includes(getBaseBindingKey(conversation.bindingKey));
    const bindingKey = pickBindingKey(resolved, conversation);
    if (sameRoute) {
      await this.updateManagedConstraint(session, {
        activePresetLane: getPresetLane(conversation.bindingKey) ?? null
      });
    }
    await this.ctx.root.parallel("chatluna/before-conversation-switch", {
      bindingKey,
      conversation,
      previousConversation
    });
    await this.setActiveConversation(bindingKey, conversation.id);
    if (current.bindingKey !== bindingKey) {
      await this.setActiveConversation(
        current.bindingKey,
        conversation.id
      );
    }
    await this.ctx.root.parallel("chatluna/after-conversation-switch", {
      bindingKey,
      conversation,
      previousConversation
    });
    return conversation;
  }
  async reopenConversation(session, options) {
    const { resolved, conversation, managed } = await this.getTarget(
      session,
      options,
      "manage",
      true
    );
    if (managed?.lockConversation ?? resolved.constraint.lockConversation) {
      throw new ConstraintLockedError("restore");
    }
    if (conversation.status !== "archived") {
      if (!(managed?.allowSwitch ?? resolved.constraint.allowSwitch)) {
        throw new ConstraintDisabledError("switch");
      }
      if (options.allPresetLanes && getLookupKeys(
        session,
        resolved.constraint.bindingKey,
        true
      ).includes(getBaseBindingKey(conversation.bindingKey))) {
        await this.updateManagedConstraint(session, {
          activePresetLane: getPresetLane(conversation.bindingKey) ?? null
        });
      }
      await this.setActiveConversation(
        pickBindingKey(resolved, conversation),
        conversation.id
      );
      return conversation;
    }
    return this.restoreConversation(session, {
      ...options,
      conversationId: conversation.id
    });
  }
  async listMessages(conversationId) {
    const [conversation, messages] = await Promise.all([
      this.getConversation(conversationId),
      this.ctx.database.get("chatluna_message", { conversationId })
    ]);
    const records = messages;
    if (records.length < 2) {
      return records;
    }
    const sortByCreatedAt = /* @__PURE__ */ __name(() => [...records].sort(
      (a, b) => (a.createdAt?.getTime() ?? 0) - (b.createdAt?.getTime() ?? 0)
    ), "sortByCreatedAt");
    if (conversation?.latestMessageId == null) {
      return sortByCreatedAt();
    }
    const map = new Map(records.map((m) => [m.id, m]));
    const ordered = [];
    const seen = /* @__PURE__ */ new Set();
    let currentId = conversation.latestMessageId;
    while (currentId != null && !seen.has(currentId)) {
      const message = map.get(currentId);
      if (message == null) break;
      ordered.unshift(message);
      seen.add(currentId);
      currentId = message.parentId;
    }
    return ordered.length === records.length ? ordered : sortByCreatedAt();
  }
  async listAcl(conversationId) {
    return await this.ctx.database.get("chatluna_acl", {
      conversationId
    });
  }
  async upsertAcl(conversationId, records) {
    if (records.length === 0) {
      return [];
    }
    await this.ctx.database.upsert(
      "chatluna_acl",
      records.map((record) => ({ conversationId, ...record }))
    );
    return this.listAcl(conversationId);
  }
  async replaceAcl(conversationId, records) {
    await this.ctx.database.remove("chatluna_acl", { conversationId });
    return this.upsertAcl(conversationId, records);
  }
  async removeAcl(conversationId, records) {
    if (records == null || records.length === 0) {
      await this.ctx.database.remove("chatluna_acl", { conversationId });
      return [];
    }
    const current = await this.listAcl(conversationId);
    const matches = /* @__PURE__ */ __name((item) => records.some(
      (record) => (record.principalType == null || record.principalType === item.principalType) && (record.principalId == null || record.principalId === item.principalId) && (record.permission == null || record.permission === item.permission)
    ), "matches");
    for (const item of current.filter(matches)) {
      await this.ctx.database.remove("chatluna_acl", item);
    }
    return this.listAcl(conversationId);
  }
  async exportConversation(session, options = {}) {
    const { resolved, conversation, managed } = await this.getTarget(
      session,
      options,
      "view",
      true
    );
    if (!(managed?.allowExport ?? resolved.constraint.allowExport)) {
      throw new ConstraintDisabledError("export");
    }
    const markdown = await this.exportMarkdown(conversation);
    const outputPath = options.outputPath ?? import_path2.default.join(
      await this.ensureDataDir("export"),
      `${conversation.id}-${Date.now()}.md`
    );
    await import_promises2.default.writeFile(outputPath, markdown, "utf8");
    return {
      conversation,
      path: outputPath,
      size: Buffer.byteLength(markdown)
    };
  }
  async archiveConversation(session, options = {}) {
    const { conversation, managed, resolved } = await this.getTarget(
      session,
      options,
      "manage"
    );
    if (managed?.lockConversation ?? resolved.constraint.lockConversation) {
      throw new ConstraintLockedError("archive");
    }
    if (!(managed?.allowArchive ?? resolved.constraint.allowArchive)) {
      throw new ConstraintDisabledError("archive");
    }
    return this.archiveConversationById(conversation.id);
  }
  async archiveConversationById(conversationId, inactiveBefore) {
    const conversation = await this.getConversation(conversationId);
    if (conversation == null) {
      throw new ConversationNotFoundError();
    }
    return this.runtime.withConversationSync(conversation, async () => {
      const current = await this.getConversation(conversationId);
      if (current == null) {
        throw new ConversationNotFoundError();
      }
      if (inactiveBefore != null && (current.status !== "active" || current.updatedAt.getTime() >= inactiveBefore.getTime())) {
        return null;
      }
      if (current.status === "archived" && current.archiveId != null) {
        const archive2 = await this.getArchive(current.archiveId);
        if (archive2 != null) {
          return {
            conversation: current,
            archive: archive2,
            path: archive2.path
          };
        }
      }
      await this.ctx.root.parallel(
        "chatluna/before-conversation-archive",
        { conversation: current }
      );
      const archiveDir = await this.ensureDataDir(
        import_path2.default.join("archive", current.id)
      );
      const messages = await this.listMessages(current.id);
      const payload = {
        formatVersion: 1,
        exportedAt: (/* @__PURE__ */ new Date()).toISOString(),
        conversation: serializeConversation(current),
        messages: messages.map(serializeMessage)
      };
      const messageBuffer = await gzipEncode(
        payload.messages.map((message) => JSON.stringify(message)).join("\n")
      );
      const checksum = (0, import_crypto2.createHash)("sha256").update(messageBuffer).digest("hex");
      const now = /* @__PURE__ */ new Date();
      const manifest = {
        format: "chatluna-archive",
        formatVersion: payload.formatVersion,
        conversationId: current.id,
        messageCount: payload.messages.length,
        checksum,
        size: messageBuffer.byteLength,
        createdAt: now.toISOString()
      };
      await import_promises2.default.writeFile(
        import_path2.default.join(archiveDir, "conversation.json"),
        JSON.stringify(payload.conversation, null, 2),
        "utf8"
      );
      await import_promises2.default.writeFile(
        import_path2.default.join(archiveDir, "messages.jsonl.gz"),
        messageBuffer
      );
      await import_promises2.default.writeFile(
        import_path2.default.join(archiveDir, "manifest.json"),
        JSON.stringify(manifest, null, 2),
        "utf8"
      );
      const archive = {
        id: (0, import_crypto2.randomUUID)(),
        conversationId: manifest.conversationId,
        path: archiveDir,
        formatVersion: manifest.formatVersion,
        messageCount: manifest.messageCount,
        checksum: manifest.checksum,
        size: manifest.size,
        state: "ready",
        createdAt: now,
        restoredAt: null
      };
      await this.ctx.database.upsert("chatluna_archive", [archive]);
      await this.touchConversation(current.id, {
        status: "archived",
        archivedAt: now,
        archiveId: archive.id
      });
      await unbindConversation(this.ctx, current.id);
      await this.ctx.database.remove("chatluna_message", {
        conversationId: current.id
      });
      const updated = await this.getConversation(current.id) ?? current;
      await this.runtime.clearConversationInterfaceLocked(updated);
      await this.ctx.root.parallel(
        "chatluna/after-conversation-archive",
        { conversation: updated, archive, path: archiveDir }
      );
      return { conversation: updated, archive, path: archiveDir };
    });
  }
  async restoreConversation(session, options = {}) {
    const { resolved, conversation, managed } = await this.getTarget(
      session,
      options,
      "manage",
      true
    );
    const archive = options.archiveId ? await this.getArchive(options.archiveId) : conversation.archiveId ? await this.getArchive(conversation.archiveId) : await this.getArchiveByConversationId(conversation.id);
    if (archive == null) {
      throw new import_error3.ChatLunaError(
        import_error3.ChatLunaErrorCode.CONVERSATION_ARCHIVE_NOT_FOUND
      );
    }
    if (archive.conversationId !== conversation.id) {
      throw new import_error3.ChatLunaError(
        import_error3.ChatLunaErrorCode.CONVERSATION_ARCHIVE_MISMATCH
      );
    }
    if (managed?.lockConversation ?? resolved.constraint.lockConversation) {
      throw new ConstraintLockedError("restore");
    }
    if (!(managed?.allowArchive ?? resolved.constraint.allowArchive)) {
      throw new ConstraintDisabledError("restore");
    }
    return this.runtime.withConversationSync(conversation, async () => {
      const current = await this.getConversation(conversation.id);
      if (current == null) {
        throw new ConversationNotFoundError();
      }
      await this.ctx.root.parallel(
        "chatluna/before-conversation-restore",
        { conversation: current, archive }
      );
      await this.ctx.database.upsert("chatluna_archive", [
        { ...archive, state: "restoring" }
      ]);
      try {
        const payload = await readArchivePayload(archive.path);
        const restored = deserializeConversation(payload.conversation);
        const restoredMessages = payload.messages.map((message) => ({
          ...deserializeMessage(message),
          conversationId: current.id
        }));
        await this.ctx.database.remove("chatluna_message", {
          conversationId: current.id
        });
        if (restoredMessages.length > 0) {
          await this.ctx.database.upsert(
            "chatluna_message",
            restoredMessages
          );
        }
        await this.ctx.database.upsert("chatluna_conversation", [
          {
            ...current,
            ...restored,
            id: current.id,
            status: "active",
            archivedAt: null,
            archiveId: null,
            updatedAt: /* @__PURE__ */ new Date()
          }
        ]);
        await this.ctx.database.upsert("chatluna_archive", [
          { ...archive, state: "ready", restoredAt: /* @__PURE__ */ new Date() }
        ]);
        const updated = await this.getConversation(current.id);
        if (updated == null) {
          throw new import_error3.ChatLunaError(
            import_error3.ChatLunaErrorCode.CONVERSATION_RESTORE_FAILED
          );
        }
        if (options.allPresetLanes && getLookupKeys(
          session,
          resolved.constraint.bindingKey,
          true
        ).includes(getBaseBindingKey(updated.bindingKey))) {
          await this.updateManagedConstraint(session, {
            activePresetLane: getPresetLane(updated.bindingKey) ?? null
          });
        }
        await this.setActiveConversation(updated.bindingKey, updated.id);
        await this.runtime.clearConversationInterfaceLocked(updated);
        await this.ctx.root.parallel(
          "chatluna/after-conversation-restore",
          { conversation: updated, archive }
        );
        return updated;
      } catch (error) {
        await this.ctx.database.upsert("chatluna_archive", [
          { ...archive, state: "broken" }
        ]);
        throw error;
      }
    });
  }
  async exportMarkdown(conversation) {
    const messages = await this.loadMessagesForExport(conversation);
    return [
      `# ${conversation.title}`,
      "",
      `- ID: ${conversation.id}`,
      `- Seq: ${conversation.seq ?? "-"}`,
      `- Route: ${conversation.bindingKey}`,
      `- Model: ${conversation.model}`,
      `- Preset: ${conversation.preset}`,
      `- Chat Mode: ${conversation.chatMode}`,
      `- Status: ${conversation.status}`,
      `- Updated At: ${conversation.updatedAt.toISOString()}`,
      "",
      ...(await Promise.all(
        messages.map(async (message) => [
          `## ${message.role} ${message.name ? `(${message.name})` : ""}`.trim(),
          "",
          await formatMessage(message),
          ""
        ])
      )).flat()
    ].join("\n");
  }
  async loadMessagesForExport(conversation) {
    if (conversation.status !== "archived" || conversation.archiveId == null) {
      return this.listMessages(conversation.id);
    }
    const archive = await this.getArchive(conversation.archiveId);
    if (archive == null) {
      return this.listMessages(conversation.id);
    }
    const payload = await readArchivePayload(archive.path);
    return payload.messages.map((message) => ({
      ...deserializeMessage(message),
      conversationId: conversation.id
    }));
  }
  async renameConversation(session, options) {
    const { resolved, conversation, managed } = await this.getTarget(
      session,
      options,
      "manage"
    );
    if (managed?.lockConversation ?? resolved.constraint.lockConversation) {
      throw new ConstraintLockedError("rename");
    }
    const updated = await this.touchConversation(conversation.id, {
      title: options.title.trim(),
      autoTitle: false
    });
    return updated;
  }
  async deleteConversation(session, options = {}) {
    const { resolved, conversation, managed } = await this.getTarget(
      session,
      options,
      "manage",
      true
    );
    if (managed?.lockConversation ?? resolved.constraint.lockConversation) {
      throw new ConstraintLockedError("delete");
    }
    return this.runtime.withConversationSync(conversation, async () => {
      const current = await this.getConversation(conversation.id);
      if (current == null) {
        throw new ConversationNotFoundError();
      }
      await this.ctx.root.parallel(
        "chatluna/before-conversation-delete",
        { conversation: current }
      );
      await removeArchive(this.ctx, current.archiveId);
      const updated = await this.touchConversation(current.id, {
        status: "deleted",
        archivedAt: null,
        archiveId: null
      });
      await unbindConversation(this.ctx, current.id);
      await this.ctx.database.remove("chatluna_message", {
        conversationId: current.id
      });
      await this.removeAcl(current.id);
      await this.runtime.clearConversationInterfaceLocked(
        updated ?? current
      );
      await this.ctx.root.parallel("chatluna/after-conversation-delete", {
        conversation: updated ?? current
      });
      return updated ?? current;
    });
  }
  async updateConversationUsage(session, options) {
    const lookupMode = options.conversationId == null ? "active" : "target";
    const lookupOptions = {
      ...options,
      ...lookupMode === "target" ? { permission: "manage" } : {},
      mode: lookupMode
    };
    const resolved = await this.resolveConversation(session, lookupOptions);
    await assertManageAllowed(session, resolved.constraint);
    const conversation = resolved.conversation;
    if (conversation == null) {
      throw new ConversationNotFoundError();
    }
    const target = await this.getManagedConstraintByBindingKey(
      conversation.bindingKey
    );
    if (target != null) {
      await assertManageAllowed(session, target);
    }
    for (const { key, constraintKey, label } of FIXED_FIELDS) {
      const fixed = target?.[constraintKey] ?? resolved.constraint[constraintKey];
      if (options[key] != null && fixed != null) {
        throw new ConstraintFixedError(label, fixed);
      }
    }
    if (target?.lockConversation ?? resolved.constraint.lockConversation) {
      throw new ConstraintLockedError("update");
    }
    this.checkChatMode(options.chatMode);
    const updated = await this.touchConversation(conversation.id, {
      model: options.model,
      preset: options.preset,
      chatMode: options.chatMode
    });
    if (updated == null) {
      throw new ConversationNotFoundError();
    }
    await this.runtime.clearConversationInterface(updated);
    return updated;
  }
  async recordCompression(conversationId, result) {
    const conversation = await this.getConversation(conversationId);
    if (!result.compressed) {
      return conversation;
    }
    if (conversation == null) {
      return void 0;
    }
    const current = JSON.parse(
      conversation.compression ?? "null"
    );
    const summaryMessage = (await this.ctx.database.get(
      "chatluna_message",
      { conversationId, name: "infinite_context" },
      { limit: 1, sort: { createdAt: "desc" } }
    ))[0];
    const summary = summaryMessage == null ? void 0 : await readText(summaryMessage);
    const updated = await this.touchConversation(conversationId, {
      compression: JSON.stringify({
        ...current,
        count: (current?.count ?? 0) + 1,
        compressedAt: (/* @__PURE__ */ new Date()).toISOString(),
        summary: summary ?? current?.summary,
        originalMessageCount: result.originalMessageCount,
        remainingMessageCount: result.remainingMessageCount,
        tokenUsage: result.outputTokens,
        inputTokens: result.inputTokens,
        outputTokens: result.outputTokens,
        reducedTokens: result.reducedTokens,
        reducedPercent: result.reducedPercent
      })
    });
    if (updated != null) {
      await this.ctx.root.parallel("chatluna/conversation-compressed", {
        conversation: updated,
        result
      });
    }
    return updated;
  }
  async getManagedConstraint(session) {
    const name = buildManagedConstraintName(session);
    return this.firstRow("chatluna_constraint", { name });
  }
  async getManagedConstraintByBindingKey(bindingKey) {
    const name = managedNameFromBindingKey(bindingKey);
    if (name == null) {
      return void 0;
    }
    return this.firstRow("chatluna_constraint", { name });
  }
  async updateManagedConstraint(session, patch) {
    this.checkChatMode(patch.defaultChatMode);
    this.checkChatMode(patch.fixedChatMode);
    const current = await this.getManagedConstraint(session);
    const now = /* @__PURE__ */ new Date();
    const guildId = session.isDirect ? null : session.guildId ?? session.channelId ?? null;
    const record = {
      id: current?.id,
      name: buildManagedConstraintName(session),
      enabled: true,
      priority: 1e3,
      createdBy: session.userId,
      createdAt: now,
      platform: session.platform,
      selfId: session.selfId,
      guildId,
      channelId: null,
      direct: session.isDirect,
      users: session.isDirect ? JSON.stringify([session.userId]) : null,
      excludeUsers: null,
      routeMode: null,
      routeKey: null,
      activePresetLane: null,
      defaultModel: null,
      defaultPreset: null,
      defaultChatMode: null,
      fixedModel: null,
      fixedPreset: null,
      fixedChatMode: null,
      lockConversation: null,
      allowNew: null,
      allowSwitch: null,
      allowArchive: null,
      allowExport: null,
      manageMode: "admin",
      ...current,
      ...patch,
      updatedAt: now
    };
    await this.ctx.database.upsert("chatluna_constraint", [record]);
    await this.ctx.root.parallel("chatluna/after-constraint-update", {
      constraint: record
    });
    return await this.getManagedConstraint(session) ?? record;
  }
  pickModel(constraint, conversation) {
    const candidates = [
      constraint.fixedModel,
      conversation?.model,
      constraint.defaultModel,
      this.config.defaultModel
    ];
    for (const model of candidates) {
      if (model == null) continue;
      const trimmed = model.trim();
      if (EMPTY_MODEL_NAMES.has(trimmed)) continue;
      const [platform, name] = (0, import_count_tokens.parseRawModelName)(model);
      if (platform == null || name == null) continue;
      const models = this.platform.listPlatformModels(
        platform,
        import_types2.ModelType.llm
      ).value;
      if (models.length > 0 && models.some((m) => m.name === name)) {
        return model;
      }
    }
    return null;
  }
  async firstRow(table, query) {
    if (table === "chatluna_conversation") {
      return (await this.ctx.database.get(
        "chatluna_conversation",
        query
      ))[0];
    }
    if (table === "chatluna_binding") {
      return (await this.ctx.database.get(
        "chatluna_binding",
        query
      ))[0];
    }
    if (table === "chatluna_archive") {
      return (await this.ctx.database.get(
        "chatluna_archive",
        query
      ))[0];
    }
    return (await this.ctx.database.get(
      "chatluna_constraint",
      query
    ))[0];
  }
  getDefaultRouteMode(session) {
    if (session.isDirect) {
      return "personal";
    }
    return this.config.defaultGroupRouteMode ?? "shared";
  }
  checkChatMode(mode) {
    if (mode != null && !this.platform.chatChains.value.some((chain) => chain.name === mode)) {
      throw new InvalidChatModeError(mode);
    }
  }
  async allocateConversationSeq(bindingKey) {
    const [latest] = await this.ctx.database.get(
      "chatluna_conversation",
      { bindingKey },
      { sort: { seq: "desc" }, limit: 1 }
    );
    return (latest?.seq ?? 0) + 1;
  }
  async findAccessibleConversations(session, options) {
    const matched = /* @__PURE__ */ __name((conversation) => {
      if (conversation.bindingKey === options.bindingKey || conversation.status === "deleted" || conversation.status === "broken" || !options.includeArchived && conversation.status === "archived") {
        return false;
      }
      const title = conversation.title.toLocaleLowerCase();
      return conversation.id === options.exactId || options.seq != null && conversation.seq === options.seq || title === options.query || title.includes(options.query);
    }, "matched");
    if (await (0, import_koishi3.checkAdmin)(session)) {
      if (options.exactId.length > 0) {
        const exact = await this.getConversation(options.exactId);
        if (exact != null && matched(exact)) {
          return [exact];
        }
      }
      const filter = options.seq != null ? { seq: options.seq } : {};
      const conversations = await this.ctx.database.get(
        "chatluna_conversation",
        filter
      );
      return conversations.filter(matched);
    }
    const ids = await this.collectAclConversationIds(
      session,
      options.permission ?? "view"
    );
    const matches = [];
    for (let i = 0; i < ids.length; i += 200) {
      const slice = ids.slice(i, i + 200);
      const conversations = await this.ctx.database.get(
        "chatluna_conversation",
        { id: { $in: slice } }
      );
      matches.push(...conversations.filter(matched));
    }
    return matches;
  }
  async collectAclConversationIds(session, required) {
    const acl = [];
    const fetch = /* @__PURE__ */ __name(async (principalType, principalId, permission) => {
      acl.push(
        ...await this.ctx.database.get("chatluna_acl", {
          principalType,
          principalId,
          permission
        })
      );
    }, "fetch");
    await fetch("user", session.userId, "manage");
    if (required === "view") {
      await fetch("user", session.userId, "view");
    }
    const guildId = session.guildId ?? session.channelId;
    if (guildId != null) {
      await fetch("guild", guildId, "manage");
      if (required === "view") {
        await fetch("guild", guildId, "view");
      }
    }
    return Array.from(new Set(acl.map((item) => item.conversationId)));
  }
  async ensureDataDir(name) {
    const target = import_path2.default.resolve(this.ctx.baseDir, "data/chatluna", name);
    await import_promises2.default.mkdir(target, { recursive: true });
    return target;
  }
};
function buildManagedConstraintName(session) {
  const route = session.isDirect ? `direct:${session.userId}` : `guild:${session.guildId ?? session.channelId ?? "unknown"}`;
  return `managed:${session.platform}:${session.selfId}:${route}`;
}
__name(buildManagedConstraintName, "buildManagedConstraintName");
function managedNameFromBindingKey(bindingKey) {
  const target = bindingKey.includes(":preset:") ? bindingKey.slice(0, bindingKey.indexOf(":preset:")) : bindingKey;
  const parts = target.split(":");
  if (parts[0] === "shared" && parts.length >= 4) {
    return `managed:${parts[1]}:${parts[2]}:guild:${parts[3]}`;
  }
  if (parts[0] === "personal" && parts.length >= 5) {
    const scope = parts[3] === "direct" ? `direct:${parts[4]}` : `guild:${parts[3]}`;
    return `managed:${parts[1]}:${parts[2]}:${scope}`;
  }
  return void 0;
}
__name(managedNameFromBindingKey, "managedNameFromBindingKey");
function isConstraintMatched(constraint, session) {
  if (constraint.platform != null && constraint.platform !== session.platform) {
    return false;
  }
  if (constraint.selfId != null && constraint.selfId !== session.selfId) {
    return false;
  }
  if (constraint.guildId != null && constraint.guildId !== session.guildId) {
    return false;
  }
  if (constraint.channelId != null && constraint.channelId !== session.channelId) {
    return false;
  }
  if (constraint.direct != null && constraint.direct !== session.isDirect) {
    return false;
  }
  const users = constraint.users == null ? null : JSON.parse(constraint.users);
  if (users != null && !users.includes(session.userId)) {
    return false;
  }
  const excludeUsers = constraint.excludeUsers == null ? null : JSON.parse(constraint.excludeUsers);
  if (excludeUsers != null && excludeUsers.includes(session.userId)) {
    return false;
  }
  return true;
}
__name(isConstraintMatched, "isConstraintMatched");
async function assertManageAllowed(session, constraint) {
  if (constraint.manageMode !== "admin") {
    return;
  }
  if (await (0, import_koishi3.checkAdmin)(session)) {
    return;
  }
  throw new AdminRequiredError();
}
__name(assertManageAllowed, "assertManageAllowed");
function assertActionAllowed(action, resolved, managed, options = {}) {
  if (managed?.lockConversation ?? resolved.constraint.lockConversation) {
    throw new ConstraintLockedError(action);
  }
  if (options.needsAllow == null) {
    return;
  }
  const allowKey = options.needsAllow === "switch" ? "allowSwitch" : options.needsAllow === "archive" ? "allowArchive" : options.needsAllow === "export" ? "allowExport" : "allowNew";
  if (!(managed?.[allowKey] ?? resolved.constraint[allowKey])) {
    throw new ConstraintDisabledError(action);
  }
}
__name(assertActionAllowed, "assertActionAllowed");
async function hasConversationPermission(ctx, session, conversation, permission, bindingKey) {
  if (conversation.bindingKey === bindingKey) {
    return true;
  }
  if (await (0, import_koishi3.checkAdmin)(session)) {
    return true;
  }
  const acl = await ctx.database.get("chatluna_acl", {
    conversationId: conversation.id
  });
  if (acl.length === 0) {
    return false;
  }
  const principals = [
    ["user", session.userId],
    ["guild", session.guildId ?? session.channelId]
  ];
  const required = permission === "view" ? ["view", "manage"] : ["manage"];
  return acl.some(
    (item) => required.includes(item.permission) && principals.some(
      ([type, id]) => id != null && item.principalType === type && item.principalId === id
    )
  );
}
__name(hasConversationPermission, "hasConversationPermission");
async function readText(message) {
  const content = JSON.parse(await gzipDecode(message.content));
  if (content == null) {
    return message.text ?? "";
  }
  if (typeof content === "string") {
    return content;
  }
  return getMessageContent(content);
}
__name(readText, "readText");
function formatUrl(url) {
  return url.length > 120 ? url.slice(0, 117) + "..." : url;
}
__name(formatUrl, "formatUrl");
function formatMediaPart(part) {
  if ((0, import_langchain.isMessageContentText)(part)) {
    return null;
  }
  if ((0, import_langchain.isMessageContentImageUrl)(part)) {
    const url = typeof part.image_url === "string" ? part.image_url : part.image_url.url;
    return `[image] ${formatUrl(url)}`;
  }
  if ((0, import_langchain.isMessageContentFileUrl)(part)) {
    const url = typeof part.file_url === "string" ? part.file_url : part.file_url.url;
    return `[file] ${formatUrl(url)}`;
  }
  if ((0, import_langchain.isMessageContentAudio)(part)) {
    const url = typeof part.audio_url === "string" ? part.audio_url : part.audio_url.url;
    return `[audio] ${formatUrl(url)}`;
  }
  if ((0, import_langchain.isMessageContentVideo)(part)) {
    const url = typeof part.video_url === "string" ? part.video_url : part.video_url.url;
    return `[video] ${formatUrl(url)}`;
  }
  return `[${part.type}]`;
}
__name(formatMediaPart, "formatMediaPart");
function formatToolBlock(body) {
  let block = body;
  let lang = "text";
  try {
    block = JSON.stringify(JSON.parse(body), null, 2);
    lang = "json";
  } catch {
  }
  return ["```" + lang, block, "```"];
}
__name(formatToolBlock, "formatToolBlock");
function formatToolCalls(toolCalls) {
  const parts = ["Tool calls:"];
  toolCalls.forEach((entry, index) => {
    const tool = entry;
    const fn = tool.function;
    const name = typeof tool.name === "string" ? tool.name : typeof fn?.name === "string" ? fn.name : "unknown";
    const id = typeof tool.id === "string" ? tool.id : "";
    const raw = tool.args ?? fn?.arguments ?? {};
    let block;
    let lang = "json";
    if (typeof raw === "string") {
      block = raw;
      try {
        block = JSON.stringify(JSON.parse(raw), null, 2);
      } catch {
        lang = "text";
      }
    } else {
      block = JSON.stringify(raw, null, 2);
    }
    if (index > 0) parts.push("");
    parts.push(`- \`${name}\`${id.length > 0 ? ` (\`${id}\`)` : ""}`);
    parts.push("```" + lang);
    parts.push(block.length > 0 ? block : "{}");
    parts.push("```");
  });
  return parts;
}
__name(formatToolCalls, "formatToolCalls");
async function formatMessage(message) {
  const content = message.content == null ? null : JSON.parse(await gzipDecode(message.content));
  const text = content == null ? (message.text ?? "").trim() : typeof content === "string" ? content.trim() : getMessageContent(content).trim();
  const media = content != null && Array.isArray(content) ? content.map(formatMediaPart).filter((line) => line != null) : [];
  const parts = [];
  if (message.role === "tool") {
    if (message.tool_call_id != null && message.tool_call_id.length > 0) {
      parts.push(`Call ID: \`${message.tool_call_id}\``);
    }
    const body = text.length > 0 ? text : media.length > 0 ? media.join("\n") : message.text ?? "";
    if (body.length > 0) {
      if (parts.length > 0) parts.push("");
      parts.push(...formatToolBlock(body));
    }
  } else {
    if (text.length > 0) {
      parts.push(text);
    }
    if (media.length > 0) {
      if (parts.length > 0) parts.push("");
      parts.push("Attachments:");
      parts.push(...media.map((line) => `- ${line}`));
    }
  }
  if (Array.isArray(message.tool_calls) && message.tool_calls.length > 0) {
    if (parts.length > 0) parts.push("");
    parts.push(...formatToolCalls(message.tool_calls));
  }
  if (parts.length > 0) {
    return parts.join("\n");
  }
  if (content != null && typeof content === "string") {
    return content;
  }
  return message.text ?? "";
}
__name(formatMessage, "formatMessage");
function firstDefined(constraints, key) {
  for (const constraint of constraints) {
    if (constraint[key] != null) {
      return constraint[key];
    }
  }
  return void 0;
}
__name(firstDefined, "firstDefined");
function firstBoolean(constraints, key, fallback) {
  for (const constraint of constraints) {
    const value = constraint[key];
    if (typeof value === "boolean") {
      return value;
    }
  }
  return fallback;
}
__name(firstBoolean, "firstBoolean");
async function runLock(locks, key, fn) {
  let lock = locks.get(key);
  if (lock == null) {
    lock = new import_lock.ObjectLock();
    locks.set(key, lock);
  }
  try {
    return await lock.runLocked(fn);
  } finally {
    if (!lock.isLocked) {
      locks.delete(key);
    }
  }
}
__name(runLock, "runLock");
function matchTargetConversation(target, normalized, conversations, entries) {
  const pick = /* @__PURE__ */ __name((matches) => {
    const active = matches.filter((c) => c.status !== "archived");
    if (active.length === 1) return active[0];
    if (active.length > 1) {
      throw new ConversationResolutionError("ambiguous_target");
    }
    if (matches.length === 1) return matches[0];
    if (matches.length > 1) {
      throw new ConversationResolutionError("ambiguous_target");
    }
    return null;
  }, "pick");
  const byId = conversations.find((c) => c.id === target);
  if (byId != null) return byId;
  if (entries != null && /^\d+$/.test(target)) {
    const seq = Number(target);
    const bySeq = entries.filter((item) => item.displaySeq === seq).map((item) => item.conversation);
    const match = pick(bySeq);
    if (match != null) return match;
  }
  const exact = pick(
    conversations.filter((c) => c.title.toLocaleLowerCase() === normalized)
  );
  if (exact != null) return exact;
  return pick(
    conversations.filter(
      (c) => c.title.toLocaleLowerCase().includes(normalized)
    )
  );
}
__name(matchTargetConversation, "matchTargetConversation");

// src/services/conversation_runtime.ts
var import_messages = require("@langchain/core/messages");
var import_lru_cache = require("lru-cache");

// src/llm-core/agent/types.ts
var MessageQueue = class {
  static {
    __name(this, "MessageQueue");
  }
  _queue = [];
  push(...messages) {
    this._queue.push(...messages);
    return true;
  }
  drain() {
    return this._queue.splice(0);
  }
  get pending() {
    return this._queue.length > 0;
  }
};

// src/services/conversation_runtime.ts
var import_queue = require("koishi-plugin-chatluna/utils/queue");
var import_crypto3 = require("crypto");
var import_error4 = require("koishi-plugin-chatluna/utils/error");
var import_langchain2 = require("koishi-plugin-chatluna/utils/langchain");

// src/utils/model.ts
function parseRawModelName2(modelName) {
  if (modelName == null || modelName.trim().length < 1) {
    return [void 0, void 0];
  }
  const value = modelName.trim();
  const index = value.indexOf("/");
  if (index === -1) {
    return [void 0, value];
  }
  if (index === 0 || index === value.length - 1) {
    return [void 0, void 0];
  }
  return [value.slice(0, index), value.slice(index + 1)];
}
__name(parseRawModelName2, "parseRawModelName");

// src/services/conversation_runtime.ts
var ConversationRuntime = class {
  constructor(service) {
    this.service = service;
  }
  static {
    __name(this, "ConversationRuntime");
  }
  interfaces = new import_lru_cache.LRUCache({
    max: 20,
    dispose: /* @__PURE__ */ __name((value) => {
      value.chatInterface.dispose?.();
    }, "dispose")
  });
  modelQueue = new import_queue.RequestIdQueue();
  conversationQueue = new import_queue.RequestIdQueue();
  activeByConversation = /* @__PURE__ */ new Map();
  get platformService() {
    return this.service.platform;
  }
  async ensureChatInterface(conversation) {
    const cached = this.interfaces.get(conversation.id);
    if (cached != null) {
      cached.conversation = conversation;
      return cached.chatInterface;
    }
    const chatInterface = await this.service.createChatInterface(conversation);
    this.interfaces.set(conversation.id, {
      conversation,
      chatInterface
    });
    return chatInterface;
  }
  async chat(session, conversation, message, options = {}) {
    return this.withConversationAndPlatformLock(
      conversation,
      async (config) => this.internalChat(
        config,
        session,
        conversation,
        message,
        options
      )
    );
  }
  async internalChat(config, session, conversation, message, options) {
    const requestId = options.requestId ?? (0, import_crypto3.randomUUID)();
    const platform = requirePlatform(conversation);
    const chatInterface = await this.ensureChatInterface(conversation);
    const abortController = new AbortController();
    const releaseSignal = linkAbortSignal(abortController, options.signal);
    const activeRequest = this.registerRequest(
      conversation.id,
      requestId,
      conversation.chatMode,
      platform,
      abortController,
      session
    );
    let lastActiveAt = Date.now();
    const touch = /* @__PURE__ */ __name(() => {
      lastActiveAt = Date.now();
    }, "touch");
    const events = wrapEvents(options.event, touch);
    let releaseIdleTimer = /* @__PURE__ */ __name(() => {
    }, "releaseIdleTimer");
    if (config.timeout > 0) {
      releaseIdleTimer = this.service.ctx.setInterval(
        () => {
          if (abortController.signal.aborted) return;
          if (Date.now() - lastActiveAt < config.timeout) return;
          abortController.abort(
            new import_error4.ChatLunaError(
              import_error4.ChatLunaErrorCode.API_REQUEST_TIMEOUT,
              void 0,
              true
            )
          );
        },
        Math.min(config.timeout, 3e4)
      );
    }
    try {
      const humanMessage = buildHumanMessage(
        session,
        message,
        conversation
      );
      const stream = options.stream ?? false;
      const variables = options.variables ?? {};
      const mask = options.toolMask ?? await this.platformService.resolveToolMask({
        session,
        conversation,
        bindingKey: conversation.bindingKey
      });
      const chainValues = await chatInterface.chat({
        message: humanMessage,
        events,
        stream,
        conversationId: conversation.id,
        requestId,
        session,
        variables,
        signal: abortController.signal,
        postHandler: options.postHandler,
        messageQueue: activeRequest.messageQueue,
        toolMask: mask,
        callbacks: await this.service.resolveCallbacks({
          session,
          conversation,
          message,
          event: events,
          stream,
          variables,
          postHandler: options.postHandler,
          requestId,
          toolMask: mask,
          callbacks: options.callbacks
        }),
        onAgentEvent: /* @__PURE__ */ __name(async (agentEvent) => {
          touch();
          if (agentEvent.type !== "round-decision") return;
          activeRequest.lastDecision = agentEvent.canContinue;
          if (agentEvent.canContinue == null) return;
          flushRoundDecision(activeRequest, agentEvent.canContinue);
        }, "onAgentEvent")
      });
      return this.buildReply(chainValues.message);
    } finally {
      releaseIdleTimer();
      releaseSignal();
      this.completeRequest(conversation.id, requestId);
    }
  }
  buildReply(aiMessage) {
    const reasoning = aiMessage.additional_kwargs?.reasoning_content;
    const reasoningTime = aiMessage.additional_kwargs?.reasoning_time;
    const usage = aiMessage.usage_metadata;
    const showThought = this.service.currentConfig.showThoughtMessage;
    const additionalReplyMessages = [];
    if (showThought && reasoning != null && reasoning.length > 0) {
      additionalReplyMessages.push({
        content: reasoningTime != null ? `Thought for ${reasoningTime / 1e3} seconds: 

${reasoning}` : `Thought: 

${reasoning}`
      });
    }
    if (showThought && usage != null && usage.total_tokens > 0) {
      additionalReplyMessages.push({
        content: formatUsageMetadataMessage(usage)
      });
    }
    return {
      content: aiMessage.content,
      additional_kwargs: aiMessage.additional_kwargs,
      additionalReplyMessages
    };
  }
  updateConversationRecord(conversation) {
    const cached = this.interfaces.get(conversation.id);
    if (cached != null) {
      cached.conversation = conversation;
    }
  }
  getCachedConversations() {
    return Array.from(this.interfaces.entries());
  }
  async withConversationLock(conversationId, callback) {
    const requestId = (0, import_crypto3.randomUUID)();
    try {
      await this.conversationQueue.add(conversationId, requestId);
      await this.conversationQueue.wait(conversationId, requestId, 0);
      return await callback();
    } finally {
      await this.conversationQueue.remove(conversationId, requestId);
    }
  }
  async withConversationSync(conversation, callback) {
    const requestId = (0, import_crypto3.randomUUID)();
    try {
      await this.conversationQueue.add(conversation.id, requestId);
      this.stopConversationRequest(conversation.id);
      await this.conversationQueue.wait(conversation.id, requestId, 0);
      return await callback();
    } finally {
      await this.conversationQueue.remove(conversation.id, requestId);
    }
  }
  async withConversationAndPlatformLock(conversation, callback) {
    const requestId = (0, import_crypto3.randomUUID)();
    const modelRequestId = (0, import_crypto3.randomUUID)();
    const platform = requirePlatform(conversation);
    const client = await this.platformService.getClient(platform);
    if (client.value == null) {
      await this.service.awaitLoadPlatform(platform);
    }
    if (client.value == null) {
      throw new import_error4.ChatLunaError(
        import_error4.ChatLunaErrorCode.CONVERSATION_PLATFORM_NOT_AVAILABLE,
        new Error(`Platform ${platform} is not available`)
      );
    }
    const config = client.value.configPool.getConfig(true).value;
    try {
      await Promise.all([
        this.conversationQueue.add(conversation.id, requestId),
        this.modelQueue.add(platform, modelRequestId)
      ]);
      await Promise.all([
        this.conversationQueue.wait(
          conversation.id,
          requestId,
          0,
          config.timeout
        ),
        this.modelQueue.wait(
          platform,
          modelRequestId,
          config.concurrentMaxSize,
          config.timeout
        )
      ]);
      return await callback(config);
    } finally {
      await Promise.all([
        this.conversationQueue.remove(conversation.id, requestId),
        this.modelQueue.remove(platform, modelRequestId)
      ]);
    }
  }
  registerRequest(conversationId, requestId, chatMode, platform, abortController, session) {
    const activeRequest = {
      requestId,
      conversationId,
      requestKey: session == null ? void 0 : buildRequestKey(session, conversationId),
      platform,
      abortController,
      chatMode,
      messageQueue: new MessageQueue(),
      roundDecisionResolvers: []
    };
    this.activeByConversation.set(conversationId, activeRequest);
    return activeRequest;
  }
  completeRequest(conversationId, requestId) {
    const active = this.activeByConversation.get(conversationId);
    if (active?.requestId === requestId) {
      flushRoundDecision(active, false);
      this.activeByConversation.delete(conversationId);
    }
  }
  stopRequest(requestId) {
    const active = Array.from(this.activeByConversation.values()).find(
      (item) => item.requestId === requestId
    );
    if (active == null || active.abortController.signal.aborted) {
      return false;
    }
    active.abortController.abort(
      new import_error4.ChatLunaError(import_error4.ChatLunaErrorCode.ABORTED, void 0, true)
    );
    return true;
  }
  stopConversationRequest(conversationId) {
    const activeRequest = this.activeByConversation.get(conversationId);
    return activeRequest == null ? false : this.stopRequest(activeRequest.requestId);
  }
  getRequestId(session, conversationId) {
    const active = this.activeByConversation.get(conversationId);
    if (active == null) return void 0;
    if (active.requestKey !== buildRequestKey(session, conversationId)) {
      return void 0;
    }
    return active.requestId;
  }
  async appendPendingMessage(conversationId, message, chatMode) {
    if (chatMode != null && chatMode !== "plugin") {
      return false;
    }
    const activeRequest = this.activeByConversation.get(conversationId);
    if (activeRequest == null || activeRequest.chatMode !== "plugin") {
      return false;
    }
    if (activeRequest.lastDecision != null) {
      if (activeRequest.lastDecision) {
        activeRequest.messageQueue.push(message);
      }
      return activeRequest.lastDecision;
    }
    return new Promise((resolve) => {
      activeRequest.roundDecisionResolvers.push((canContinue) => {
        if (canContinue) {
          activeRequest.messageQueue.push(message);
        }
        resolve(canContinue);
      });
    });
  }
  async clearConversationCache(conversationId) {
    return this.interfaces.delete(conversationId);
  }
  async clearConversationHistory(conversation) {
    return this.withConversationLock(conversation.id, async () => {
      const chatInterface = await this.ensureChatInterface(conversation);
      await this.service.ctx.root.parallel(
        "chatluna/before-conversation-clear-history",
        { conversation, chatInterface }
      );
      await this.service.ctx.root.parallel(
        "chatluna/clear-chat-history",
        conversation.id,
        chatInterface
      );
      await chatInterface.clearChatHistory();
      this.interfaces.delete(conversation.id);
      await this.service.ctx.root.parallel(
        "chatluna/after-conversation-clear-history",
        { conversation, chatInterface }
      );
    });
  }
  async compressConversation(conversation, force = false, instruction) {
    return this.withConversationAndPlatformLock(conversation, async () => {
      const chatInterface = await this.ensureChatInterface(conversation);
      return await chatInterface.compressContext(force, instruction);
    });
  }
  async clearConversationInterfaceLocked(conversation) {
    const cached = this.interfaces.get(conversation.id);
    const existed = cached != null;
    await this.service.ctx.root.parallel(
      "chatluna/before-conversation-cache-clear",
      { conversation, chatInterface: cached?.chatInterface }
    );
    this.interfaces.delete(conversation.id);
    await this.service.ctx.root.parallel(
      "chatluna/after-conversation-cache-clear",
      { conversation }
    );
    return existed;
  }
  async clearConversationInterface(conversation) {
    return this.withConversationLock(
      conversation.id,
      () => this.clearConversationInterfaceLocked(conversation)
    );
  }
  dispose(platform) {
    const abortActive = /* @__PURE__ */ __name((active) => {
      flushRoundDecision(active, false);
      active.abortController.abort(
        new import_error4.ChatLunaError(import_error4.ChatLunaErrorCode.ABORTED, void 0, true)
      );
    }, "abortActive");
    if (platform == null) {
      for (const active of this.activeByConversation.values()) {
        abortActive(active);
      }
      this.interfaces.clear();
      this.activeByConversation.clear();
      return;
    }
    for (const active of Array.from(this.activeByConversation.values())) {
      if (active.platform !== platform) continue;
      abortActive(active);
      this.activeByConversation.delete(active.conversationId);
      this.interfaces.delete(active.conversationId);
    }
    for (const [conversationId, entry] of Array.from(
      this.interfaces.entries()
    )) {
      if (parseRawModelName2(entry.conversation.model)[0] === platform) {
        this.interfaces.delete(conversationId);
      }
    }
  }
};
var EVENT_KEYS = [
  "llm-new-token",
  "llm-queue-waiting",
  "llm-used-token-count",
  "llm-usage",
  "llm-call-tool",
  "llm-new-chunk"
];
function wrapEvents(source, touch) {
  const out = {};
  for (const key of EVENT_KEYS) {
    const fn = source?.[key];
    out[key] = async (...args) => {
      touch();
      await fn?.(...args);
    };
  }
  return out;
}
__name(wrapEvents, "wrapEvents");
function linkAbortSignal(controller, upstream) {
  if (upstream == null) return () => {
  };
  if (upstream.aborted) {
    controller.abort(upstream.reason);
    return () => {
    };
  }
  const onAbort = /* @__PURE__ */ __name(() => controller.abort(upstream.reason), "onAbort");
  upstream.addEventListener("abort", onAbort, { once: true });
  return () => upstream.removeEventListener("abort", onAbort);
}
__name(linkAbortSignal, "linkAbortSignal");
function buildHumanMessage(session, message, conversation) {
  const humanMessage = new import_messages.HumanMessage({
    content: message.content,
    name: message.name,
    id: session.userId,
    additional_kwargs: {
      ...message.additional_kwargs,
      preset: conversation.preset
    }
  });
  (0, import_langchain2.markChatLunaUserMessage)(humanMessage);
  return humanMessage;
}
__name(buildHumanMessage, "buildHumanMessage");
function buildRequestKey(session, conversationId) {
  return JSON.stringify([
    session.userId,
    session.guildId ?? "",
    conversationId
  ]);
}
__name(buildRequestKey, "buildRequestKey");
function requirePlatform(conversation) {
  const [platform] = parseRawModelName2(conversation.model);
  if (platform == null) {
    throw new import_error4.ChatLunaError(
      import_error4.ChatLunaErrorCode.CONVERSATION_INVALID_MODEL,
      new Error(`Invalid conversation model: ${conversation.model}`)
    );
  }
  return platform;
}
__name(requirePlatform, "requirePlatform");
function formatTokenDetail(detail, fields) {
  if (detail == null) return [];
  const parts = [];
  for (const { key, positiveOnly } of fields) {
    const value = detail[key];
    if (value == null) continue;
    if (positiveOnly && !(value > 0)) continue;
    parts.push(`${key}=${value}`);
  }
  return parts;
}
__name(formatTokenDetail, "formatTokenDetail");
function formatUsageMetadataMessage(usage) {
  const input = formatTokenDetail(usage.input_token_details, [
    { key: "audio", positiveOnly: true },
    { key: "image", positiveOnly: true },
    { key: "cache_read" },
    { key: "cache_creation" }
  ]);
  const output = formatTokenDetail(usage.output_token_details, [
    { key: "audio", positiveOnly: true },
    { key: "image", positiveOnly: true },
    { key: "reasoning" }
  ]);
  const lines = [
    "Token usage:",
    `- input: ${usage.input_tokens}`,
    `- output: ${usage.output_tokens}`,
    `- total: ${usage.total_tokens}`
  ];
  if (input.length > 0) lines.push(`- input details: ${input.join(", ")}`);
  if (output.length > 0) lines.push(`- output details: ${output.join(", ")}`);
  return lines.join("\n");
}
__name(formatUsageMetadataMessage, "formatUsageMetadataMessage");
function flushRoundDecision(active, canContinue) {
  for (const resolve of active.roundDecisionResolvers) {
    resolve(canContinue);
  }
  active.roundDecisionResolvers = [];
}
__name(flushRoundDecision, "flushRoundDecision");

// src/services/chat.ts
var import_request = require("koishi-plugin-chatluna/utils/request");
var import_koishi_plugin_chatluna3 = require("koishi-plugin-chatluna");
var import_promise = require("koishi-plugin-chatluna/utils/promise");
var import_in_memory = require("koishi-plugin-chatluna/llm-core/model/in_memory");

// src/services/prompt_renderer.ts
var import_messages2 = require("@langchain/core/messages");
var import_koishi4 = require("koishi");
var import_koishi_plugin_chatluna2 = require("koishi-plugin-chatluna");
var import_shared_prompt_renderer = require("@chatluna/shared-prompt-renderer");
var import_string2 = require("koishi-plugin-chatluna/utils/string");
var ChatLunaPromptRenderService = class {
  static {
    __name(this, "ChatLunaPromptRenderService");
  }
  _renderer;
  constructor() {
    this._renderer = new import_shared_prompt_renderer.ChatLunaPromptRenderer();
    this._initBuiltinFunctions();
  }
  _initBuiltinFunctions() {
    this.registerFunctionProvider("time_UTC", (args) => {
      const date = /* @__PURE__ */ new Date();
      const utcOffset = args[0] ? parseInt(args[0]) : 0;
      if (isNaN(utcOffset)) {
        import_koishi_plugin_chatluna2.logger.warn(`Invalid UTC offset: ${args[0]}`);
        return "Invalid UTC offset";
      }
      const offsetDate = new Date(+date + utcOffset * import_koishi4.Time.hour);
      return offsetDate.toISOString().replace("T", " ").slice(0, -5);
    });
    this.registerFunctionProvider("timeDiff", (args) => {
      return (0, import_string2.getTimeDiff)(args[0], args[1]);
    });
    this.registerFunctionProvider("date", () => {
      const date = /* @__PURE__ */ new Date();
      const offsetDate = new Date(
        +date - date.getTimezoneOffset() * import_koishi4.Time.minute
      );
      return offsetDate.toISOString().split("T")[0];
    });
    this.registerFunctionProvider("weekday", () => {
      const date = /* @__PURE__ */ new Date();
      return [
        "Sunday",
        "Monday",
        "Tuesday",
        "Wednesday",
        "Thursday",
        "Friday",
        "Saturday"
      ][date.getDay()];
    });
    this.registerFunctionProvider("isotime", () => {
      const date = /* @__PURE__ */ new Date();
      const offsetDate = new Date(
        +date - date.getTimezoneOffset() * import_koishi4.Time.minute
      );
      return offsetDate.toISOString().slice(11, 19);
    });
    this.registerFunctionProvider("isodate", () => {
      const date = /* @__PURE__ */ new Date();
      const offsetDate = new Date(
        +date - date.getTimezoneOffset() * import_koishi4.Time.minute
      );
      return offsetDate.toISOString().split("T")[0];
    });
    this.registerFunctionProvider("random", (args) => {
      if (args.length === 2) {
        const [min, max] = args.map(Number);
        if (!isNaN(min) && !isNaN(max)) {
          return Math.floor(
            Math.random() * (max - min + 1) + min
          ).toString();
        }
      }
      return (0, import_string2.selectFromList)(args.join(","), false);
    });
    this.registerFunctionProvider("pick", (args, _variables, cfg) => {
      return (0, import_string2.selectFromList)(
        args.join(","),
        true,
        cfg?.conversationId ?? ""
      );
    });
    this.registerFunctionProvider("roll", (args) => {
      return (0, import_string2.rollDice)(args[0]).toString();
    });
    this.registerFunctionProvider("url", async (args) => {
      return await (0, import_string2.fetchUrl)(
        args[1],
        args[0],
        args[2],
        parseInt(args[3] ?? "1000")
      );
    });
  }
  registerFunctionProvider(name, handler) {
    return this._renderer.registerFunctionProvider(name, handler);
  }
  registerVariableProvider(provider) {
    return this._renderer.registerVariableProvider(provider);
  }
  setVariable(name, value) {
    this._renderer.setStaticVariable(name, value);
  }
  getVariable(name) {
    return this._renderer.getStaticVariable(name);
  }
  removeVariable(name) {
    this._renderer.removeStaticVariable(name);
  }
  async renderTemplate(source, variables = {}, options) {
    return await this._renderer.render(source, variables, options);
  }
  async renderMessages(messages, variables = {}, options) {
    return await Promise.all(
      messages.map(async (message) => {
        const content = await this.renderTemplate(
          (0, import_string2.getMessageContent)(message.content),
          variables,
          options
        );
        const messageInstance = new {
          human: import_messages2.HumanMessage,
          ai: import_messages2.AIMessage,
          system: import_messages2.SystemMessage
        }[message.getType()]({
          content: content.text,
          additional_kwargs: message.additional_kwargs
        });
        return messageInstance;
      })
    );
  }
  async renderPresetTemplate(presetTemplate, variables = {}, options) {
    const collectedVariables = /* @__PURE__ */ new Set();
    const formattedMessages = await Promise.all(
      presetTemplate.messages.map(async (message) => {
        const content = await this.renderTemplate(
          (0, import_string2.getMessageContent)(message.content),
          variables,
          options
        );
        const messageInstance = new {
          human: import_messages2.HumanMessage,
          ai: import_messages2.AIMessage,
          system: import_messages2.SystemMessage
        }[message.getType()]({
          content: content.text,
          additional_kwargs: message.additional_kwargs
        });
        for (const variable of content.variables) {
          collectedVariables.add(variable);
        }
        return messageInstance;
      })
    );
    return {
      messages: formattedMessages,
      variables: Array.from(collectedVariables)
    };
  }
};

// src/services/chat.ts
var import_reactivity = require("@vue/reactivity");
var import_prompt = require("koishi-plugin-chatluna/llm-core/prompt");
var import_prompt2 = require("koishi-plugin-chatluna/llm-core/chain/prompt");

// src/services/types.ts
var types_exports = {};
__export(types_exports, {
  AdminRequiredError: () => AdminRequiredError,
  ConstraintDisabledError: () => ConstraintDisabledError,
  ConstraintFixedError: () => ConstraintFixedError,
  ConstraintLockedError: () => ConstraintLockedError,
  ConversationNotFoundError: () => ConversationNotFoundError,
  ConversationResolutionError: () => ConversationResolutionError,
  InvalidChatModeError: () => InvalidChatModeError,
  applyPresetLane: () => applyPresetLane,
  computeBaseBindingKey: () => computeBaseBindingKey,
  getBaseBindingKey: () => getBaseBindingKey,
  getPresetLane: () => getPresetLane
});
__reExport(types_exports, require("@chatluna/shared-prompt-renderer"));

// src/services/chat.ts
__reExport(chat_exports, types_exports, module.exports);

// src/llm-core/prompt/context_manager.ts
var import_messages3 = require("@langchain/core/messages");
var STAGE_ORDER = {
  system_prompts: 0,
  after_system_prompts: 50,
  chat_history: 100,
  long_history: 200,
  injections: 300,
  input: 400,
  scratchpad: 500,
  after_scratchpad: 600
};
var ChatLunaContextManagerService = class _ChatLunaContextManagerService {
  static {
    __name(this, "ChatLunaContextManagerService");
  }
  // -- injection middlewares (per-name, handles a single AnchoredInjection) --
  _middlewares = /* @__PURE__ */ new Map();
  // -- pipeline middlewares (per-stage, handles the whole stage) --
  _pipelineMiddlewares = [];
  // -- storage --
  _conversationPersistent = /* @__PURE__ */ new Map();
  _conversationQueue = /* @__PURE__ */ new Map();
  _skillProviders = /* @__PURE__ */ new Set();
  _coreRegistered = false;
  ensureCoreMiddlewares(register) {
    if (this._coreRegistered) return;
    this._coreRegistered = true;
    register();
  }
  constructor(ctx) {
    const clearQueue = /* @__PURE__ */ __name((conversationId) => {
      this._conversationQueue.delete(conversationId);
    }, "clearQueue");
    ctx.on(
      "chatluna/after-conversation-clear-history",
      async (payload) => this.clearConversation(payload.conversation.id)
    );
    ctx.on(
      "chatluna/after-conversation-archive",
      async (payload) => clearQueue(payload.conversation.id)
    );
    ctx.on(
      "chatluna/after-conversation-restore",
      async (payload) => clearQueue(payload.conversation.id)
    );
    ctx.on(
      "chatluna/after-conversation-delete",
      async (payload) => this.clearConversation(payload.conversation.id)
    );
  }
  // -----------------------------------------------------------------------
  // Pipeline middleware registration
  // -----------------------------------------------------------------------
  /**
   * Register a pipeline middleware for a given stage.
   *
   * Pipeline middlewares execute in `(stage-order, priority)` order.
   * Lower priority = earlier within the same stage.
   *
   * Returns a disposer function.
   */
  pipeline(stage, middleware, priority = 0) {
    const entry = { stage, middleware, priority };
    this._pipelineMiddlewares.push(entry);
    this._pipelineMiddlewares.sort((a, b) => {
      const sa = STAGE_ORDER[a.stage] ?? 999;
      const sb = STAGE_ORDER[b.stage] ?? 999;
      if (sa !== sb) return sa - sb;
      return a.priority - b.priority;
    });
    return () => {
      const idx = this._pipelineMiddlewares.indexOf(entry);
      if (idx !== -1) this._pipelineMiddlewares.splice(idx, 1);
    };
  }
  /**
   * Execute the full pipeline.  Each registered pipeline middleware is
   * called in order.  The `next()` function advances to the next
   * middleware.
   */
  async runPipeline(runtime) {
    const entries = [...this._pipelineMiddlewares];
    let index = -1;
    const dispatch = /* @__PURE__ */ __name(async (step) => {
      if (step <= index) {
        throw new Error("Pipeline middleware called next() twice");
      }
      index = step;
      const current = entries[step];
      if (!current) return;
      await current.middleware(runtime, () => dispatch(step + 1));
    }, "dispatch");
    await dispatch(0);
  }
  // -----------------------------------------------------------------------
  // Injection middleware registration (per-name, backward-compatible)
  // -----------------------------------------------------------------------
  intercept(name, middleware, priority = 0) {
    const wrappers = this._middlewares.get(name) ?? [];
    const wrapper = { middleware, priority };
    const insertAt = wrappers.findIndex((item) => item.priority > priority);
    if (insertAt === -1) {
      wrappers.push(wrapper);
    } else {
      wrappers.splice(insertAt, 0, wrapper);
    }
    this._middlewares.set(name, wrappers);
    return () => {
      const current = this._middlewares.get(name);
      if (!current) return;
      const idx = current.findIndex(
        (item) => item.middleware === middleware
      );
      if (idx === -1) return;
      if (current.length === 1) {
        this._middlewares.delete(name);
      } else {
        current.splice(idx, 1);
      }
    };
  }
  replace(name, middleware) {
    this._middlewares.set(name, [{ middleware, priority: 0 }]);
    return () => this._middlewares.delete(name);
  }
  has(name) {
    const wrappers = this._middlewares.get(name);
    return wrappers != null && wrappers.length > 0;
  }
  // -----------------------------------------------------------------------
  // Injection storage
  // -----------------------------------------------------------------------
  /**
   * Inject content into a conversation's context.
   *
   * Content can be anchored between two messages (by message ID).  As long
   * as both anchor messages exist in the assembled prompt, the injection
   * will be placed between them.
   *
   * If `once` is true the injection is consumed after one collection.
   * Otherwise it persists until the anchor messages are gone or it is
   * cleared.
   */
  inject(options) {
    const injection = this._createInjection(options);
    if (!options.conversationId) return;
    const store = options.once ? this._conversationQueue : this._conversationPersistent;
    this._addToStore(store, options.conversationId, injection);
  }
  /**
   * Remove a specific persistent injection by id.
   */
  removeInjection(conversationId, injectionId) {
    const list = this._conversationPersistent.get(conversationId);
    if (!list) return false;
    const idx = list.findIndex((item) => item.id === injectionId);
    if (idx === -1) return false;
    list.splice(idx, 1);
    return true;
  }
  // -----------------------------------------------------------------------
  // Collecting injections for the current prompt render
  // -----------------------------------------------------------------------
  collectInjections({
    variables,
    configurable,
    afterUserMessage,
    currentMessages
  }) {
    const conversationId = configurable?.conversationId;
    for (const msg of currentMessages) {
      if (!msg.id) {
        msg.id = this._createId();
      }
    }
    const liveIds = new Set(
      currentMessages.map((m) => m.id).filter(Boolean)
    );
    const collected = [];
    if (conversationId) {
      const persistent = this._conversationPersistent.get(conversationId) ?? [];
      const alive = persistent.filter((inj) => {
        if (inj.beforeMessageId && !liveIds.has(inj.beforeMessageId)) {
          return false;
        }
        if (!inj.afterMessageId) return true;
        return liveIds.has(inj.afterMessageId);
      });
      if (alive.length !== persistent.length) {
        this._conversationPersistent.set(conversationId, alive);
      }
      collected.push(...alive);
      const queued = this._conversationQueue.get(conversationId) ?? [];
      if (queued.length > 0) {
        this._conversationQueue.delete(conversationId);
      }
      collected.push(...queued);
    }
    if (Array.isArray(variables?.["lore_books"]) && variables["lore_books"].length > 0) {
      collected.push(
        this._createInjection({
          name: "lore_books",
          value: variables["lore_books"],
          stage: "injections"
        })
      );
    }
    if (variables?.["authors_note"]) {
      collected.push(
        this._createInjection({
          name: "authors_note",
          value: variables["authors_note"],
          stage: "injections"
        })
      );
    }
    if (afterUserMessage) {
      collected.push(
        this._createInjection({
          name: "after_user_message",
          value: afterUserMessage,
          stage: "after_scratchpad"
        })
      );
    }
    collected.sort((a, b) => {
      if (a.priority !== b.priority) return a.priority - b.priority;
      return a.createdAt - b.createdAt;
    });
    return {
      beforeScratchpad: collected.filter(
        (item) => item.stage !== "after_scratchpad" && item.stage !== "scratchpad"
      ),
      afterScratchpad: collected.filter(
        (item) => item.stage === "after_scratchpad"
      )
    };
  }
  // -----------------------------------------------------------------------
  // Applying injections (runs per-name middleware chains)
  // -----------------------------------------------------------------------
  async applyInjections(injections, runtime) {
    for (const injection of injections) {
      await this._applySingleInjection(injection, runtime);
    }
    return runtime;
  }
  // -----------------------------------------------------------------------
  // Anchor helpers – find position by message ID
  // -----------------------------------------------------------------------
  /**
   * Find the index in `messages` where content anchored by
   * `afterMessageId` / `beforeMessageId` should be inserted.
   *
   * Matches against `BaseMessage.id`.  Returns the splice index.
   */
  static findAnchorIndex(messages, afterMessageId, beforeMessageId) {
    if (afterMessageId) {
      const idx = messages.findIndex((m) => m.id === afterMessageId);
      if (idx !== -1) return idx + 1;
    }
    if (beforeMessageId) {
      const idx = messages.findIndex((m) => m.id === beforeMessageId);
      if (idx !== -1) return idx;
    }
    return messages.length;
  }
  /**
   * Return true when all specified anchor messages are present in
   * `messages` (matched by `BaseMessage.id`).
   */
  static anchorsExist(messages, afterMessageId, beforeMessageId) {
    if (!afterMessageId && !beforeMessageId) return true;
    const ids = new Set(messages.map((m) => m.id).filter(Boolean));
    if (afterMessageId && !ids.has(afterMessageId)) return false;
    if (beforeMessageId && !ids.has(beforeMessageId)) return false;
    return true;
  }
  // -----------------------------------------------------------------------
  // Cleanup
  // -----------------------------------------------------------------------
  clearConversation(conversationId) {
    this._conversationQueue.delete(conversationId);
    this._conversationPersistent.delete(conversationId);
  }
  clearAll() {
    this._conversationQueue.clear();
    this._conversationPersistent.clear();
  }
  registerSkillProvider(provider) {
    this._skillProviders.add(provider);
    return () => {
      this._skillProviders.delete(provider);
    };
  }
  // -----------------------------------------------------------------------
  // Private helpers
  // -----------------------------------------------------------------------
  _createInjection(options) {
    return {
      id: this._createId(),
      name: options.name,
      value: options.value,
      afterMessageId: options.afterMessageId,
      beforeMessageId: options.beforeMessageId,
      stage: options.stage ?? this._resolveDefaultStage(options.name),
      createdAt: Date.now(),
      priority: options.priority ?? 0,
      once: options.once
    };
  }
  _resolveDefaultStage(name) {
    if (name === "after_user_message" || name === "tool_observation") {
      return "after_scratchpad";
    }
    return "injections";
  }
  _createId() {
    return `${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
  }
  _addToStore(store, key, injection) {
    const current = store.get(key);
    if (!current) {
      store.set(key, [injection]);
      return;
    }
    current.push(injection);
  }
  async _applySingleInjection(injection, runtime) {
    const wrappers = [
      ...this._middlewares.get("*") ?? [],
      ...this._middlewares.get(injection.name) ?? []
    ];
    const context = {
      injection,
      runtime,
      handled: false,
      markHandled: /* @__PURE__ */ __name(() => {
        context.handled = true;
      }, "markHandled"),
      appendMessages: /* @__PURE__ */ __name((input) => {
        const messages = toMessages(input);
        if (messages.length > 0) {
          runtime.result.push(...messages);
        }
        return messages;
      }, "appendMessages"),
      insertAtAnchor: /* @__PURE__ */ __name((input) => {
        const messages = Array.isArray(input) ? input : [input];
        if (messages.length === 0) return runtime.result.length;
        const idx = _ChatLunaContextManagerService.findAnchorIndex(
          runtime.result,
          injection.afterMessageId,
          injection.beforeMessageId
        );
        runtime.result.splice(idx, 0, ...messages);
        return idx;
      }, "insertAtAnchor")
    };
    if (wrappers.length > 0) {
      let index = -1;
      const dispatch = /* @__PURE__ */ __name(async (step) => {
        if (step <= index) {
          throw new Error(
            "Prompt context middleware called next() twice"
          );
        }
        index = step;
        const current = wrappers[step];
        if (!current) return;
        await current.middleware(context, () => dispatch(step + 1));
      }, "dispatch");
      await dispatch(0);
    }
    if (!context.handled) {
      const messages = toMessages(injection.value);
      if (messages.length > 0) {
        if (injection.afterMessageId || injection.beforeMessageId) {
          context.insertAtAnchor(messages);
        } else {
          runtime.result.push(...messages);
        }
      }
    }
  }
};
function toMessages(input) {
  if (input == null) return [];
  if (Array.isArray(input)) {
    return input.flatMap((item) => toMessages(item));
  }
  if (input instanceof import_messages3.BaseMessage) return [input];
  if (typeof input === "string") {
    if (input.trim().length < 1) return [];
    return [new import_messages3.HumanMessage(input)];
  }
  return [];
}
__name(toMessages, "toMessages");

// src/services/chat.ts
var ChatLunaService = class extends import_koishi5.Service {
  constructor(ctx, config) {
    super(ctx, "chatluna");
    this.ctx = ctx;
    this.config = config;
    this.currentConfig = config;
    this._chain = new ChatChain(ctx, config);
    this._keysCache = new Cache(this.ctx, config, "chatluna/keys");
    this._preset = new import_preset.PresetService(ctx, config);
    this._platformService = new import_service.PlatformService(ctx);
    this._messageTransformer = new MessageTransformer(config);
    this._renderer = new import_koishi_plugin_chatluna3.DefaultRenderer(ctx, config);
    this._promptRenderer = new ChatLunaPromptRenderService();
    this._contextManager = new import_prompt.ChatLunaContextManagerService(ctx);
    this._conversationRuntime = new ConversationRuntime(this);
    this._conversation = new ConversationService(
      ctx,
      config,
      this._conversationRuntime,
      this._platformService
    );
    this._createTempDir();
    this._defineDatabase();
    this.ctx.on("ready", async () => {
      await this._dedupeConstraintNames();
    });
  }
  static {
    __name(this, "ChatLunaService");
  }
  _plugins = {};
  _chain;
  _keysCache;
  _preset;
  _platformService;
  _messageTransformer;
  _renderer;
  _promptRenderer;
  _contextManager;
  _conversation;
  _conversationRuntime;
  _callbackProviders = /* @__PURE__ */ new Set();
  async installPlugin(plugin) {
    const platformName = plugin.platformName;
    if (this._plugins[platformName]) {
      throw new import_error5.ChatLunaError(
        import_error5.ChatLunaErrorCode.PLUGIN_ALREADY_REGISTERED,
        new Error(`Plugin ${platformName} already registered`)
      );
    }
    this._plugins[platformName] = plugin;
    this.ctx.logger.success(`Plugin %c installed`, platformName);
  }
  async awaitLoadPlatform(plugin, timeout = 3e4) {
    const pluginName = typeof plugin === "string" ? plugin : plugin.platformName;
    const { promise, resolve, reject } = (0, import_promise.withResolver)();
    const models = this._platformService.listPlatformModels(
      pluginName,
      import_types6.ModelType.all
    );
    if (models.value.length > 0) {
      resolve();
      return promise;
    }
    let timeoutError = null;
    try {
      throw new Error(
        `Timeout waiting for platform ${pluginName} to load`
      );
    } catch (e) {
      timeoutError = e;
    }
    const timeoutId = this.ctx.setTimeout(() => {
      reject(timeoutError);
    }, timeout);
    const disposable = (0, import_reactivity.watch)(
      models,
      () => {
        if ((models.value?.length ?? 0) > 0) {
          resolve();
          timeoutId();
          disposable.stop();
        }
      },
      { deep: true }
    );
    this[import_koishi5.Context.origin].effect(() => () => disposable.stop());
    return promise;
  }
  uninstallPlugin(plugin) {
    const platformName = typeof plugin === "string" ? plugin : plugin.platformName;
    const targetPlugin = this._plugins[platformName];
    if (!targetPlugin) {
      return;
    }
    const platform = targetPlugin.platformName;
    this._conversationRuntime.dispose(platform);
    delete this._plugins[platform];
    this.ctx.logger.success(
      "Plugin %c uninstalled",
      targetPlugin.platformName
    );
  }
  registerToolMaskResolver(name, resolver) {
    return this._platformService.registerToolMaskResolver(name, resolver);
  }
  async resolveToolMask(arg) {
    return this._platformService.resolveToolMask(arg);
  }
  getPlugin(platformName) {
    return this._plugins[platformName];
  }
  chat(session, conversation, message, options = {}) {
    return this._conversationRuntime.chat(
      session,
      conversation,
      message,
      options
    );
  }
  registerCallbacksProvider(provider) {
    this._callbackProviders.add(provider);
    return () => {
      this._callbackProviders.delete(provider);
    };
  }
  async resolveCallbacks(input) {
    let merged = input.callbacks;
    for (const provider of this._callbackProviders) {
      merged = import_manager.CallbackManager.configure(merged, await provider(input));
    }
    return merged;
  }
  async clearCache(conversation) {
    return this._conversationRuntime.clearConversationInterface(
      conversation
    );
  }
  async createChatInterface(conversation) {
    const config = this.currentConfig;
    const chatInterface = new import_app.ChatInterface(
      this.ctx,
      {
        chatMode: conversation.chatMode,
        autoTitle: conversation.autoTitle === true,
        botName: config.botNames[0],
        preset: this.preset.getPreset(conversation.preset),
        model: conversation.model,
        conversationId: conversation.id,
        embeddings: config.defaultEmbeddings && config.defaultEmbeddings.length > 0 ? config.defaultEmbeddings : void 0,
        vectorStoreName: config.defaultVectorStore && config.defaultVectorStore.length > 0 ? config.defaultVectorStore : void 0
      },
      this
    );
    return chatInterface;
  }
  async createChatModel(platformName, model) {
    const service = this._platformService;
    if (model == null) {
      ;
      [platformName, model] = (0, import_count_tokens2.parseRawModelName)(platformName);
    }
    const client = await service.getClient(platformName);
    return (0, import_reactivity.computed)(() => {
      if (client.value == null) {
        return void 0;
      }
      try {
        return client.value.createModel(model);
      } catch (error) {
        this.ctx.logger.warn(`The model ${model} not found`, error);
      }
      return void 0;
    });
  }
  async createEmbeddings(platformName, modelName) {
    const service = this._platformService;
    if (modelName == null) {
      ;
      [platformName, modelName] = (0, import_count_tokens2.parseRawModelName)(platformName);
    }
    if (platformName == null || modelName == null || modelName.length < 1 || modelName === "无") {
      return (0, import_reactivity.computed)(() => import_in_memory.emptyEmbeddings);
    }
    const client = await service.getClient(platformName);
    const info = service.findModel(platformName, modelName);
    return (0, import_reactivity.computed)(() => {
      if (info.value == null) {
        this.ctx.logger.warn(
          `The embeddings model ${modelName} not found, return empty embeddings`
        );
        return import_in_memory.emptyEmbeddings;
      }
      if (info.value.type !== import_types6.ModelType.embeddings) {
        this.ctx.logger.warn(
          `The model ${modelName} is not embeddings, return empty embeddings`
        );
        return import_in_memory.emptyEmbeddings;
      }
      if (client.value == null) {
        this.ctx.logger.warn(
          `The platform ${platformName} no available, return empty embeddings`
        );
        return import_in_memory.emptyEmbeddings;
      }
      try {
        const model = client.value.createModel(modelName);
        if (model instanceof import_model2.ChatLunaBaseEmbeddings) {
          return model;
        }
      } catch (error) {
        this.ctx.logger.warn(`The model ${modelName} not found`, error);
        return import_in_memory.emptyEmbeddings;
      }
      this.ctx.logger.warn(
        `The model ${modelName} is not embeddings, return empty embeddings`
      );
      return import_in_memory.emptyEmbeddings;
    });
  }
  async createReranker(platformName, modelName) {
    const service = this._platformService;
    if (modelName == null) {
      ;
      [platformName, modelName] = (0, import_count_tokens2.parseRawModelName)(platformName);
    }
    const client = await service.getClient(platformName);
    return (0, import_reactivity.computed)(() => {
      if (client.value == null) {
        if (platformName !== "无") {
          this.ctx.logger.warn(
            `The platform ${platformName} is not available`
          );
        }
        return void 0;
      }
      try {
        const model = client.value.createModel(modelName);
        if (model instanceof import_rerank.ChatLunaReranker) {
          return model;
        }
      } catch (error) {
        this.ctx.logger.warn(
          `The rerank model ${modelName} not found`,
          error
        );
        return void 0;
      }
      this.ctx.logger.warn(`The model ${modelName} is not a reranker`);
      return void 0;
    });
  }
  async createAgent(options) {
    const llm = await (0, import_agent.resolveAgentModel)(
      options.model,
      (name) => this.createChatModel(name)
    );
    const embeddings = await (0, import_agent.resolveAgentEmbeddings)(
      options.embeddings,
      (name) => this.createEmbeddings(name),
      this.currentConfig.defaultEmbeddings
    );
    const tools = (0, import_agent.resolveAgentTools)(
      options.tools,
      (name) => this.platform.getTool(name)
    );
    const { preset, instructions } = (0, import_agent.resolveAgentPreset)(
      options,
      (name) => (0, import_reactivity.computed)(() => this._preset.getPreset(name).value)
    );
    const model = llm.value;
    const prompt = options.prompt ?? new import_prompt2.ChatLunaChatPrompt({
      preset,
      tokenCounter: /* @__PURE__ */ __name((text) => model.getNumTokens(text), "tokenCounter"),
      sendTokenLimit: model.invocationParams().maxTokenLimit ?? model.getModelMaxContextSize(),
      contextManager: this._contextManager,
      promptRenderService: this._promptRenderer
    });
    return (0, import_agent.createAgent)({
      id: options.id,
      name: options.name,
      description: options.description,
      llm,
      embeddings,
      tools,
      prompt,
      mode: options.mode,
      maxSteps: options.maxSteps,
      handleParsingErrors: options.handleParsingErrors,
      instructions,
      returnIntermediateSteps: options.returnIntermediateSteps,
      toolMask: options.toolMask
    });
  }
  get platform() {
    return this._platformService;
  }
  get cache() {
    return this._keysCache;
  }
  get preset() {
    return this._preset;
  }
  get chatChain() {
    return this._chain;
  }
  get messageTransformer() {
    return this._messageTransformer;
  }
  get renderer() {
    return this._renderer;
  }
  get promptRenderer() {
    return this._promptRenderer;
  }
  get contextManager() {
    return this._contextManager;
  }
  get conversation() {
    return this._conversation;
  }
  get conversationRuntime() {
    return this._conversationRuntime;
  }
  async stop() {
    this._conversationRuntime.dispose();
    this._platformService.dispose();
    this._contextManager.clearAll();
  }
  _createTempDir() {
    const tempPath = import_path3.default.resolve(this.ctx.baseDir, "data/chatluna/temp");
    if (!import_fs.default.existsSync(tempPath)) {
      import_fs.default.mkdirSync(tempPath, { recursive: true });
    }
  }
  async _dedupeConstraintNames() {
    const rows = await this.ctx.database.get(
      "chatluna_constraint",
      {}
    );
    if (rows.length < 2) {
      return;
    }
    const names = /* @__PURE__ */ new Set();
    const ids = [...rows].sort((left, right) => {
      const leftTime = left.updatedAt?.getTime() ?? 0;
      const rightTime = right.updatedAt?.getTime() ?? 0;
      if (leftTime !== rightTime) {
        return rightTime - leftTime;
      }
      return (right.id ?? 0) - (left.id ?? 0);
    }).filter((row) => {
      if (!names.has(row.name)) {
        names.add(row.name);
        return false;
      }
      return row.id != null;
    }).map((row) => row.id);
    if (ids.length === 0) {
      return;
    }
    this.ctx.logger.warn(
      `Removing ${ids.length} duplicate chatluna_constraint rows.`
    );
    await this.ctx.database.remove("chatluna_constraint", {
      id: ids
    });
  }
  _defineDatabase() {
    const ctx = this.ctx;
    ctx.database.extend(
      "chatluna_conversation",
      {
        id: {
          type: "char",
          length: 255
        },
        seq: {
          type: "unsigned",
          nullable: true
        },
        bindingKey: {
          type: "string",
          length: 255
        },
        title: "string",
        model: {
          type: "char",
          length: 100
        },
        preset: {
          type: "char",
          length: 255
        },
        chatMode: {
          type: "char",
          length: 20
        },
        createdBy: {
          type: "char",
          length: 255
        },
        createdAt: {
          type: "timestamp",
          nullable: false,
          initial: /* @__PURE__ */ new Date()
        },
        updatedAt: {
          type: "timestamp",
          nullable: false,
          initial: /* @__PURE__ */ new Date()
        },
        lastChatAt: {
          type: "timestamp",
          nullable: true
        },
        status: {
          type: "char",
          length: 20
        },
        latestMessageId: {
          type: "char",
          length: 255,
          nullable: true
        },
        additional_kwargs: {
          type: "text",
          nullable: true
        },
        compression: {
          type: "text",
          nullable: true
        },
        archivedAt: {
          type: "timestamp",
          nullable: true
        },
        archiveId: {
          type: "char",
          length: 255,
          nullable: true
        },
        legacyRoomId: {
          type: "unsigned",
          nullable: true
        },
        legacyMeta: {
          type: "text",
          nullable: true
        },
        autoTitle: {
          type: "boolean",
          nullable: true
        }
      },
      {
        autoInc: false,
        primary: "id",
        unique: ["id"]
      }
    );
    ctx.database.extend(
      "chatluna_message",
      {
        id: {
          type: "char",
          length: 255
        },
        conversationId: {
          type: "char",
          length: 255
        },
        parentId: {
          type: "char",
          length: 255,
          nullable: true
        },
        role: {
          type: "char",
          length: 20
        },
        text: {
          type: "text",
          nullable: true
        },
        content: {
          type: "binary",
          nullable: true
        },
        name: {
          type: "char",
          length: 255,
          nullable: true
        },
        tool_call_id: {
          type: "string",
          nullable: true
        },
        tool_calls: {
          type: "json",
          nullable: true
        },
        additional_kwargs_binary: {
          type: "binary",
          nullable: true
        },
        response_metadata_binary: {
          type: "binary",
          nullable: true
        },
        rawId: {
          type: "char",
          length: 255,
          nullable: true
        },
        createdAt: {
          type: "timestamp",
          nullable: true
        }
      },
      {
        autoInc: false,
        primary: "id",
        unique: ["id"]
      }
    );
    ctx.database.extend(
      "chatluna_binding",
      {
        bindingKey: {
          type: "string",
          length: 255
        },
        activeConversationId: {
          type: "char",
          length: 255,
          nullable: true
        },
        lastConversationId: {
          type: "char",
          length: 255,
          nullable: true
        },
        updatedAt: {
          type: "timestamp",
          nullable: false,
          initial: /* @__PURE__ */ new Date()
        }
      },
      {
        autoInc: false,
        primary: "bindingKey",
        unique: ["bindingKey"]
      }
    );
    ctx.database.extend(
      "chatluna_constraint",
      {
        id: "unsigned",
        name: "string",
        enabled: {
          type: "boolean",
          initial: true
        },
        priority: {
          type: "integer",
          initial: 0
        },
        createdBy: {
          type: "char",
          length: 255
        },
        createdAt: {
          type: "timestamp",
          nullable: false,
          initial: /* @__PURE__ */ new Date()
        },
        updatedAt: {
          type: "timestamp",
          nullable: false,
          initial: /* @__PURE__ */ new Date()
        },
        platform: {
          type: "char",
          length: 255,
          nullable: true
        },
        selfId: {
          type: "char",
          length: 255,
          nullable: true
        },
        guildId: {
          type: "char",
          length: 255,
          nullable: true
        },
        channelId: {
          type: "char",
          length: 255,
          nullable: true
        },
        direct: {
          type: "boolean",
          nullable: true
        },
        users: {
          type: "text",
          nullable: true
        },
        excludeUsers: {
          type: "text",
          nullable: true
        },
        routeMode: {
          type: "char",
          length: 20,
          nullable: true
        },
        routeKey: {
          type: "string",
          length: 255,
          nullable: true
        },
        activePresetLane: {
          type: "char",
          length: 255,
          nullable: true
        },
        defaultModel: {
          type: "char",
          length: 100,
          nullable: true
        },
        defaultPreset: {
          type: "char",
          length: 255,
          nullable: true
        },
        defaultChatMode: {
          type: "char",
          length: 20,
          nullable: true
        },
        fixedModel: {
          type: "char",
          length: 100,
          nullable: true
        },
        fixedPreset: {
          type: "char",
          length: 255,
          nullable: true
        },
        fixedChatMode: {
          type: "char",
          length: 20,
          nullable: true
        },
        lockConversation: {
          type: "boolean",
          nullable: true
        },
        allowNew: {
          type: "boolean",
          nullable: true
        },
        allowSwitch: {
          type: "boolean",
          nullable: true
        },
        allowArchive: {
          type: "boolean",
          nullable: true
        },
        allowExport: {
          type: "boolean",
          nullable: true
        },
        manageMode: {
          type: "char",
          length: 20,
          nullable: true
        }
      },
      {
        autoInc: true,
        primary: "id",
        unique: ["name"]
      }
    );
    ctx.database.extend(
      "chatluna_archive",
      {
        id: {
          type: "char",
          length: 255
        },
        conversationId: {
          type: "char",
          length: 255
        },
        path: "string",
        formatVersion: {
          type: "unsigned"
        },
        messageCount: {
          type: "unsigned"
        },
        checksum: {
          type: "char",
          length: 255,
          nullable: true
        },
        size: {
          type: "unsigned"
        },
        state: {
          type: "char",
          length: 20
        },
        createdAt: {
          type: "timestamp",
          nullable: false,
          initial: /* @__PURE__ */ new Date()
        },
        restoredAt: {
          type: "timestamp",
          nullable: true
        }
      },
      {
        autoInc: false,
        primary: "id",
        unique: ["id"]
      }
    );
    ctx.database.extend(
      "chatluna_acl",
      {
        conversationId: {
          type: "char",
          length: 255
        },
        principalType: {
          type: "char",
          length: 20
        },
        principalId: {
          type: "char",
          length: 255
        },
        permission: {
          type: "char",
          length: 20
        }
      },
      {
        autoInc: false,
        primary: [
          "conversationId",
          "principalType",
          "principalId",
          "permission"
        ]
      }
    );
    ctx.database.extend(
      "chatluna_meta",
      {
        key: {
          type: "string",
          length: 255
        },
        value: {
          type: "text",
          nullable: true
        },
        updatedAt: {
          type: "timestamp",
          nullable: false,
          initial: /* @__PURE__ */ new Date()
        }
      },
      {
        autoInc: false,
        primary: "key",
        unique: ["key"]
      }
    );
    ctx.database.extend(
      "chatluna_docstore",
      {
        key: {
          type: "char",
          length: 255
        },
        id: {
          type: "char",
          length: 255
        },
        pageContent: "text",
        metadata: "json",
        createdAt: "date"
      },
      {
        autoInc: false,
        primary: ["key", "id"]
      }
    );
  }
  static inject = ["database"];
};
var ChatLunaPlugin = class {
  constructor(ctx, config, platformName, createConfigPool = true) {
    this.ctx = ctx;
    this.config = config;
    this.platformName = platformName;
    ctx.on("dispose", async () => {
      ctx.chatluna.uninstallPlugin(this);
    });
    ctx.on("ready", async () => {
      ctx.chatluna.installPlugin(this);
    });
    if (createConfigPool) {
      if (config == null) {
        const error = new Error("Check Config!");
        this.ctx.scope.cancel(error);
        throw error;
      }
      this.platformConfigPool = new import_config.ClientConfigPool(
        ctx,
        config.configMode === "default" ? import_config.ClientConfigPoolMode.AlwaysTheSame : import_config.ClientConfigPoolMode.LoadBalancing
      );
    }
    this._platformService = ctx.chatluna.platform;
    const models = this._platformService.listPlatformModels(
      this.platformName,
      import_types6.ModelType.llm
    );
    const watcher = (0, import_reactivity.watch)(
      models,
      () => {
        this._supportModels = (models.value ?? []).map(
          (model) => `${this.platformName}/${model.name}`
        );
      },
      { deep: true }
    );
    const stop = /* @__PURE__ */ __name(() => watcher.stop(), "stop");
    this.ctx.effect(() => stop);
  }
  static {
    __name(this, "ChatLunaPlugin");
  }
  _supportModels = [];
  platformConfigPool;
  _platformService;
  parseConfig(f) {
    const configs = f(this.config);
    for (const config of configs) {
      this.platformConfigPool.addConfig(config);
    }
  }
  createRunnableConfig() {
    const abortController = new AbortController();
    const abort = /* @__PURE__ */ __name(() => abortController.abort(
      new import_error5.ChatLunaError(import_error5.ChatLunaErrorCode.ABORTED, void 0, true)
    ), "abort");
    this.ctx.effect(() => abort);
    return {
      signal: abortController.signal
    };
  }
  async initClient() {
    let notification;
    let result;
    this.ctx.inject(["notifier"], (ctx) => {
      if (notification) return;
      if (result) {
        notification = ctx.notifier.create({
          content: result.content,
          type: result.type
        });
      } else {
        notification = ctx.notifier.create({
          content: `适配器 ${this.platformName} 加载中...`,
          type: "primary"
        });
      }
      ctx.effect(() => () => notification?.dispose());
    });
    try {
      await this._platformService.createClient(
        this.platformName,
        this.createRunnableConfig()
      );
      const content = `适配器 ${this.platformName} 加载成功，共加载了 ${this._supportModels.length} 个模型。`;
      if (notification) {
        notification.update({ content, type: "success" });
      } else {
        result = { type: "success", content };
      }
    } catch (e) {
      const content = `适配器 ${this.platformName} 加载失败: ${e.message}`;
      if (notification) {
        notification.update({ content, type: "danger" });
      } else {
        result = { type: "danger", content };
      }
      this.ctx.chatluna.uninstallPlugin(this);
      this.ctx.scope.cancel(e);
      throw e;
    }
  }
  get supportedModels() {
    return this._supportModels;
  }
  registerToService() {
    try {
      throw new Error("Please remove this method");
    } catch (e) {
      this.ctx.logger.warn(
        `Now the plugin support auto installation, Please remove call this method`,
        e
      );
    }
  }
  registerClient(func, platformName = this.platformName) {
    this.ctx.effect(
      () => this._platformService.registerClient(platformName, func)
    );
  }
  registerVectorStore(name, func) {
    this.ctx.effect(
      () => this._platformService.registerVectorStore(name, func)
    );
  }
  registerTool(name, tool) {
    this.ctx.effect(() => this._platformService.registerTool(name, tool));
  }
  registerChatChainProvider(name, description, func) {
    this.ctx.effect(
      () => this._platformService.registerChatChain(name, description, func)
    );
  }
  registerRenderer(name, renderer) {
    this.ctx.effect(
      () => this.ctx.chatluna.renderer.addRenderer(name, renderer)
    );
  }
  fetch(info, init, proxy) {
    if (proxy != null) {
      return (0, import_request.chatLunaFetch)(info, init, proxy);
    }
    const proxyMode = this.config.proxyMode;
    switch (proxyMode) {
      case "system":
        return (0, import_request.chatLunaFetch)(info, init);
      case "off":
        return (0, import_request.chatLunaFetch)(info, init, "null");
      case "on":
        return (0, import_request.chatLunaFetch)(info, init, this.config.proxyAddress);
      default:
        return (0, import_request.chatLunaFetch)(info, init);
    }
  }
  ws(url, options) {
    const proxyMode = this.config.proxyMode;
    let webSocket;
    switch (proxyMode) {
      case "system":
        webSocket = (0, import_request.ws)(url, options);
        break;
      case "off":
        webSocket = (0, import_request.ws)(url, options, "null");
        break;
      case "on":
        webSocket = (0, import_request.ws)(url, options, this.config.proxyAddress);
        break;
      default:
        webSocket = (0, import_request.ws)(url, options);
        break;
    }
    this.ctx.effect(() => webSocket.close);
    webSocket.on("error", (err) => {
      this.ctx.logger.error(err);
    });
    return webSocket;
  }
};
((ChatLunaPlugin2) => {
  ChatLunaPlugin2.Config = import_koishi5.Schema.intersect([
    import_koishi5.Schema.object({
      chatConcurrentMaxSize: import_koishi5.Schema.number().min(1).max(8).default(3),
      chatTimeLimit: import_koishi5.Schema.computed(
        import_koishi5.Schema.number().min(1).max(2e3)
      ).default(200),
      configMode: import_koishi5.Schema.union([
        import_koishi5.Schema.const("default"),
        import_koishi5.Schema.const("balance")
      ]).default("default"),
      maxRetries: import_koishi5.Schema.number().min(0).max(6).default(5),
      timeout: import_koishi5.Schema.number().default(300 * 1e3),
      proxyMode: import_koishi5.Schema.union([
        import_koishi5.Schema.const("system"),
        import_koishi5.Schema.const("off"),
        import_koishi5.Schema.const("on")
      ]).default("system")
    }),
    import_koishi5.Schema.union([
      import_koishi5.Schema.object({
        proxyMode: import_koishi5.Schema.const("on").required(),
        proxyAddress: import_koishi5.Schema.string().default("http://127.0.0.1:7897")
      }),
      import_koishi5.Schema.object({
        proxyMode: import_koishi5.Schema.const("off").required()
      }),
      import_koishi5.Schema.object({
        proxyMode: import_koishi5.Schema.const("system")
      })
    ])
  ]).i18n({
    "zh-CN": require_zh_CN_schema_plugin(),
    "en-US": require_en_US_schema_plugin()
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  });
})(ChatLunaPlugin || (ChatLunaPlugin = {}));
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AdminRequiredError,
  ChatLunaContextManagerService,
  ChatLunaPlugin,
  ChatLunaPromptRenderService,
  ChatLunaService,
  ConstraintDisabledError,
  ConstraintFixedError,
  ConstraintLockedError,
  ConversationNotFoundError,
  ConversationResolutionError,
  InvalidChatModeError,
  MessageTransformer,
  STAGE_ORDER,
  applyPresetLane,
  computeBaseBindingKey,
  getBaseBindingKey,
  getPresetLane,
  toMessages
});
