// koishi-plugin-chatluna/lib/llm-core/utils/count_tokens.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/llm-core/utils/tiktoken.ts
import path from "path";
import {
  getEncodingNameForModel,
  Tiktoken
} from "js-tiktoken/lite";
import {
  chatLunaFetch,
  globalProxyAddress
} from "koishi-plugin-chatluna/utils/request";
import os from "os";
import fs from "fs/promises";
var cache = {};
var tiktokenCache = {};
async function getEncoding(encoding, options) {
  options = options ?? {};
  const cacheDir = path.resolve(os.tmpdir(), "chatluna", "tiktoken");
  const cachePath = path.join(cacheDir, `${encoding}.json`);
  if (tiktokenCache[encoding]) {
    return tiktokenCache[encoding];
  }
  if (cache[encoding]) {
    const tiktoken2 = new Tiktoken(
      cache[encoding],
      options?.extendedSpecialTokens
    );
    tiktokenCache[encoding] = tiktoken2;
    return tiktoken2;
  }
  await fs.mkdir(cacheDir, { recursive: true });
  try {
    const cacheContent = await fs.readFile(cachePath, "utf-8");
    cache[encoding] = JSON.parse(cacheContent);
    const tiktoken2 = new Tiktoken(
      cache[encoding],
      options?.extendedSpecialTokens
    );
    tiktokenCache[encoding] = tiktoken2;
    return tiktoken2;
  } catch (e) {
  }
  const url = (globalProxyAddress?.length ?? 0) > 0 ? `https://tiktoken.pages.dev/js/${encoding}.json` : `https://jsd.onmicrosoft.cn/npm/tiktoken@latest/encoders/${encoding}.json`;
  cache[encoding] = await chatLunaFetch(url).then((res) => {
    if (!res.ok) {
      throw new Error(
        `Failed to fetch tiktoken encoding: ${res.status}`
      );
    }
    return res.json();
  }).catch((e) => {
    delete cache[encoding];
    throw e;
  });
  await fs.writeFile(cachePath, JSON.stringify(cache[encoding]));
  const tiktoken = new Tiktoken(
    cache[encoding],
    options?.extendedSpecialTokens
  );
  tiktokenCache[encoding] = tiktoken;
  return tiktoken;
}
__name(getEncoding, "getEncoding");
async function encodingForModel(model, options) {
  const result = await getEncoding(getEncodingNameForModel(model), options);
  return result;
}
__name(encodingForModel, "encodingForModel");

// src/llm-core/utils/count_tokens.ts
import { getMessageContent } from "koishi-plugin-chatluna/utils/string";
var tiktokenModels = [
  "davinci-002",
  "babbage-002",
  "text-davinci-003",
  "text-davinci-002",
  "text-davinci-001",
  "text-curie-001",
  "text-babbage-001",
  "text-ada-001",
  "davinci",
  "curie",
  "babbage",
  "ada",
  "code-davinci-002",
  "code-davinci-001",
  "code-cushman-002",
  "code-cushman-001",
  "davinci-codex",
  "cushman-codex",
  "text-davinci-edit-001",
  "code-davinci-edit-001",
  "text-embedding-ada-002",
  "text-embedding-3-small",
  "text-embedding-3-large",
  "text-similarity-davinci-001",
  "text-similarity-curie-001",
  "text-similarity-babbage-001",
  "text-similarity-ada-001",
  "text-search-davinci-doc-001",
  "text-search-curie-doc-001",
  "text-search-babbage-doc-001",
  "text-search-ada-doc-001",
  "code-search-babbage-code-001",
  "code-search-ada-code-001",
  "gpt2",
  "gpt-3.5-turbo",
  "gpt-35-turbo",
  "gpt-3.5-turbo-0301",
  "gpt-3.5-turbo-0613",
  "gpt-3.5-turbo-1106",
  "gpt-3.5-turbo-0125",
  "gpt-3.5-turbo-16k",
  "gpt-3.5-turbo-16k-0613",
  "gpt-3.5-turbo-instruct",
  "gpt-3.5-turbo-instruct-0914",
  "gpt-4",
  "gpt-4-0314",
  "gpt-4-0613",
  "gpt-4-32k",
  "gpt-4-32k-0314",
  "gpt-4-32k-0613",
  "gpt-4-turbo",
  "gpt-4-turbo-2024-04-09",
  "gpt-4-turbo-preview",
  "gpt-4-1106-preview",
  "gpt-4-0125-preview",
  "gpt-4-vision-preview",
  "gpt-4o",
  "gpt-4o-2024-05-13",
  "gpt-4o-2024-08-06",
  "gpt-4o-mini-2024-07-18",
  "gpt-4o-mini",
  "o1-mini",
  "o1-preview",
  "o1-preview-2024-09-12",
  "o1-mini-2024-09-12",
  "chatgpt-4o-latest",
  "gpt-4o-realtime",
  "gpt-4o-realtime-preview-2024-10-01"
];
var getModelNameForTiktoken = /* @__PURE__ */ __name((modelName) => {
  if (modelName.startsWith("gpt-3.5-turbo-16k")) {
    return "gpt-3.5-turbo-16k";
  }
  if (modelName.startsWith("gpt-3.5-turbo-")) {
    return "gpt-3.5-turbo";
  }
  if (modelName.startsWith("gpt-4-32k")) {
    return "gpt-4-32k";
  }
  if (modelName.startsWith("gpt-4-")) {
    return "gpt-4";
  }
  if (modelName.startsWith("gpt-4o")) {
    return "gpt-4o";
  }
  if (modelName.startsWith("o1")) {
    return "o1-mini";
  }
  if (tiktokenModels.includes(modelName)) {
    return modelName;
  }
  return "o1-mini";
}, "getModelNameForTiktoken");
var getEmbeddingContextSize = /* @__PURE__ */ __name((modelName) => {
  switch (modelName) {
    case "text-embedding-3-large":
    case "text-embedding-3-small":
    case "text-embedding-ada-002":
      return 8191;
    default:
      return 2046;
  }
}, "getEmbeddingContextSize");
var calculateMaxTokens = /* @__PURE__ */ __name(async ({
  prompt,
  modelName
}) => {
  let numTokens = Math.ceil(prompt.length / 4);
  try {
    numTokens = (await encodingForModel(modelName)).encode(prompt).length;
  } catch (error) {
  }
  const maxTokens = getModelContextSize(modelName);
  return maxTokens - numTokens;
}, "calculateMaxTokens");
function messageTypeToOpenAIRole(type) {
  switch (type) {
    case "system":
      return "system";
    case "ai":
      return "assistant";
    case "human":
      return "user";
    case "function":
      return "function";
    case "tool":
      return "tool";
    default:
      throw new Error(`Unknown message type: ${type}`);
  }
}
__name(messageTypeToOpenAIRole, "messageTypeToOpenAIRole");
var getModelContextSize = /* @__PURE__ */ __name((modelName) => {
  switch (getModelNameForTiktoken(modelName)) {
    case "gpt-4o":
    case "gpt-4o-2024-05-13":
    case "gpt-4-0125-preview":
    case "o1-mini":
      return 128e3;
    case "gpt-3.5-turbo-16k":
      return 16384;
    case "gpt-3.5-turbo":
      return 4096;
    case "gpt-4-32k":
      return 32768;
    case "gpt-4":
      return 8192;
    case "text-davinci-003":
      return 4097;
    case "text-curie-001":
      return 2048;
    case "text-babbage-001":
      return 2048;
    case "text-ada-001":
      return 2048;
    case "code-davinci-002":
      return 8e3;
    case "code-cushman-001":
      return 2048;
    default:
      return 4097;
  }
}, "getModelContextSize");
function parseRawModelName(modelName) {
  if (modelName == null || modelName.trim().length < 1) {
    return [void 0, void 0];
  }
  const value = modelName.trim();
  const index = value.indexOf("/");
  if (index === -1) {
    return [void 0, value];
  }
  if (index === 0 || index === value.length - 1) {
    return [void 0, void 0];
  }
  return [value.slice(0, index), value.slice(index + 1)];
}
__name(parseRawModelName, "parseRawModelName");
async function countMessageTokens(message, tokenCounter) {
  let content = getMessageContent(message.content);
  if (content.includes("![image]") && content.includes("base64") && message.additional_kwargs?.["images"]) {
    content = content.replaceAll(/!\[.*?\]\(.*?\)/g, "");
  }
  let tokens = await tokenCounter(content) + await tokenCounter(messageTypeToOpenAIRole(message.getType())) + (message.name ? await tokenCounter(message.name) : 0);
  const toolCalls = message.tool_calls ?? message.additional_kwargs?.tool_calls;
  if (Array.isArray(toolCalls) && toolCalls.length > 0) {
    tokens += await tokenCounter(JSON.stringify(toolCalls));
  }
  return tokens;
}
__name(countMessageTokens, "countMessageTokens");
async function countMessagesTokens(messages, tokenCounter, presetTokens = 0) {
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i].getType() !== "ai") continue;
    const usage = messages[i].usage_metadata;
    if (usage?.input_tokens > 0) {
      let tail = 0;
      for (let j = i; j < messages.length; j++) {
        tail += await countMessageTokens(messages[j], tokenCounter);
      }
      return Math.max(usage.input_tokens - presetTokens + tail, 0);
    }
  }
  let total = 0;
  for (const msg of messages) {
    total += await countMessageTokens(msg, tokenCounter);
  }
  return total;
}
__name(countMessagesTokens, "countMessagesTokens");
export {
  calculateMaxTokens,
  countMessageTokens,
  countMessagesTokens,
  getEmbeddingContextSize,
  getModelContextSize,
  getModelNameForTiktoken,
  messageTypeToOpenAIRole,
  parseRawModelName
};
