// koishi-plugin-chatluna/lib/llm-core/platform/api.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/llm-core/platform/api.ts
var api_exports = {};
__export(api_exports, {
  MIN_LATENCY_MS: () => MIN_LATENCY_MS,
  ModelRequester: () => ModelRequester,
  attachInvocationMetrics: () => attachInvocationMetrics,
  createModelUsageTiming: () => createModelUsageTiming,
  readInvocationMetrics: () => readInvocationMetrics
});
module.exports = __toCommonJS(api_exports);
var import_messages = require("@langchain/core/messages");
var import_outputs = require("@langchain/core/outputs");
var import_error = require("koishi-plugin-chatluna/utils/error");

// src/llm-core/platform/metrics.ts
var chatlunaMetricsKey = "chatluna_invocation_metrics";
function attachInvocationMetrics(chunk, metrics) {
  chunk.generationInfo = {
    ...chunk.generationInfo,
    [chatlunaMetricsKey]: metrics
  };
  const message = chunk.message;
  if (message != null) {
    message.response_metadata = {
      ...message.response_metadata,
      [chatlunaMetricsKey]: metrics
    };
  }
}
__name(attachInvocationMetrics, "attachInvocationMetrics");
function readInvocationMetrics(chunk) {
  const message = chunk?.message;
  const info = chunk?.generationInfo;
  const metadata = message?.response_metadata;
  const metrics = info?.[chatlunaMetricsKey] ?? metadata?.[chatlunaMetricsKey];
  return {
    usageMetadata: metrics?.usageMetadata ?? message?.usage_metadata,
    timing: metrics?.timing
  };
}
__name(readInvocationMetrics, "readInvocationMetrics");
var MIN_LATENCY_MS = 10;
function createModelUsageTiming(start, firstAt, usage) {
  const totalMs = Math.max(Date.now() - start, MIN_LATENCY_MS);
  const outputTokens = usage?.output_tokens + (usage?.output_token_details?.reasoning ?? 0);
  if (firstAt == null) {
    return {
      totalMs,
      tps: outputTokens == null ? void 0 : outputTokens * 1e3 / totalMs
    };
  }
  const ttftMs = Math.max(firstAt - start, MIN_LATENCY_MS);
  return {
    ttftMs,
    totalMs,
    tps: outputTokens == null ? void 0 : outputTokens * 1e3 / totalMs
  };
}
__name(createModelUsageTiming, "createModelUsageTiming");
function hasResponseChunk(chunk) {
  const message = chunk.message;
  const content = message?.content;
  const kwargs = message?.additional_kwargs;
  return chunk.text.length > 0 || (typeof content === "string" ? content.trim().length > 0 : Array.isArray(content) && content.length > 0) || (message?.tool_call_chunks?.length ?? 0) > 0 || (message?.tool_calls?.length ?? 0) > 0 || (message?.invalid_tool_calls?.length ?? 0) > 0 || (kwargs?.tool_calls?.length ?? 0) > 0 || kwargs?.function_call != null || kwargs?.thought_data != null;
}
__name(hasResponseChunk, "hasResponseChunk");
var StreamMetricsTracker = class {
  static {
    __name(this, "StreamMetricsTracker");
  }
  start = Date.now();
  firstAt;
  usage;
  observe(chunk) {
    const message = chunk.message;
    if (message?.usage_metadata != null) {
      this.usage = message.usage_metadata;
    }
    if (this.firstAt == null && hasResponseChunk(chunk)) {
      this.firstAt = Date.now();
    }
  }
  attachTo(chunk) {
    attachInvocationMetrics(chunk, {
      usageMetadata: this.usage,
      timing: createModelUsageTiming(this.start, this.firstAt, this.usage)
    });
    return chunk;
  }
};

// src/llm-core/platform/api.ts
var ModelRequester = class {
  constructor(ctx, _configPool, _pluginConfig, _plugin) {
    this.ctx = ctx;
    this._configPool = _configPool;
    this._pluginConfig = _pluginConfig;
    this._plugin = _plugin;
  }
  static {
    __name(this, "ModelRequester");
  }
  _errorCounts = {};
  async completion(params) {
    const stream = this.completionStream(params);
    let result;
    for await (const chunk of stream) {
      result = result != null ? result.concat(chunk) : chunk;
    }
    return result;
  }
  async *completionStream(params) {
    this._configPool.getConfig(false);
    const config = this._config;
    const tracker = new StreamMetricsTracker();
    try {
      for await (const chunk of this.completionStreamInternal(params)) {
        tracker.observe(chunk);
        yield chunk;
      }
      yield tracker.attachTo(
        new import_outputs.ChatGenerationChunk({
          message: new import_messages.AIMessageChunk({ content: "" }),
          text: ""
        })
      );
    } catch (e) {
      if (e instanceof import_error.ChatLunaError && (e.errorCode === import_error.ChatLunaErrorCode.NETWORK_ERROR || e.errorCode === import_error.ChatLunaErrorCode.API_REQUEST_TIMEOUT || e.errorCode === import_error.ChatLunaErrorCode.ABORTED || e.errorCode === import_error.ChatLunaErrorCode.API_UNSAFE_CONTENT) || e.name === "AbortError") {
        throw e;
      }
      this._errorCounts[config.md5()] = (this._errorCounts[config.md5()] || 0) + 1;
      if (this._errorCounts[config.md5()] > this._pluginConfig.maxRetries) {
        this._configPool.markConfigStatus(config.value, false);
        delete this._errorCounts[config.md5()];
      }
      throw e;
    }
  }
  async init() {
  }
  async dispose(model, id) {
  }
  post(url, data, params = {}) {
    const requestUrl = this.concatUrl(url);
    for (const key in data) {
      if (data[key] === void 0) {
        delete data[key];
      }
    }
    const body = JSON.stringify(data);
    const { headers: initHeaders, method: _m, ...rest } = params;
    return this._plugin.fetch(requestUrl, {
      ...rest,
      method: "POST",
      headers: {
        ...this.buildHeaders(),
        ...initHeaders
      },
      body
    });
  }
  get(url, headers, params = {}) {
    const requestUrl = this.concatUrl(url);
    const { headers: initHeaders, method: _m, ...rest } = params;
    return this._plugin.fetch(requestUrl, {
      ...rest,
      method: "GET",
      headers: {
        ...this.buildHeaders(),
        ...initHeaders,
        ...headers
      }
    });
  }
  get _config() {
    return this._configPool.getConfig(true);
  }
  buildHeaders() {
    return {
      Authorization: `Bearer ${this._config.value.apiKey}`,
      "Content-Type": "application/json",
      "HTTP-Referer": "https://github.com/ChatLunaLab/chatluna",
      // Optional. Site URL for rankings on openrouter.ai.
      "X-Title": "ChatLuna"
      // Optional. Site title for rankings on openrouter.ai.
    };
  }
  concatUrl(url) {
    const apiEndPoint = this._config.value.apiEndpoint;
    if (apiEndPoint.endsWith("/")) {
      return apiEndPoint + url;
    }
    return apiEndPoint + "/" + url;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MIN_LATENCY_MS,
  ModelRequester,
  attachInvocationMetrics,
  createModelUsageTiming,
  readInvocationMetrics
});
