// koishi-plugin-chatluna/lib/llm-core/platform/client.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/llm-core/platform/client.ts
import { sleep } from "koishi";
import {
  ModelCapabilities
} from "koishi-plugin-chatluna/llm-core/platform/types";
import { ObjectLock } from "koishi-plugin-chatluna/utils/lock";
import {
  ChatLunaError,
  ChatLunaErrorCode
} from "koishi-plugin-chatluna/utils/error";
import {
  createModelUsageReporter
} from "koishi-plugin-chatluna/llm-core/platform/usage";
import { usageSourceFromStack } from "koishi-plugin-chatluna/utils/usage_source";
var BasePlatformClient = class {
  constructor(ctx, configPool) {
    this.ctx = ctx;
    this.configPool = configPool;
  }
  static {
    __name(this, "BasePlatformClient");
  }
  _modelPool = {};
  _reports = {};
  _modelInfos = {};
  _lock = new ObjectLock();
  async isAvailable(config) {
    if (Object.values(this._modelInfos).length > 0) {
      return true;
    }
    const unlock = await this._lock.lock();
    let retryCount = 0;
    const cfg = this.configPool.findAvailableConfig() ?? this.configPool.getConfigs()[0];
    if (cfg == null) {
      unlock();
      return false;
    }
    const maxRetries = cfg.value.maxRetries ?? 5;
    while (retryCount <= maxRetries) {
      let oldConfig;
      try {
        oldConfig = this.configPool.getConfig(true);
        await this.init(config);
        unlock();
        return true;
      } catch (e) {
        if (e instanceof ChatLunaError && e.errorCode === ChatLunaErrorCode.ABORTED) {
          unlock();
          throw e;
        }
        if (e instanceof ChatLunaError && e.errorCode === ChatLunaErrorCode.NOT_AVAILABLE_CONFIG) {
          unlock();
          return false;
        }
        if (retryCount >= maxRetries) {
          if (oldConfig == null) {
            this.ctx.logger.error(e);
            unlock();
            return false;
          }
          this.configPool.getConfig(false);
          this.configPool.markConfigStatus(oldConfig.value, false);
          this.ctx.logger.error(e);
          if (this.configPool.findAvailableConfig() != null) {
            retryCount = 0;
            continue;
          }
          unlock();
          return false;
        }
      }
      await sleep(1e3 * 2 ** retryCount);
      retryCount++;
    }
    unlock();
    return false;
  }
  get config() {
    return this.configPool.getConfig(true).value;
  }
  async getModels(config) {
    let models = Object.values(this._modelInfos);
    if (models.length > 0) {
      return models;
    }
    try {
      models = await this.refreshModels(config);
      this._modelInfos = {};
      for (const model of models) {
        model.capabilities = model.capabilities.includes(
          ModelCapabilities.ImageGeneration
        ) ? [ModelCapabilities.ImageGeneration] : Array.from(
          /* @__PURE__ */ new Set([
            ...model.capabilities,
            ModelCapabilities.TextInput
          ])
        );
        this._modelInfos[model.name] = model;
      }
      return models;
    } catch (e) {
      if (e instanceof ChatLunaError && e.errorCode === ChatLunaErrorCode.ABORTED) {
        throw e;
      }
      const err = e instanceof Error ? e : new Error(String(e));
      const cause = err instanceof ChatLunaError ? err.originError : null;
      const reason = cause?.message ?? err.message;
      err.message = `获取模型列表失败 (${this.platform}): ${reason}`;
      this.ctx.logger.error(err);
      this._modelInfos = {};
      throw err;
    }
  }
  async init(config) {
    await this.getModels(config);
  }
  /**
   * Returns file handling configuration for this platform, or `null` if the
   * platform does not support inline file uploads beyond basic image input.
   *
   * Override in subclasses to provide platform-specific MIME types, size
   * limits, and inline data support.
   */
  getFileHandlingConfig() {
    return null;
  }
  createModel(model) {
    const limit = Error.stackTraceLimit;
    Error.stackTraceLimit = Math.max(limit, 50);
    const stack = new Error().stack;
    Error.stackTraceLimit = limit;
    const source = usageSourceFromStack(stack);
    const key = `${source}:${model}`;
    const report = this._reports[key] ?? createModelUsageReporter(this.ctx, this.platform, model, stack);
    this._reports[key] = report;
    if (!this._modelPool[key]) {
      this._modelPool[key] = this._createModel(model, report);
    }
    return this._modelPool[key];
  }
};
var PlatformModelClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformModelClient");
  }
  async clearContext() {
  }
};
var PlatformEmbeddingsClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformEmbeddingsClient");
  }
};
var PlatformModelAndEmbeddingsClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformModelAndEmbeddingsClient");
  }
};
var PlatformRerankerClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformRerankerClient");
  }
};
var PlatformModelEmbeddingsAndRerankerClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformModelEmbeddingsAndRerankerClient");
  }
};
export {
  BasePlatformClient,
  PlatformEmbeddingsClient,
  PlatformModelAndEmbeddingsClient,
  PlatformModelClient,
  PlatformModelEmbeddingsAndRerankerClient,
  PlatformRerankerClient
};
