// koishi-plugin-chatluna/lib/llm-core/platform/client.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/llm-core/platform/client.ts
var client_exports = {};
__export(client_exports, {
  BasePlatformClient: () => BasePlatformClient,
  PlatformEmbeddingsClient: () => PlatformEmbeddingsClient,
  PlatformModelAndEmbeddingsClient: () => PlatformModelAndEmbeddingsClient,
  PlatformModelClient: () => PlatformModelClient,
  PlatformModelEmbeddingsAndRerankerClient: () => PlatformModelEmbeddingsAndRerankerClient,
  PlatformRerankerClient: () => PlatformRerankerClient
});
module.exports = __toCommonJS(client_exports);
var import_koishi = require("koishi");
var import_types = require("koishi-plugin-chatluna/llm-core/platform/types");
var import_lock = require("koishi-plugin-chatluna/utils/lock");
var import_error = require("koishi-plugin-chatluna/utils/error");
var import_usage = require("koishi-plugin-chatluna/llm-core/platform/usage");
var import_usage_source = require("koishi-plugin-chatluna/utils/usage_source");
var BasePlatformClient = class {
  constructor(ctx, configPool) {
    this.ctx = ctx;
    this.configPool = configPool;
  }
  static {
    __name(this, "BasePlatformClient");
  }
  _modelPool = {};
  _reports = {};
  _modelInfos = {};
  _lock = new import_lock.ObjectLock();
  async isAvailable(config) {
    if (Object.values(this._modelInfos).length > 0) {
      return true;
    }
    const unlock = await this._lock.lock();
    let retryCount = 0;
    const cfg = this.configPool.findAvailableConfig() ?? this.configPool.getConfigs()[0];
    if (cfg == null) {
      unlock();
      return false;
    }
    const maxRetries = cfg.value.maxRetries ?? 5;
    while (retryCount <= maxRetries) {
      let oldConfig;
      try {
        oldConfig = this.configPool.getConfig(true);
        await this.init(config);
        unlock();
        return true;
      } catch (e) {
        if (e instanceof import_error.ChatLunaError && e.errorCode === import_error.ChatLunaErrorCode.ABORTED) {
          unlock();
          throw e;
        }
        if (e instanceof import_error.ChatLunaError && e.errorCode === import_error.ChatLunaErrorCode.NOT_AVAILABLE_CONFIG) {
          unlock();
          return false;
        }
        if (retryCount >= maxRetries) {
          if (oldConfig == null) {
            this.ctx.logger.error(e);
            unlock();
            return false;
          }
          this.configPool.getConfig(false);
          this.configPool.markConfigStatus(oldConfig.value, false);
          this.ctx.logger.error(e);
          if (this.configPool.findAvailableConfig() != null) {
            retryCount = 0;
            continue;
          }
          unlock();
          return false;
        }
      }
      await (0, import_koishi.sleep)(1e3 * 2 ** retryCount);
      retryCount++;
    }
    unlock();
    return false;
  }
  get config() {
    return this.configPool.getConfig(true).value;
  }
  async getModels(config) {
    let models = Object.values(this._modelInfos);
    if (models.length > 0) {
      return models;
    }
    try {
      models = await this.refreshModels(config);
      this._modelInfos = {};
      for (const model of models) {
        model.capabilities = model.capabilities.includes(
          import_types.ModelCapabilities.ImageGeneration
        ) ? [import_types.ModelCapabilities.ImageGeneration] : Array.from(
          /* @__PURE__ */ new Set([
            ...model.capabilities,
            import_types.ModelCapabilities.TextInput
          ])
        );
        this._modelInfos[model.name] = model;
      }
      return models;
    } catch (e) {
      if (e instanceof import_error.ChatLunaError && e.errorCode === import_error.ChatLunaErrorCode.ABORTED) {
        throw e;
      }
      const err = e instanceof Error ? e : new Error(String(e));
      const cause = err instanceof import_error.ChatLunaError ? err.originError : null;
      const reason = cause?.message ?? err.message;
      err.message = `获取模型列表失败 (${this.platform}): ${reason}`;
      this.ctx.logger.error(err);
      this._modelInfos = {};
      throw err;
    }
  }
  async init(config) {
    await this.getModels(config);
  }
  /**
   * Returns file handling configuration for this platform, or `null` if the
   * platform does not support inline file uploads beyond basic image input.
   *
   * Override in subclasses to provide platform-specific MIME types, size
   * limits, and inline data support.
   */
  getFileHandlingConfig() {
    return null;
  }
  createModel(model) {
    const limit = Error.stackTraceLimit;
    Error.stackTraceLimit = Math.max(limit, 50);
    const stack = new Error().stack;
    Error.stackTraceLimit = limit;
    const source = (0, import_usage_source.usageSourceFromStack)(stack);
    const key = `${source}:${model}`;
    const report = this._reports[key] ?? (0, import_usage.createModelUsageReporter)(this.ctx, this.platform, model, stack);
    this._reports[key] = report;
    if (!this._modelPool[key]) {
      this._modelPool[key] = this._createModel(model, report);
    }
    return this._modelPool[key];
  }
};
var PlatformModelClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformModelClient");
  }
  async clearContext() {
  }
};
var PlatformEmbeddingsClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformEmbeddingsClient");
  }
};
var PlatformModelAndEmbeddingsClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformModelAndEmbeddingsClient");
  }
};
var PlatformRerankerClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformRerankerClient");
  }
};
var PlatformModelEmbeddingsAndRerankerClient = class extends BasePlatformClient {
  static {
    __name(this, "PlatformModelEmbeddingsAndRerankerClient");
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BasePlatformClient,
  PlatformEmbeddingsClient,
  PlatformModelAndEmbeddingsClient,
  PlatformModelClient,
  PlatformModelEmbeddingsAndRerankerClient,
  PlatformRerankerClient
});
