// koishi-plugin-chatluna/lib/llm-core/platform/rerank.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/llm-core/platform/rerank.ts
import { BaseDocumentCompressor } from "@langchain/core/retrievers/document_compressors";
import {
  AsyncCaller
} from "@langchain/core/utils/async_caller";
import {
  ChatLunaError,
  ChatLunaErrorCode
} from "koishi-plugin-chatluna/utils/error";
import { logger } from "koishi-plugin-chatluna";
import { estimateTextTokens } from "koishi-plugin-chatluna/llm-core/platform/usage";
var ChatLunaReranker = class extends BaseDocumentCompressor {
  static {
    __name(this, "ChatLunaReranker");
  }
  modelName = "bge-reranker-v2-m3";
  // topN limits direct rerank() calls; compressDocuments reranks all inputs.
  topN = 3;
  maxChunksPerDoc;
  timeout;
  caller;
  _client;
  _report;
  constructor(fields) {
    super();
    this.caller = new AsyncCaller(fields);
    this.timeout = fields.timeout ?? 1e3 * 60;
    this.modelName = fields.model ?? this.modelName;
    this.topN = fields.topN ?? this.topN;
    this.maxChunksPerDoc = fields.maxChunksPerDoc;
    this._client = fields.client;
    this._report = fields.usageReporter;
  }
  async compressDocuments(documents, query) {
    if (documents == null || documents.length === 0) {
      return [];
    }
    const results = await this.rerank(documents, query, {
      topN: documents.length
    });
    return results.map((result) => {
      const original = documents[result.index];
      return {
        ...original,
        metadata: {
          ...original.metadata,
          relevanceScore: result.relevanceScore
        }
      };
    });
  }
  async rerank(documents, query, options) {
    const docStrings = documents.map(
      (doc) => typeof doc === "string" ? doc : "pageContent" in doc ? doc.pageContent : JSON.stringify(doc)
    );
    const request = {
      model: options?.model ?? this.modelName,
      query,
      documents: docStrings,
      topN: options?.topN ?? this.topN,
      maxChunksPerDoc: options?.maxChunksPerDoc ?? this.maxChunksPerDoc
    };
    let data;
    try {
      data = await this._rerankWithRetry(request);
    } catch (e) {
      await this._reportFailedUsage();
      throw e;
    }
    const results = Array.isArray(data) ? data : data.results;
    await this._reportUsage(
      [request.query, ...request.documents],
      Array.isArray(data) ? void 0 : data.usage
    );
    return results.map((result) => ({
      index: result.index,
      relevanceScore: result.relevanceScore
    }));
  }
  async _rerankWithRetry(request) {
    const timeoutError = new ChatLunaError(
      ChatLunaErrorCode.API_REQUEST_TIMEOUT,
      new Error(`timeout when calling ${request.model} reranker`),
      true
    );
    const makeRequest = /* @__PURE__ */ __name(async () => {
      let timeoutId;
      const timeoutPromise = new Promise(
        // eslint-disable-next-line promise/param-names
        (_, reject) => {
          timeoutId = setTimeout(() => {
            reject(timeoutError);
          }, this.timeout);
        }
      );
      try {
        return await Promise.race([
          this._client.rerank(request),
          timeoutPromise
        ]);
      } catch (e) {
        if (e instanceof ChatLunaError) {
          throw e;
        }
        throw new ChatLunaError(ChatLunaErrorCode.API_REQUEST_FAILED, e);
      } finally {
        clearTimeout(timeoutId);
      }
    }, "makeRequest");
    try {
      return await this.caller.call(makeRequest);
    } catch (e) {
      if (e instanceof ChatLunaError) {
        throw e;
      }
      throw new ChatLunaError(ChatLunaErrorCode.API_REQUEST_FAILED, e);
    }
  }
  async _reportUsage(input, usage) {
    if (this._report == null) return;
    try {
      const estimated = usage?.input_tokens == null && usage?.total_tokens == null;
      const inputTokens = usage?.input_tokens ?? usage?.total_tokens ?? await estimateTextTokens(input);
      const outputTokens = usage?.output_tokens ?? 0;
      await this._report({
        callType: "reranker",
        usageMetadata: {
          ...usage,
          input_tokens: usage?.input_tokens ?? inputTokens,
          output_tokens: outputTokens,
          total_tokens: usage?.total_tokens ?? (usage?.input_tokens ?? inputTokens) + outputTokens
        },
        estimated,
        success: true
      });
    } catch (e) {
      logger.warn("Failed to report reranker usage", e);
    }
  }
  async _reportFailedUsage() {
    if (this._report == null) return;
    try {
      await this._report({
        callType: "reranker",
        usageMetadata: {
          input_tokens: 0,
          output_tokens: 0,
          total_tokens: 0
        },
        estimated: false,
        success: false
      });
    } catch (e) {
      logger.warn("Failed to report reranker usage", e);
    }
  }
};
export {
  ChatLunaReranker
};
