// koishi-plugin-chatluna/lib/llm-core/platform/service.cjs
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/llm-core/platform/service.ts
var service_exports = {};
__export(service_exports, {
  PlatformService: () => PlatformService
});
module.exports = __toCommonJS(service_exports);
var import_client = require("koishi-plugin-chatluna/llm-core/platform/client");
var import_types = require("koishi-plugin-chatluna/llm-core/platform/types");
var import_lru_cache = require("lru-cache");
var import_count_tokens = require("koishi-plugin-chatluna/llm-core/utils/count_tokens");
var import_reactivity = require("@vue/reactivity");
var import_crypto = require("crypto");
var PlatformService = class {
  constructor(ctx) {
    this.ctx = ctx;
    const clear = /* @__PURE__ */ __name(async (payload) => {
      this._tmpVectorStores.delete(payload.conversation.id);
    }, "clear");
    this.ctx.on("chatluna/after-conversation-clear-history", clear);
    this.ctx.on("chatluna/after-conversation-cache-clear", clear);
    this.ctx.on("chatluna/after-conversation-archive", clear);
    this.ctx.on("chatluna/after-conversation-restore", clear);
    this.ctx.on("chatluna/after-conversation-delete", clear);
    this.ctx.on("chatluna/conversation-compressed", clear);
  }
  static {
    __name(this, "PlatformService");
  }
  _platformClients = (0, import_reactivity.reactive)({});
  _createClientFunctions = (0, import_reactivity.reactive)({});
  _tools = (0, import_reactivity.reactive)({});
  _tmpTools = {};
  _toolMaskResolvers = {};
  _models = (0, import_reactivity.reactive)({});
  _chatChains = (0, import_reactivity.reactive)({});
  _vectorStore = (0, import_reactivity.reactive)(
    {}
  );
  _tmpVectorStores = new import_lru_cache.LRUCache({
    max: 20,
    dispose: /* @__PURE__ */ __name((value) => {
      value.free?.();
    }, "dispose")
  });
  registerClient(name, createClientFunction) {
    if (this._createClientFunctions[name]) {
      throw new Error(`Client ${name} already exists`);
    }
    this._createClientFunctions[name] = createClientFunction;
    return () => this.unregisterClient(name);
  }
  registerTool(name, toolCreator) {
    toolCreator.id = (0, import_crypto.randomUUID)();
    toolCreator.name = name;
    this._tools[name] = (0, import_reactivity.markRaw)(toolCreator);
    delete this._tmpTools[name];
    this.ctx.emit("chatluna/tool-updated", this);
    return () => this.unregisterTool(name);
  }
  unregisterTool(name) {
    delete this._tools[name];
    delete this._tmpTools[name];
    this.ctx.emit("chatluna/tool-updated", this);
  }
  unregisterClient(platform) {
    delete this._models[platform];
    const client = this._platformClients[platform];
    if (client == null) {
      delete this._createClientFunctions[platform];
      return;
    }
    delete this._platformClients[platform];
    if (client instanceof import_client.PlatformModelEmbeddingsAndRerankerClient) {
      this.ctx.emit("chatluna/embeddings-removed", this, platform, client);
      this.ctx.emit("chatluna/model-removed", this, platform, client);
      this.ctx.emit("chatluna/reranker-removed", this, platform, client);
    } else if (client instanceof import_client.PlatformModelAndEmbeddingsClient) {
      this.ctx.emit("chatluna/embeddings-removed", this, platform, client);
      this.ctx.emit("chatluna/model-removed", this, platform, client);
    } else if (client instanceof import_client.PlatformModelClient) {
      this.ctx.emit("chatluna/model-removed", this, platform, client);
    } else if (client instanceof import_client.PlatformEmbeddingsClient) {
      this.ctx.emit("chatluna/embeddings-removed", this, platform, client);
    } else if (client instanceof import_client.PlatformRerankerClient) {
      this.ctx.emit("chatluna/reranker-removed", this, platform, client);
    }
    delete this._createClientFunctions[platform];
  }
  unregisterVectorStore(name) {
    delete this._vectorStore[name];
    this.ctx.emit("chatluna/vector-store-removed", this, name);
  }
  registerVectorStore(name, vectorStoreRetrieverCreator) {
    this._vectorStore[name] = vectorStoreRetrieverCreator;
    this.ctx.emit("chatluna/vector-store-added", this, name);
    return () => this.unregisterVectorStore(name);
  }
  registerChatChain(name, description, createChatChainFunction) {
    this._chatChains[name] = {
      name,
      description,
      createFunction: createChatChainFunction
    };
    this.ctx.emit("chatluna/chat-chain-added", this, this._chatChains[name]);
    return () => this.unregisterChatChain(name);
  }
  unregisterChatChain(name) {
    const chain = this._chatChains[name];
    delete this._chatChains[name];
    this.ctx.emit("chatluna/chat-chain-removed", this, chain);
  }
  listPlatformModels(platform, type) {
    return (0, import_reactivity.computed)(() => {
      const models = this._models[platform] ?? [];
      if (models.length === 0) {
        return [];
      }
      return models.filter((m) => type === import_types.ModelType.all || m.type === type).sort((a, b) => {
        if (!a?.name || !b?.name) return 0;
        return a.name.localeCompare(b.name, void 0, {
          numeric: true,
          sensitivity: "base"
        });
      });
    });
  }
  findModel(platform, name) {
    if (name == null) {
      ;
      [platform, name] = (0, import_count_tokens.parseRawModelName)(platform);
    }
    return (0, import_reactivity.computed)(
      () => this._models[platform]?.find((m) => m.name === name) ?? null
    );
  }
  getTools() {
    return (0, import_reactivity.computed)(() => Object.keys(this._tools));
  }
  getToolRegistry() {
    return Object.fromEntries(
      Object.entries(this._tools).map(([name, tool]) => [
        name,
        {
          name,
          description: tool.description,
          meta: tool.meta
        }
      ])
    );
  }
  getFilteredTools(mask) {
    const allNames = Object.keys(this._tools);
    const names = mask.tools ? allNames.filter((name) => mask.tools.includes(name)) : allNames;
    if (mask.mode === "all") {
      return names;
    }
    if (mask.mode === "allow") {
      return names.filter((name) => mask.allow.includes(name));
    }
    return names.filter((name) => !mask.deny.includes(name));
  }
  registerToolMaskResolver(name, resolver) {
    this._toolMaskResolvers[name] = resolver;
    return () => {
      delete this._toolMaskResolvers[name];
    };
  }
  async resolveToolMask(arg) {
    for (const name in this._toolMaskResolvers) {
      const mask = await this._toolMaskResolvers[name](arg);
      if (mask) {
        return mask;
      }
    }
  }
  static buildToolMask(rule) {
    if (rule.mode === "allow") {
      return {
        mode: "allow",
        allow: rule.allow ?? [],
        deny: []
      };
    }
    if (rule.mode === "deny") {
      return {
        mode: "deny",
        allow: [],
        deny: rule.deny ?? []
      };
    }
    return {
      mode: "all",
      allow: [],
      deny: []
    };
  }
  listAllModels(type) {
    return (0, import_reactivity.computed)(() => {
      const allModel = [];
      for (const platform in this._models) {
        const models = this._models[platform];
        for (const model of models) {
          if (type === import_types.ModelType.all || model.type === type) {
            allModel.push({
              ...model,
              platform,
              toModelName: /* @__PURE__ */ __name(() => platform + "/" + model.name, "toModelName")
            });
          }
        }
      }
      return allModel.sort();
    });
  }
  get vectorStores() {
    return (0, import_reactivity.computed)(() => Object.keys(this._vectorStore));
  }
  get chatChains() {
    return (0, import_reactivity.computed)(() => Object.values(this._chatChains));
  }
  async createVectorStore(name, params) {
    const vectorStoreRetriever = this._vectorStore[name];
    if (!vectorStoreRetriever) {
      throw new Error(`Vector store retriever ${name} not found`);
    }
    const key = params.key;
    if (key == null) {
      return await vectorStoreRetriever(params);
    }
    const cacheVectorStore = this._tmpVectorStores.get(key);
    if (cacheVectorStore) {
      return cacheVectorStore;
    }
    const vectorStore = await vectorStoreRetriever(params);
    this._tmpVectorStores.set(key, vectorStore);
    return vectorStore;
  }
  async getClient(platform) {
    if (!this._platformClients[platform]) {
      await this.createClient(platform);
    }
    return (0, import_reactivity.computed)(() => this._platformClients[platform]);
  }
  async refreshClient(client, platform, config) {
    const isAvailable = await client.isAvailable(config);
    if (!isAvailable) {
      return;
    }
    const models = await client.getModels(config);
    if (models == null) {
      return;
    }
    const availableModels = this._models[platform] ?? [];
    this._models[platform] = availableModels.concat(
      models.filter(
        (m) => !availableModels.some((am) => am.name === m.name)
      )
    );
    if (client instanceof import_client.PlatformModelEmbeddingsAndRerankerClient) {
      this.ctx.emit("chatluna/embeddings-added", this, platform, client);
      this.ctx.emit("chatluna/model-added", this, platform, client);
      this.ctx.emit("chatluna/reranker-added", this, platform, client);
    } else if (client instanceof import_client.PlatformModelAndEmbeddingsClient) {
      this.ctx.emit("chatluna/embeddings-added", this, platform, client);
      this.ctx.emit("chatluna/model-added", this, platform, client);
    } else if (client instanceof import_client.PlatformModelClient) {
      this.ctx.emit("chatluna/model-added", this, platform, client);
    } else if (client instanceof import_client.PlatformEmbeddingsClient) {
      this.ctx.emit("chatluna/embeddings-added", this, platform, client);
    } else if (client instanceof import_client.PlatformRerankerClient) {
      this.ctx.emit("chatluna/reranker-added", this, platform, client);
    }
  }
  async createClient(platform, config) {
    const createClientFunction = this._createClientFunctions[platform];
    if (!createClientFunction) {
      return;
    }
    if (this._platformClients[platform]) {
      this.ctx.logger.warn(
        `Client ${platform} already exists, skip creating`
      );
      return this._platformClients[platform];
    }
    const client = createClientFunction();
    await this.refreshClient(client, platform, config);
    this._platformClients[platform] = (0, import_reactivity.markRaw)(client);
    return client;
  }
  getTool(name) {
    const tool = this._tools[name];
    const that = this;
    return {
      ...tool,
      createTool(params) {
        return that._createTool(name, params);
      }
    };
  }
  _createTool(name, params) {
    if (this._tmpTools[name]) {
      return this._tmpTools[name];
    }
    const chatLunaTool = this._tools[name];
    if (chatLunaTool == null) {
      throw new Error(`Tool ${name} not found`);
    }
    const tool = chatLunaTool.createTool(params);
    this._tmpTools[name] = (0, import_reactivity.markRaw)(tool);
    return tool;
  }
  createChatChain(name, params) {
    const chatChain = this._chatChains[name];
    if (!chatChain) {
      throw new Error(`Chat chain ${name} not found`);
    }
    return chatChain.createFunction(params);
  }
  dispose() {
    this._tmpVectorStores.clear();
    this._platformClients = (0, import_reactivity.reactive)({});
    this._models = (0, import_reactivity.reactive)({});
    this._tools = (0, import_reactivity.reactive)({});
    this._tmpTools = {};
    this._chatChains = (0, import_reactivity.reactive)({});
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PlatformService
});
