// koishi-plugin-chatluna/lib/llm-core/platform/types.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/llm-core/platform/types.ts
var ModelCapabilities = /* @__PURE__ */ ((ModelCapabilities2) => {
  ModelCapabilities2["TextInput"] = "text_input";
  ModelCapabilities2["ToolCall"] = "tool_call";
  ModelCapabilities2["ImageInput"] = "image_input";
  ModelCapabilities2["Thinking"] = "thinking";
  ModelCapabilities2["ImageGeneration"] = "image_generation";
  ModelCapabilities2["AudioInput"] = "audio_input";
  ModelCapabilities2["VideoInput"] = "video_input";
  ModelCapabilities2["FileInput"] = "file_input";
  return ModelCapabilities2;
})(ModelCapabilities || {});
var ModelType = /* @__PURE__ */ ((ModelType2) => {
  ModelType2[ModelType2["all"] = 0] = "all";
  ModelType2[ModelType2["llm"] = 1] = "llm";
  ModelType2[ModelType2["embeddings"] = 2] = "embeddings";
  ModelType2[ModelType2["reranker"] = 3] = "reranker";
  return ModelType2;
})(ModelType || {});
function isModelInfo(model) {
  return model.name != null && model.type != null && model.maxTokens != null;
}
__name(isModelInfo, "isModelInfo");
function isPlatformModelInfo(model) {
  return isModelInfo(model) && model["platform"] != null && model["platform"] !== "default";
}
__name(isPlatformModelInfo, "isPlatformModelInfo");
export {
  ModelCapabilities,
  ModelType,
  isModelInfo,
  isPlatformModelInfo
};
