// koishi-plugin-chatluna/lib/llm-core/platform/usage.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/llm-core/platform/usage.ts
var usage_exports = {};
__export(usage_exports, {
  createModelUsageReporter: () => createModelUsageReporter,
  estimateTextTokens: () => estimateTextTokens,
  warmupTokenEncoder: () => warmupTokenEncoder
});
module.exports = __toCommonJS(usage_exports);
var import_koishi_plugin_chatluna = require("koishi-plugin-chatluna");
var import_usage_source = require("koishi-plugin-chatluna/utils/usage_source");

// src/llm-core/utils/tiktoken.ts
var import_path = __toESM(require("path"), 1);
var import_lite = require("js-tiktoken/lite");
var import_request = require("koishi-plugin-chatluna/utils/request");
var import_os = __toESM(require("os"), 1);
var import_promises = __toESM(require("fs/promises"), 1);
var cache = {};
var tiktokenCache = {};
async function getEncoding(encoding, options) {
  options = options ?? {};
  const cacheDir = import_path.default.resolve(import_os.default.tmpdir(), "chatluna", "tiktoken");
  const cachePath = import_path.default.join(cacheDir, `${encoding}.json`);
  if (tiktokenCache[encoding]) {
    return tiktokenCache[encoding];
  }
  if (cache[encoding]) {
    const tiktoken2 = new import_lite.Tiktoken(
      cache[encoding],
      options?.extendedSpecialTokens
    );
    tiktokenCache[encoding] = tiktoken2;
    return tiktoken2;
  }
  await import_promises.default.mkdir(cacheDir, { recursive: true });
  try {
    const cacheContent = await import_promises.default.readFile(cachePath, "utf-8");
    cache[encoding] = JSON.parse(cacheContent);
    const tiktoken2 = new import_lite.Tiktoken(
      cache[encoding],
      options?.extendedSpecialTokens
    );
    tiktokenCache[encoding] = tiktoken2;
    return tiktoken2;
  } catch (e) {
  }
  const url = (import_request.globalProxyAddress?.length ?? 0) > 0 ? `https://tiktoken.pages.dev/js/${encoding}.json` : `https://jsd.onmicrosoft.cn/npm/tiktoken@latest/encoders/${encoding}.json`;
  cache[encoding] = await (0, import_request.chatLunaFetch)(url).then((res) => {
    if (!res.ok) {
      throw new Error(
        `Failed to fetch tiktoken encoding: ${res.status}`
      );
    }
    return res.json();
  }).catch((e) => {
    delete cache[encoding];
    throw e;
  });
  await import_promises.default.writeFile(cachePath, JSON.stringify(cache[encoding]));
  const tiktoken = new import_lite.Tiktoken(
    cache[encoding],
    options?.extendedSpecialTokens
  );
  tiktokenCache[encoding] = tiktoken;
  return tiktoken;
}
__name(getEncoding, "getEncoding");

// src/llm-core/platform/usage.ts
function createModelUsageReporter(ctx, platform, model, stack) {
  const report = /* @__PURE__ */ __name(async (usage) => {
    const limit = Error.stackTraceLimit;
    Error.stackTraceLimit = Math.max(limit, 50);
    const currentStack = new Error().stack;
    Error.stackTraceLimit = limit;
    const source = (0, import_usage_source.usageSourceFromStack)(currentStack);
    const fallback = source === "chatluna" || source === "unknown" ? (0, import_usage_source.usageSourceFromStack)(stack) : source;
    const payload = {
      ...usage,
      source: fallback === "unknown" ? source : fallback,
      platform: usage.platform ?? platform,
      model: usage.model ?? model,
      usageMetadata: usage.usageMetadata ?? {
        input_tokens: 0,
        output_tokens: 0,
        total_tokens: 0
      },
      estimated: usage.estimated ?? false,
      success: usage.success ?? true,
      createdAt: usage.createdAt ?? /* @__PURE__ */ new Date()
    };
    try {
      await ctx.root.parallel("chatluna/model-usage", payload);
    } catch (e) {
      ctx.logger.error(e);
    }
  }, "report");
  return report;
}
__name(createModelUsageReporter, "createModelUsageReporter");
var encoder = null;
var warmup = null;
var failed = false;
function warmupTokenEncoder() {
  if (encoder != null || failed) return Promise.resolve();
  if (warmup != null) return warmup;
  warmup = getEncoding("cl100k_base").then((e) => {
    encoder = e;
  }).catch((e) => {
    failed = true;
    import_koishi_plugin_chatluna.logger.warn("tiktoken init failed; falling back to heuristic", e);
  }).finally(() => {
    warmup = null;
  });
  return warmup;
}
__name(warmupTokenEncoder, "warmupTokenEncoder");
async function estimateTextTokens(input) {
  const text = Array.isArray(input) ? input.join("\n") : input;
  if (encoder == null && !failed) await warmupTokenEncoder();
  if (encoder != null) {
    try {
      return encoder.encode(text).length;
    } catch {
    }
  }
  let count = 0;
  for (const char of text) {
    count += char.charCodeAt(0) <= 127 ? 0.25 : 2 / 3;
  }
  return Math.ceil(count);
}
__name(estimateTextTokens, "estimateTextTokens");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createModelUsageReporter,
  estimateTextTokens,
  warmupTokenEncoder
});
