// koishi-plugin-chatluna/lib/llm-core/platform/usage.mjs
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/llm-core/platform/usage.ts
import { logger } from "koishi-plugin-chatluna";
import { usageSourceFromStack } from "koishi-plugin-chatluna/utils/usage_source";

// src/llm-core/utils/tiktoken.ts
import path from "path";
import {
  getEncodingNameForModel,
  Tiktoken
} from "js-tiktoken/lite";
import {
  chatLunaFetch,
  globalProxyAddress
} from "koishi-plugin-chatluna/utils/request";
import os from "os";
import fs from "fs/promises";
var cache = {};
var tiktokenCache = {};
async function getEncoding(encoding, options) {
  options = options ?? {};
  const cacheDir = path.resolve(os.tmpdir(), "chatluna", "tiktoken");
  const cachePath = path.join(cacheDir, `${encoding}.json`);
  if (tiktokenCache[encoding]) {
    return tiktokenCache[encoding];
  }
  if (cache[encoding]) {
    const tiktoken2 = new Tiktoken(
      cache[encoding],
      options?.extendedSpecialTokens
    );
    tiktokenCache[encoding] = tiktoken2;
    return tiktoken2;
  }
  await fs.mkdir(cacheDir, { recursive: true });
  try {
    const cacheContent = await fs.readFile(cachePath, "utf-8");
    cache[encoding] = JSON.parse(cacheContent);
    const tiktoken2 = new Tiktoken(
      cache[encoding],
      options?.extendedSpecialTokens
    );
    tiktokenCache[encoding] = tiktoken2;
    return tiktoken2;
  } catch (e) {
  }
  const url = (globalProxyAddress?.length ?? 0) > 0 ? `https://tiktoken.pages.dev/js/${encoding}.json` : `https://jsd.onmicrosoft.cn/npm/tiktoken@latest/encoders/${encoding}.json`;
  cache[encoding] = await chatLunaFetch(url).then((res) => {
    if (!res.ok) {
      throw new Error(
        `Failed to fetch tiktoken encoding: ${res.status}`
      );
    }
    return res.json();
  }).catch((e) => {
    delete cache[encoding];
    throw e;
  });
  await fs.writeFile(cachePath, JSON.stringify(cache[encoding]));
  const tiktoken = new Tiktoken(
    cache[encoding],
    options?.extendedSpecialTokens
  );
  tiktokenCache[encoding] = tiktoken;
  return tiktoken;
}
__name(getEncoding, "getEncoding");

// src/llm-core/platform/usage.ts
function createModelUsageReporter(ctx, platform, model, stack) {
  const report = /* @__PURE__ */ __name(async (usage) => {
    const limit = Error.stackTraceLimit;
    Error.stackTraceLimit = Math.max(limit, 50);
    const currentStack = new Error().stack;
    Error.stackTraceLimit = limit;
    const source = usageSourceFromStack(currentStack);
    const fallback = source === "chatluna" || source === "unknown" ? usageSourceFromStack(stack) : source;
    const payload = {
      ...usage,
      source: fallback === "unknown" ? source : fallback,
      platform: usage.platform ?? platform,
      model: usage.model ?? model,
      usageMetadata: usage.usageMetadata ?? {
        input_tokens: 0,
        output_tokens: 0,
        total_tokens: 0
      },
      estimated: usage.estimated ?? false,
      success: usage.success ?? true,
      createdAt: usage.createdAt ?? /* @__PURE__ */ new Date()
    };
    try {
      await ctx.root.parallel("chatluna/model-usage", payload);
    } catch (e) {
      ctx.logger.error(e);
    }
  }, "report");
  return report;
}
__name(createModelUsageReporter, "createModelUsageReporter");
var encoder = null;
var warmup = null;
var failed = false;
function warmupTokenEncoder() {
  if (encoder != null || failed) return Promise.resolve();
  if (warmup != null) return warmup;
  warmup = getEncoding("cl100k_base").then((e) => {
    encoder = e;
  }).catch((e) => {
    failed = true;
    logger.warn("tiktoken init failed; falling back to heuristic", e);
  }).finally(() => {
    warmup = null;
  });
  return warmup;
}
__name(warmupTokenEncoder, "warmupTokenEncoder");
async function estimateTextTokens(input) {
  const text = Array.isArray(input) ? input.join("\n") : input;
  if (encoder == null && !failed) await warmupTokenEncoder();
  if (encoder != null) {
    try {
      return encoder.encode(text).length;
    } catch {
    }
  }
  let count = 0;
  for (const char of text) {
    count += char.charCodeAt(0) <= 127 ? 0.25 : 2 / 3;
  }
  return Math.ceil(count);
}
__name(estimateTextTokens, "estimateTextTokens");
export {
  createModelUsageReporter,
  estimateTextTokens,
  warmupTokenEncoder
};
